<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>sis package &mdash; Spots-in-Spaaaaaace! 0.0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=d45e8c67"></script>
        <script src="_static/doctools.js?v=888ff710"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            Spots-in-Spaaaaaace!
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li><a class="reference internal" href="#">sis package</a><ul>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-sis.celltype_mapping">sis.celltype_mapping module</a><ul>
<li><a class="reference internal" href="#sis.celltype_mapping.CKMapping"><code class="docutils literal notranslate"><span class="pre">CKMapping</span></code></a><ul>
<li><a class="reference internal" href="#sis.celltype_mapping.CKMapping.plot_best_mapping"><code class="docutils literal notranslate"><span class="pre">CKMapping.plot_best_mapping()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.CKMapping.plot_best_mapping_corr"><code class="docutils literal notranslate"><span class="pre">CKMapping.plot_best_mapping_corr()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.CKMapping.plot_best_mapping_probability"><code class="docutils literal notranslate"><span class="pre">CKMapping.plot_best_mapping_probability()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.CKMapping.plot_mapping_performance"><code class="docutils literal notranslate"><span class="pre">CKMapping.plot_mapping_performance()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.celltype_mapping.CellTypeMapping"><code class="docutils literal notranslate"><span class="pre">CellTypeMapping</span></code></a><ul>
<li><a class="reference internal" href="#sis.celltype_mapping.CellTypeMapping.load_from_timestamp"><code class="docutils literal notranslate"><span class="pre">CellTypeMapping.load_from_timestamp()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.CellTypeMapping.mapping"><code class="docutils literal notranslate"><span class="pre">CellTypeMapping.mapping()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.CellTypeMapping.ordered_labels"><code class="docutils literal notranslate"><span class="pre">CellTypeMapping.ordered_labels</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.CellTypeMapping.qc_mapping"><code class="docutils literal notranslate"><span class="pre">CellTypeMapping.qc_mapping()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.CellTypeMapping.reset_qc"><code class="docutils literal notranslate"><span class="pre">CellTypeMapping.reset_qc()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.CellTypeMapping.save_mapping"><code class="docutils literal notranslate"><span class="pre">CellTypeMapping.save_mapping()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.CellTypeMapping.spatial_umap"><code class="docutils literal notranslate"><span class="pre">CellTypeMapping.spatial_umap()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.celltype_mapping.ScrattchMapping"><code class="docutils literal notranslate"><span class="pre">ScrattchMapping</span></code></a><ul>
<li><a class="reference internal" href="#sis.celltype_mapping.ScrattchMapping.get_confusion_matrix"><code class="docutils literal notranslate"><span class="pre">ScrattchMapping.get_confusion_matrix()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.ScrattchMapping.load_scrattch_mapping_results"><code class="docutils literal notranslate"><span class="pre">ScrattchMapping.load_scrattch_mapping_results()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.ScrattchMapping.load_taxonomy_anndata"><code class="docutils literal notranslate"><span class="pre">ScrattchMapping.load_taxonomy_anndata()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.ScrattchMapping.make_mapping_anndata"><code class="docutils literal notranslate"><span class="pre">ScrattchMapping.make_mapping_anndata()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.ScrattchMapping.plot_best_mapping"><code class="docutils literal notranslate"><span class="pre">ScrattchMapping.plot_best_mapping()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.ScrattchMapping.plot_best_mapping_corr"><code class="docutils literal notranslate"><span class="pre">ScrattchMapping.plot_best_mapping_corr()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.ScrattchMapping.plot_class_proportions"><code class="docutils literal notranslate"><span class="pre">ScrattchMapping.plot_class_proportions()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.ScrattchMapping.run_on_hpc"><code class="docutils literal notranslate"><span class="pre">ScrattchMapping.run_on_hpc()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.celltype_mapping.TangramMapping"><code class="docutils literal notranslate"><span class="pre">TangramMapping</span></code></a><ul>
<li><a class="reference internal" href="#sis.celltype_mapping.TangramMapping.compare_spatial_gene_exp"><code class="docutils literal notranslate"><span class="pre">TangramMapping.compare_spatial_gene_exp()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.TangramMapping.evaluate_mapping"><code class="docutils literal notranslate"><span class="pre">TangramMapping.evaluate_mapping()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.TangramMapping.get_discrete_cell_mapping"><code class="docutils literal notranslate"><span class="pre">TangramMapping.get_discrete_cell_mapping()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.TangramMapping.mapping"><code class="docutils literal notranslate"><span class="pre">TangramMapping.mapping()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.TangramMapping.plot_cell_probability"><code class="docutils literal notranslate"><span class="pre">TangramMapping.plot_cell_probability()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.TangramMapping.plot_discrete_mapping"><code class="docutils literal notranslate"><span class="pre">TangramMapping.plot_discrete_mapping()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.TangramMapping.plot_discrete_mapping_probability"><code class="docutils literal notranslate"><span class="pre">TangramMapping.plot_discrete_mapping_probability()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.TangramMapping.project_genes"><code class="docutils literal notranslate"><span class="pre">TangramMapping.project_genes()</span></code></a></li>
<li><a class="reference internal" href="#sis.celltype_mapping.TangramMapping.set_training_genes"><code class="docutils literal notranslate"><span class="pre">TangramMapping.set_training_genes()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#module-sis.expression_dataset">sis.expression_dataset module</a><ul>
<li><a class="reference internal" href="#sis.expression_dataset.ExpressionDataset"><code class="docutils literal notranslate"><span class="pre">ExpressionDataset</span></code></a><ul>
<li><a class="reference internal" href="#sis.expression_dataset.ExpressionDataset.copy"><code class="docutils literal notranslate"><span class="pre">ExpressionDataset.copy()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.ExpressionDataset.genes"><code class="docutils literal notranslate"><span class="pre">ExpressionDataset.genes</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.ExpressionDataset.load_arrow"><code class="docutils literal notranslate"><span class="pre">ExpressionDataset.load_arrow()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.ExpressionDataset.load_from_timestamp"><code class="docutils literal notranslate"><span class="pre">ExpressionDataset.load_from_timestamp()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.ExpressionDataset.normalize"><code class="docutils literal notranslate"><span class="pre">ExpressionDataset.normalize()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.ExpressionDataset.select_genes"><code class="docutils literal notranslate"><span class="pre">ExpressionDataset.select_genes()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.ExpressionDataset.subsample_by_type"><code class="docutils literal notranslate"><span class="pre">ExpressionDataset.subsample_by_type()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.expression_dataset.GeneBasisMethod"><code class="docutils literal notranslate"><span class="pre">GeneBasisMethod</span></code></a><ul>
<li><a class="reference internal" href="#sis.expression_dataset.GeneBasisMethod.celltype_mapping"><code class="docutils literal notranslate"><span class="pre">GeneBasisMethod.celltype_mapping()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.GeneBasisMethod.df_to_sce"><code class="docutils literal notranslate"><span class="pre">GeneBasisMethod.df_to_sce()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.GeneBasisMethod.filter_genes"><code class="docutils literal notranslate"><span class="pre">GeneBasisMethod.filter_genes()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.GeneBasisMethod.load_gene_panel"><code class="docutils literal notranslate"><span class="pre">GeneBasisMethod.load_gene_panel()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.GeneBasisMethod.neighborhood_score"><code class="docutils literal notranslate"><span class="pre">GeneBasisMethod.neighborhood_score()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.GeneBasisMethod.panel_eval"><code class="docutils literal notranslate"><span class="pre">GeneBasisMethod.panel_eval()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.GeneBasisMethod.raw_to_sce"><code class="docutils literal notranslate"><span class="pre">GeneBasisMethod.raw_to_sce()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.GeneBasisMethod.run_on_hpc"><code class="docutils literal notranslate"><span class="pre">GeneBasisMethod.run_on_hpc()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.GeneBasisMethod.save_to_csv"><code class="docutils literal notranslate"><span class="pre">GeneBasisMethod.save_to_csv()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.GeneBasisMethod.select_gene_panel"><code class="docutils literal notranslate"><span class="pre">GeneBasisMethod.select_gene_panel()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.expression_dataset.GenePanelMethod"><code class="docutils literal notranslate"><span class="pre">GenePanelMethod</span></code></a><ul>
<li><a class="reference internal" href="#sis.expression_dataset.GenePanelMethod.evaluate_gene_panel"><code class="docutils literal notranslate"><span class="pre">GenePanelMethod.evaluate_gene_panel()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.GenePanelMethod.select_gene_panel"><code class="docutils literal notranslate"><span class="pre">GenePanelMethod.select_gene_panel()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.expression_dataset.GenePanelSelection"><code class="docutils literal notranslate"><span class="pre">GenePanelSelection</span></code></a><ul>
<li><a class="reference internal" href="#sis.expression_dataset.GenePanelSelection.eval_method"><code class="docutils literal notranslate"><span class="pre">GenePanelSelection.eval_method()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.GenePanelSelection.expression_dataset"><code class="docutils literal notranslate"><span class="pre">GenePanelSelection.expression_dataset()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.GenePanelSelection.get_confusion_matrix"><code class="docutils literal notranslate"><span class="pre">GenePanelSelection.get_confusion_matrix()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.GenePanelSelection.get_fraction_mapped"><code class="docutils literal notranslate"><span class="pre">GenePanelSelection.get_fraction_mapped()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.GenePanelSelection.load_from_timestamp"><code class="docutils literal notranslate"><span class="pre">GenePanelSelection.load_from_timestamp()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.GenePanelSelection.report"><code class="docutils literal notranslate"><span class="pre">GenePanelSelection.report()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.GenePanelSelection.save_selection"><code class="docutils literal notranslate"><span class="pre">GenePanelSelection.save_selection()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.expression_dataset.PROPOSEMethod"><code class="docutils literal notranslate"><span class="pre">PROPOSEMethod</span></code></a><ul>
<li><a class="reference internal" href="#sis.expression_dataset.PROPOSEMethod.select_gene_panel"><code class="docutils literal notranslate"><span class="pre">PROPOSEMethod.select_gene_panel()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.expression_dataset.ScranMethod"><code class="docutils literal notranslate"><span class="pre">ScranMethod</span></code></a><ul>
<li><a class="reference internal" href="#sis.expression_dataset.ScranMethod.select_gene_panel"><code class="docutils literal notranslate"><span class="pre">ScranMethod.select_gene_panel()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.expression_dataset.SeuratMethod"><code class="docutils literal notranslate"><span class="pre">SeuratMethod</span></code></a><ul>
<li><a class="reference internal" href="#sis.expression_dataset.SeuratMethod.select_gene_panel"><code class="docutils literal notranslate"><span class="pre">SeuratMethod.select_gene_panel()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.expression_dataset.gene_basis_panel_eval"><code class="docutils literal notranslate"><span class="pre">gene_basis_panel_eval()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.importr"><code class="docutils literal notranslate"><span class="pre">importr()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.load_mfishtools_fraction_mapped"><code class="docutils literal notranslate"><span class="pre">load_mfishtools_fraction_mapped()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.load_panel_eval"><code class="docutils literal notranslate"><span class="pre">load_panel_eval()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.mFISHtoolsMethod"><code class="docutils literal notranslate"><span class="pre">mFISHtoolsMethod</span></code></a><ul>
<li><a class="reference internal" href="#sis.expression_dataset.mFISHtoolsMethod.build_panel_ad"><code class="docutils literal notranslate"><span class="pre">mFISHtoolsMethod.build_panel_ad()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.mFISHtoolsMethod.filter_genes"><code class="docutils literal notranslate"><span class="pre">mFISHtoolsMethod.filter_genes()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.mFISHtoolsMethod.load_gene_panel"><code class="docutils literal notranslate"><span class="pre">mFISHtoolsMethod.load_gene_panel()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.mFISHtoolsMethod.run_on_hpc"><code class="docutils literal notranslate"><span class="pre">mFISHtoolsMethod.run_on_hpc()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.mFISHtoolsMethod.select_gene_panel"><code class="docutils literal notranslate"><span class="pre">mFISHtoolsMethod.select_gene_panel()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.expression_dataset.marker_gene_eval"><code class="docutils literal notranslate"><span class="pre">marker_gene_eval()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.norm_confusion_matrix"><code class="docutils literal notranslate"><span class="pre">norm_confusion_matrix()</span></code></a></li>
<li><a class="reference internal" href="#sis.expression_dataset.run_mfishtools_fraction_mapped"><code class="docutils literal notranslate"><span class="pre">run_mfishtools_fraction_mapped()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-sis.hpc">sis.hpc module</a><ul>
<li><a class="reference internal" href="#sis.hpc.JobState"><code class="docutils literal notranslate"><span class="pre">JobState</span></code></a><ul>
<li><a class="reference internal" href="#sis.hpc.JobState.state_codes"><code class="docutils literal notranslate"><span class="pre">JobState.state_codes</span></code></a></li>
<li><a class="reference internal" href="#sis.hpc.JobState.state_descriptions"><code class="docutils literal notranslate"><span class="pre">JobState.state_descriptions</span></code></a></li>
<li><a class="reference internal" href="#sis.hpc.JobState.state_to_code"><code class="docutils literal notranslate"><span class="pre">JobState.state_to_code</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.hpc.SlurmJob"><code class="docutils literal notranslate"><span class="pre">SlurmJob</span></code></a><ul>
<li><a class="reference internal" href="#sis.hpc.SlurmJob.cancel"><code class="docutils literal notranslate"><span class="pre">SlurmJob.cancel()</span></code></a></li>
<li><a class="reference internal" href="#sis.hpc.SlurmJob.error"><code class="docutils literal notranslate"><span class="pre">SlurmJob.error</span></code></a></li>
<li><a class="reference internal" href="#sis.hpc.SlurmJob.error_file"><code class="docutils literal notranslate"><span class="pre">SlurmJob.error_file</span></code></a></li>
<li><a class="reference internal" href="#sis.hpc.SlurmJob.is_done"><code class="docutils literal notranslate"><span class="pre">SlurmJob.is_done()</span></code></a></li>
<li><a class="reference internal" href="#sis.hpc.SlurmJob.output"><code class="docutils literal notranslate"><span class="pre">SlurmJob.output</span></code></a></li>
<li><a class="reference internal" href="#sis.hpc.SlurmJob.output_file"><code class="docutils literal notranslate"><span class="pre">SlurmJob.output_file</span></code></a></li>
<li><a class="reference internal" href="#sis.hpc.SlurmJob.state"><code class="docutils literal notranslate"><span class="pre">SlurmJob.state()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.hpc.SlurmJobArray"><code class="docutils literal notranslate"><span class="pre">SlurmJobArray</span></code></a><ul>
<li><a class="reference internal" href="#sis.hpc.SlurmJobArray.cancel"><code class="docutils literal notranslate"><span class="pre">SlurmJobArray.cancel()</span></code></a></li>
<li><a class="reference internal" href="#sis.hpc.SlurmJobArray.finished_jobs"><code class="docutils literal notranslate"><span class="pre">SlurmJobArray.finished_jobs()</span></code></a></li>
<li><a class="reference internal" href="#sis.hpc.SlurmJobArray.is_done"><code class="docutils literal notranslate"><span class="pre">SlurmJobArray.is_done()</span></code></a></li>
<li><a class="reference internal" href="#sis.hpc.SlurmJobArray.state"><code class="docutils literal notranslate"><span class="pre">SlurmJobArray.state()</span></code></a></li>
<li><a class="reference internal" href="#sis.hpc.SlurmJobArray.state_counts"><code class="docutils literal notranslate"><span class="pre">SlurmJobArray.state_counts()</span></code></a></li>
<li><a class="reference internal" href="#sis.hpc.SlurmJobArray.unfinished_jobs"><code class="docutils literal notranslate"><span class="pre">SlurmJobArray.unfinished_jobs()</span></code></a></li>
<li><a class="reference internal" href="#sis.hpc.SlurmJobArray.wait_iter"><code class="docutils literal notranslate"><span class="pre">SlurmJobArray.wait_iter()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.hpc.close_ssh_connections"><code class="docutils literal notranslate"><span class="pre">close_ssh_connections()</span></code></a></li>
<li><a class="reference internal" href="#sis.hpc.hpc_worker"><code class="docutils literal notranslate"><span class="pre">hpc_worker()</span></code></a></li>
<li><a class="reference internal" href="#sis.hpc.run"><code class="docutils literal notranslate"><span class="pre">run()</span></code></a></li>
<li><a class="reference internal" href="#sis.hpc.run_slurm"><code class="docutils literal notranslate"><span class="pre">run_slurm()</span></code></a></li>
<li><a class="reference internal" href="#sis.hpc.run_slurm_func"><code class="docutils literal notranslate"><span class="pre">run_slurm_func()</span></code></a></li>
<li><a class="reference internal" href="#sis.hpc.sacct"><code class="docutils literal notranslate"><span class="pre">sacct()</span></code></a></li>
<li><a class="reference internal" href="#sis.hpc.squeue"><code class="docutils literal notranslate"><span class="pre">squeue()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-sis.image">sis.image module</a><ul>
<li><a class="reference internal" href="#sis.image.Image"><code class="docutils literal notranslate"><span class="pre">Image</span></code></a><ul>
<li><a class="reference internal" href="#sis.image.Image.get_data"><code class="docutils literal notranslate"><span class="pre">Image.get_data()</span></code></a></li>
<li><a class="reference internal" href="#sis.image.Image.get_sub_data"><code class="docutils literal notranslate"><span class="pre">Image.get_sub_data()</span></code></a></li>
<li><a class="reference internal" href="#sis.image.Image.shape"><code class="docutils literal notranslate"><span class="pre">Image.shape</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.image.ImageBase"><code class="docutils literal notranslate"><span class="pre">ImageBase</span></code></a><ul>
<li><a class="reference internal" href="#sis.image.ImageBase.bounds"><code class="docutils literal notranslate"><span class="pre">ImageBase.bounds()</span></code></a></li>
<li><a class="reference internal" href="#sis.image.ImageBase.get_channel"><code class="docutils literal notranslate"><span class="pre">ImageBase.get_channel()</span></code></a></li>
<li><a class="reference internal" href="#sis.image.ImageBase.get_data"><code class="docutils literal notranslate"><span class="pre">ImageBase.get_data()</span></code></a></li>
<li><a class="reference internal" href="#sis.image.ImageBase.get_frame"><code class="docutils literal notranslate"><span class="pre">ImageBase.get_frame()</span></code></a></li>
<li><a class="reference internal" href="#sis.image.ImageBase.get_pixel_subregion"><code class="docutils literal notranslate"><span class="pre">ImageBase.get_pixel_subregion()</span></code></a></li>
<li><a class="reference internal" href="#sis.image.ImageBase.get_subregion"><code class="docutils literal notranslate"><span class="pre">ImageBase.get_subregion()</span></code></a></li>
<li><a class="reference internal" href="#sis.image.ImageBase.shape"><code class="docutils literal notranslate"><span class="pre">ImageBase.shape</span></code></a></li>
<li><a class="reference internal" href="#sis.image.ImageBase.show"><code class="docutils literal notranslate"><span class="pre">ImageBase.show()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.image.ImageFile"><code class="docutils literal notranslate"><span class="pre">ImageFile</span></code></a><ul>
<li><a class="reference internal" href="#sis.image.ImageFile.get_data"><code class="docutils literal notranslate"><span class="pre">ImageFile.get_data()</span></code></a></li>
<li><a class="reference internal" href="#sis.image.ImageFile.get_sub_data"><code class="docutils literal notranslate"><span class="pre">ImageFile.get_sub_data()</span></code></a></li>
<li><a class="reference internal" href="#sis.image.ImageFile.load_merscope"><code class="docutils literal notranslate"><span class="pre">ImageFile.load_merscope()</span></code></a></li>
<li><a class="reference internal" href="#sis.image.ImageFile.shape"><code class="docutils literal notranslate"><span class="pre">ImageFile.shape</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.image.ImageStack"><code class="docutils literal notranslate"><span class="pre">ImageStack</span></code></a><ul>
<li><a class="reference internal" href="#sis.image.ImageStack.channels"><code class="docutils literal notranslate"><span class="pre">ImageStack.channels</span></code></a></li>
<li><a class="reference internal" href="#sis.image.ImageStack.get_data"><code class="docutils literal notranslate"><span class="pre">ImageStack.get_data()</span></code></a></li>
<li><a class="reference internal" href="#sis.image.ImageStack.get_sub_data"><code class="docutils literal notranslate"><span class="pre">ImageStack.get_sub_data()</span></code></a></li>
<li><a class="reference internal" href="#sis.image.ImageStack.load_merscope_stacks"><code class="docutils literal notranslate"><span class="pre">ImageStack.load_merscope_stacks()</span></code></a></li>
<li><a class="reference internal" href="#sis.image.ImageStack.shape"><code class="docutils literal notranslate"><span class="pre">ImageStack.shape</span></code></a></li>
<li><a class="reference internal" href="#sis.image.ImageStack.transform"><code class="docutils literal notranslate"><span class="pre">ImageStack.transform</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.image.ImageTransform"><code class="docutils literal notranslate"><span class="pre">ImageTransform</span></code></a><ul>
<li><a class="reference internal" href="#sis.image.ImageTransform.inverse_matrix"><code class="docutils literal notranslate"><span class="pre">ImageTransform.inverse_matrix</span></code></a></li>
<li><a class="reference internal" href="#sis.image.ImageTransform.map_from_pixels"><code class="docutils literal notranslate"><span class="pre">ImageTransform.map_from_pixels()</span></code></a></li>
<li><a class="reference internal" href="#sis.image.ImageTransform.map_to_pixels"><code class="docutils literal notranslate"><span class="pre">ImageTransform.map_to_pixels()</span></code></a></li>
<li><a class="reference internal" href="#sis.image.ImageTransform.translated"><code class="docutils literal notranslate"><span class="pre">ImageTransform.translated()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.image.ImageView"><code class="docutils literal notranslate"><span class="pre">ImageView</span></code></a><ul>
<li><a class="reference internal" href="#sis.image.ImageView.channels"><code class="docutils literal notranslate"><span class="pre">ImageView.channels</span></code></a></li>
<li><a class="reference internal" href="#sis.image.ImageView.get_data"><code class="docutils literal notranslate"><span class="pre">ImageView.get_data()</span></code></a></li>
<li><a class="reference internal" href="#sis.image.ImageView.get_sub_data"><code class="docutils literal notranslate"><span class="pre">ImageView.get_sub_data()</span></code></a></li>
<li><a class="reference internal" href="#sis.image.ImageView.name"><code class="docutils literal notranslate"><span class="pre">ImageView.name</span></code></a></li>
<li><a class="reference internal" href="#sis.image.ImageView.shape"><code class="docutils literal notranslate"><span class="pre">ImageView.shape</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#module-sis.optional_import">sis.optional_import module</a><ul>
<li><a class="reference internal" href="#sis.optional_import.OptionalImportError"><code class="docutils literal notranslate"><span class="pre">OptionalImportError</span></code></a></li>
<li><a class="reference internal" href="#sis.optional_import.optional_import"><code class="docutils literal notranslate"><span class="pre">optional_import()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-sis.qc">sis.qc module</a><ul>
<li><a class="reference internal" href="#sis.qc.calc_n_blanks"><code class="docutils literal notranslate"><span class="pre">calc_n_blanks()</span></code></a></li>
<li><a class="reference internal" href="#sis.qc.calc_n_genes"><code class="docutils literal notranslate"><span class="pre">calc_n_genes()</span></code></a></li>
<li><a class="reference internal" href="#sis.qc.calc_n_transcripts"><code class="docutils literal notranslate"><span class="pre">calc_n_transcripts()</span></code></a></li>
<li><a class="reference internal" href="#sis.qc.get_counts_from_cbg"><code class="docutils literal notranslate"><span class="pre">get_counts_from_cbg()</span></code></a></li>
<li><a class="reference internal" href="#sis.qc.run_doublet_detection"><code class="docutils literal notranslate"><span class="pre">run_doublet_detection()</span></code></a></li>
<li><a class="reference internal" href="#sis.qc.run_solo_doublet_detection"><code class="docutils literal notranslate"><span class="pre">run_solo_doublet_detection()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-sis.segmentation">sis.segmentation module</a><ul>
<li><a class="reference internal" href="#sis.segmentation.BaysorSegmentationMethod"><code class="docutils literal notranslate"><span class="pre">BaysorSegmentationMethod</span></code></a><ul>
<li><a class="reference internal" href="#sis.segmentation.BaysorSegmentationMethod.run"><code class="docutils literal notranslate"><span class="pre">BaysorSegmentationMethod.run()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.segmentation.BaysorSegmentationResult"><code class="docutils literal notranslate"><span class="pre">BaysorSegmentationResult</span></code></a><ul>
<li><a class="reference internal" href="#sis.segmentation.BaysorSegmentationResult.cell_ids"><code class="docutils literal notranslate"><span class="pre">BaysorSegmentationResult.cell_ids</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.segmentation.CellposeSegmentationMethod"><code class="docutils literal notranslate"><span class="pre">CellposeSegmentationMethod</span></code></a><ul>
<li><a class="reference internal" href="#sis.segmentation.CellposeSegmentationMethod.get_total_mrna_image"><code class="docutils literal notranslate"><span class="pre">CellposeSegmentationMethod.get_total_mrna_image()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.CellposeSegmentationMethod.map_spots_to_img_px"><code class="docutils literal notranslate"><span class="pre">CellposeSegmentationMethod.map_spots_to_img_px()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.CellposeSegmentationMethod.run"><code class="docutils literal notranslate"><span class="pre">CellposeSegmentationMethod.run()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.segmentation.CellposeSegmentationResult"><code class="docutils literal notranslate"><span class="pre">CellposeSegmentationResult</span></code></a><ul>
<li><a class="reference internal" href="#sis.segmentation.CellposeSegmentationResult.cell_ids"><code class="docutils literal notranslate"><span class="pre">CellposeSegmentationResult.cell_ids</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.CellposeSegmentationResult.mask_image"><code class="docutils literal notranslate"><span class="pre">CellposeSegmentationResult.mask_image</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.CellposeSegmentationResult.show"><code class="docutils literal notranslate"><span class="pre">CellposeSegmentationResult.show()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.segmentation.MerscopeSegmentationRun"><code class="docutils literal notranslate"><span class="pre">MerscopeSegmentationRun</span></code></a><ul>
<li><a class="reference internal" href="#sis.segmentation.MerscopeSegmentationRun.get_load_args"><code class="docutils literal notranslate"><span class="pre">MerscopeSegmentationRun.get_load_args()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.MerscopeSegmentationRun.get_load_func"><code class="docutils literal notranslate"><span class="pre">MerscopeSegmentationRun.get_load_func()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationMethod"><code class="docutils literal notranslate"><span class="pre">SegmentationMethod</span></code></a><ul>
<li><a class="reference internal" href="#sis.segmentation.SegmentationMethod.run"><code class="docutils literal notranslate"><span class="pre">SegmentationMethod.run()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationResult"><code class="docutils literal notranslate"><span class="pre">SegmentationResult</span></code></a><ul>
<li><a class="reference internal" href="#sis.segmentation.SegmentationResult.cell_ids"><code class="docutils literal notranslate"><span class="pre">SegmentationResult.cell_ids</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationResult.save"><code class="docutils literal notranslate"><span class="pre">SegmentationResult.save()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationResult.spot_table"><code class="docutils literal notranslate"><span class="pre">SegmentationResult.spot_table()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun"><code class="docutils literal notranslate"><span class="pre">SegmentationRun</span></code></a><ul>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.clean_up"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.clean_up()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.create_cell_by_gene"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.create_cell_by_gene()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.from_json"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.from_json()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.from_spatial_dataset"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.from_spatial_dataset()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.get_load_args"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.get_load_args()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.get_load_func"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.get_load_func()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.get_polygon_run_spec"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.get_polygon_run_spec()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.get_seg_run_spec"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.get_seg_run_spec()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.load_cbg"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.load_cbg()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.load_cell_ids"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.load_cell_ids()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.load_metadata"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.load_metadata()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.load_regions"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.load_regions()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.load_results"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.load_results()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.load_run_spec"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.load_run_spec()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.load_spot_table"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.load_spot_table()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.merge_cell_polygons"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.merge_cell_polygons()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.merge_segmented_tiles"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.merge_segmented_tiles()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.resume"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.resume()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.run"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.run()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.save_cbg"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.save_cbg()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.save_cell_ids"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.save_cell_ids()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.save_metadata"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.save_metadata()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.save_regions"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.save_regions()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.save_run_spec"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.save_run_spec()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.set_intermediate_file_paths"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.set_intermediate_file_paths()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.submit_jobs"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.submit_jobs()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.tile_seg_region"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.tile_seg_region()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.track_job_progress"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.track_job_progress()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.SegmentationRun.update_metadata"><code class="docutils literal notranslate"><span class="pre">SegmentationRun.update_metadata()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.segmentation.StereoSeqSegmentationRun"><code class="docutils literal notranslate"><span class="pre">StereoSeqSegmentationRun</span></code></a><ul>
<li><a class="reference internal" href="#sis.segmentation.StereoSeqSegmentationRun.get_load_args"><code class="docutils literal notranslate"><span class="pre">StereoSeqSegmentationRun.get_load_args()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.StereoSeqSegmentationRun.get_load_func"><code class="docutils literal notranslate"><span class="pre">StereoSeqSegmentationRun.get_load_func()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.segmentation.dilate_labels"><code class="docutils literal notranslate"><span class="pre">dilate_labels()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.load_baysor_result"><code class="docutils literal notranslate"><span class="pre">load_baysor_result()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.merge_segmentation_results"><code class="docutils literal notranslate"><span class="pre">merge_segmentation_results()</span></code></a></li>
<li><a class="reference internal" href="#sis.segmentation.run_segmentation"><code class="docutils literal notranslate"><span class="pre">run_segmentation()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-sis.spatial_dataset">sis.spatial_dataset module</a><ul>
<li><a class="reference internal" href="#sis.spatial_dataset.MERSCOPESection"><code class="docutils literal notranslate"><span class="pre">MERSCOPESection</span></code></a><ul>
<li><a class="reference internal" href="#sis.spatial_dataset.MERSCOPESection.calc_cell_qc_metrics"><code class="docutils literal notranslate"><span class="pre">MERSCOPESection.calc_cell_qc_metrics()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.MERSCOPESection.check_doublet_detection_status"><code class="docutils literal notranslate"><span class="pre">MERSCOPESection.check_doublet_detection_status()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.MERSCOPESection.get_section_data_paths"><code class="docutils literal notranslate"><span class="pre">MERSCOPESection.get_section_data_paths()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.MERSCOPESection.get_section_metadata"><code class="docutils literal notranslate"><span class="pre">MERSCOPESection.get_section_metadata()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.MERSCOPESection.load_doublet_anndata"><code class="docutils literal notranslate"><span class="pre">MERSCOPESection.load_doublet_anndata()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.MERSCOPESection.load_segmentation_results"><code class="docutils literal notranslate"><span class="pre">MERSCOPESection.load_segmentation_results()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.MERSCOPESection.load_spottable"><code class="docutils literal notranslate"><span class="pre">MERSCOPESection.load_spottable()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.MERSCOPESection.make_anndata"><code class="docutils literal notranslate"><span class="pre">MERSCOPESection.make_anndata()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.MERSCOPESection.make_cirro"><code class="docutils literal notranslate"><span class="pre">MERSCOPESection.make_cirro()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.MERSCOPESection.run_doublet_detection_on_section"><code class="docutils literal notranslate"><span class="pre">MERSCOPESection.run_doublet_detection_on_section()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.MERSCOPESection.run_mapping_on_section"><code class="docutils literal notranslate"><span class="pre">MERSCOPESection.run_mapping_on_section()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.MERSCOPESection.run_qc_filtering_on_section"><code class="docutils literal notranslate"><span class="pre">MERSCOPESection.run_qc_filtering_on_section()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.MERSCOPESection.run_segmentation_on_section"><code class="docutils literal notranslate"><span class="pre">MERSCOPESection.run_segmentation_on_section()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.spatial_dataset.SpatialDataset"><code class="docutils literal notranslate"><span class="pre">SpatialDataset</span></code></a><ul>
<li><a class="reference internal" href="#sis.spatial_dataset.SpatialDataset.check_segmentation_status"><code class="docutils literal notranslate"><span class="pre">SpatialDataset.check_segmentation_status()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.SpatialDataset.get_analysis_status"><code class="docutils literal notranslate"><span class="pre">SpatialDataset.get_analysis_status()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.SpatialDataset.get_mappings"><code class="docutils literal notranslate"><span class="pre">SpatialDataset.get_mappings()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.SpatialDataset.load_from_barcode"><code class="docutils literal notranslate"><span class="pre">SpatialDataset.load_from_barcode()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.SpatialDataset.save_dataset"><code class="docutils literal notranslate"><span class="pre">SpatialDataset.save_dataset()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.SpatialDataset.show_gene_transcripts"><code class="docutils literal notranslate"><span class="pre">SpatialDataset.show_gene_transcripts()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.SpatialDataset.spatial_corr_to_bulk"><code class="docutils literal notranslate"><span class="pre">SpatialDataset.spatial_corr_to_bulk()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.SpatialDataset.view_mapping_results"><code class="docutils literal notranslate"><span class="pre">SpatialDataset.view_mapping_results()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.spatial_dataset.StereoSeqSection"><code class="docutils literal notranslate"><span class="pre">StereoSeqSection</span></code></a><ul>
<li><a class="reference internal" href="#sis.spatial_dataset.StereoSeqSection.evaluate_qc_metric"><code class="docutils literal notranslate"><span class="pre">StereoSeqSection.evaluate_qc_metric()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.StereoSeqSection.get_section_data_paths"><code class="docutils literal notranslate"><span class="pre">StereoSeqSection.get_section_data_paths()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.StereoSeqSection.get_section_metadata"><code class="docutils literal notranslate"><span class="pre">StereoSeqSection.get_section_metadata()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.StereoSeqSection.load_spottable"><code class="docutils literal notranslate"><span class="pre">StereoSeqSection.load_spottable()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.StereoSeqSection.qc_bin200"><code class="docutils literal notranslate"><span class="pre">StereoSeqSection.qc_bin200()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.StereoSeqSection.qc_corr_to_pair"><code class="docutils literal notranslate"><span class="pre">StereoSeqSection.qc_corr_to_pair()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.StereoSeqSection.qc_gene_expression"><code class="docutils literal notranslate"><span class="pre">StereoSeqSection.qc_gene_expression()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.StereoSeqSection.qc_mapping_results"><code class="docutils literal notranslate"><span class="pre">StereoSeqSection.qc_mapping_results()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.StereoSeqSection.qc_spatial_corr_to_bulk"><code class="docutils literal notranslate"><span class="pre">StereoSeqSection.qc_spatial_corr_to_bulk()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.StereoSeqSection.qc_widget"><code class="docutils literal notranslate"><span class="pre">StereoSeqSection.qc_widget()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.StereoSeqSection.run_mapping_on_section"><code class="docutils literal notranslate"><span class="pre">StereoSeqSection.run_mapping_on_section()</span></code></a></li>
<li><a class="reference internal" href="#sis.spatial_dataset.StereoSeqSection.set_partner_chips"><code class="docutils literal notranslate"><span class="pre">StereoSeqSection.set_partner_chips()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.spatial_dataset.StereoSeqSectionCollection"><code class="docutils literal notranslate"><span class="pre">StereoSeqSectionCollection</span></code></a><ul>
<li><a class="reference internal" href="#sis.spatial_dataset.StereoSeqSectionCollection.get_counts"><code class="docutils literal notranslate"><span class="pre">StereoSeqSectionCollection.get_counts()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#module-sis.spot_table">sis.spot_table module</a><ul>
<li><a class="reference internal" href="#sis.spot_table.SpotTable"><code class="docutils literal notranslate"><span class="pre">SpotTable</span></code></a><ul>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.add_image"><code class="docutils literal notranslate"><span class="pre">SpotTable.add_image()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.binned_expression_counts"><code class="docutils literal notranslate"><span class="pre">SpotTable.binned_expression_counts()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.bounds"><code class="docutils literal notranslate"><span class="pre">SpotTable.bounds()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.calculate_cell_features"><code class="docutils literal notranslate"><span class="pre">SpotTable.calculate_cell_features()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.calculate_cell_polygons"><code class="docutils literal notranslate"><span class="pre">SpotTable.calculate_cell_polygons()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.calculate_optimal_polygon"><code class="docutils literal notranslate"><span class="pre">SpotTable.calculate_optimal_polygon()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.cell_bounds"><code class="docutils literal notranslate"><span class="pre">SpotTable.cell_bounds()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.cell_by_gene_anndata"><code class="docutils literal notranslate"><span class="pre">SpotTable.cell_by_gene_anndata()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.cell_by_gene_dense_matrix"><code class="docutils literal notranslate"><span class="pre">SpotTable.cell_by_gene_dense_matrix()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.cell_by_gene_sparse_matrix"><code class="docutils literal notranslate"><span class="pre">SpotTable.cell_by_gene_sparse_matrix()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.cell_centroids"><code class="docutils literal notranslate"><span class="pre">SpotTable.cell_centroids()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.cell_ids"><code class="docutils literal notranslate"><span class="pre">SpotTable.cell_ids</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.cell_ids_changed"><code class="docutils literal notranslate"><span class="pre">SpotTable.cell_ids_changed()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.cell_indices"><code class="docutils literal notranslate"><span class="pre">SpotTable.cell_indices()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.cell_indices_within_padding"><code class="docutils literal notranslate"><span class="pre">SpotTable.cell_indices_within_padding()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.cell_mask"><code class="docutils literal notranslate"><span class="pre">SpotTable.cell_mask()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.cell_palette"><code class="docutils literal notranslate"><span class="pre">SpotTable.cell_palette()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.cell_polygon"><code class="docutils literal notranslate"><span class="pre">SpotTable.cell_polygon()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.cell_scatter_plot"><code class="docutils literal notranslate"><span class="pre">SpotTable.cell_scatter_plot()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.cells_inside_region"><code class="docutils literal notranslate"><span class="pre">SpotTable.cells_inside_region()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.convert_cell_id"><code class="docutils literal notranslate"><span class="pre">SpotTable.convert_cell_id()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.copy"><code class="docutils literal notranslate"><span class="pre">SpotTable.copy()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.dataframe"><code class="docutils literal notranslate"><span class="pre">SpotTable.dataframe()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.filter_cells"><code class="docutils literal notranslate"><span class="pre">SpotTable.filter_cells()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.gene_indices"><code class="docutils literal notranslate"><span class="pre">SpotTable.gene_indices()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.gene_names"><code class="docutils literal notranslate"><span class="pre">SpotTable.gene_names</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.generate_production_cell_ids"><code class="docutils literal notranslate"><span class="pre">SpotTable.generate_production_cell_ids()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.get_cell_features"><code class="docutils literal notranslate"><span class="pre">SpotTable.get_cell_features()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.get_genes"><code class="docutils literal notranslate"><span class="pre">SpotTable.get_genes()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.get_geojson_collection"><code class="docutils literal notranslate"><span class="pre">SpotTable.get_geojson_collection()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.get_image"><code class="docutils literal notranslate"><span class="pre">SpotTable.get_image()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.get_subregion"><code class="docutils literal notranslate"><span class="pre">SpotTable.get_subregion()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.grid_tiles"><code class="docutils literal notranslate"><span class="pre">SpotTable.grid_tiles()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.load_baysor"><code class="docutils literal notranslate"><span class="pre">SpotTable.load_baysor()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.load_cell_polygons"><code class="docutils literal notranslate"><span class="pre">SpotTable.load_cell_polygons()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.load_json"><code class="docutils literal notranslate"><span class="pre">SpotTable.load_json()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.load_merscope"><code class="docutils literal notranslate"><span class="pre">SpotTable.load_merscope()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.load_npz"><code class="docutils literal notranslate"><span class="pre">SpotTable.load_npz()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.load_pickle"><code class="docutils literal notranslate"><span class="pre">SpotTable.load_pickle()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.load_stereoseq"><code class="docutils literal notranslate"><span class="pre">SpotTable.load_stereoseq()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.map_gene_ids_to_names"><code class="docutils literal notranslate"><span class="pre">SpotTable.map_gene_ids_to_names()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.map_gene_names_to_ids"><code class="docutils literal notranslate"><span class="pre">SpotTable.map_gene_names_to_ids()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.map_indices_from_parent"><code class="docutils literal notranslate"><span class="pre">SpotTable.map_indices_from_parent()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.map_indices_to_parent"><code class="docutils literal notranslate"><span class="pre">SpotTable.map_indices_to_parent()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.map_mask_to_parent"><code class="docutils literal notranslate"><span class="pre">SpotTable.map_mask_to_parent()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.merge_cells"><code class="docutils literal notranslate"><span class="pre">SpotTable.merge_cells()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.plot_rect"><code class="docutils literal notranslate"><span class="pre">SpotTable.plot_rect()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.production_cell_ids"><code class="docutils literal notranslate"><span class="pre">SpotTable.production_cell_ids</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.reduced_expression_map"><code class="docutils literal notranslate"><span class="pre">SpotTable.reduced_expression_map()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.save_cell_polygons"><code class="docutils literal notranslate"><span class="pre">SpotTable.save_cell_polygons()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.save_csv"><code class="docutils literal notranslate"><span class="pre">SpotTable.save_csv()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.save_json"><code class="docutils literal notranslate"><span class="pre">SpotTable.save_json()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.save_npz"><code class="docutils literal notranslate"><span class="pre">SpotTable.save_npz()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.scatter_plot"><code class="docutils literal notranslate"><span class="pre">SpotTable.scatter_plot()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.set_cell_ids_from_tiles"><code class="docutils literal notranslate"><span class="pre">SpotTable.set_cell_ids_from_tiles()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.show_binned_heatmap"><code class="docutils literal notranslate"><span class="pre">SpotTable.show_binned_heatmap()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.show_image"><code class="docutils literal notranslate"><span class="pre">SpotTable.show_image()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.show_subregion_images"><code class="docutils literal notranslate"><span class="pre">SpotTable.show_subregion_images()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.split_tiles"><code class="docutils literal notranslate"><span class="pre">SpotTable.split_tiles()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.unique_cell_ids"><code class="docutils literal notranslate"><span class="pre">SpotTable.unique_cell_ids()</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.x"><code class="docutils literal notranslate"><span class="pre">SpotTable.x</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.y"><code class="docutils literal notranslate"><span class="pre">SpotTable.y</span></code></a></li>
<li><a class="reference internal" href="#sis.spot_table.SpotTable.z"><code class="docutils literal notranslate"><span class="pre">SpotTable.z</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#sis.spot_table.run_cell_polygon_calculation"><code class="docutils literal notranslate"><span class="pre">run_cell_polygon_calculation()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-sis.util">sis.util module</a><ul>
<li><a class="reference internal" href="#sis.util.example_function"><code class="docutils literal notranslate"><span class="pre">example_function()</span></code></a></li>
<li><a class="reference internal" href="#sis.util.load_config"><code class="docutils literal notranslate"><span class="pre">load_config()</span></code></a></li>
<li><a class="reference internal" href="#sis.util.log_plus_1"><code class="docutils literal notranslate"><span class="pre">log_plus_1()</span></code></a></li>
<li><a class="reference internal" href="#sis.util.make_cirro_compatible"><code class="docutils literal notranslate"><span class="pre">make_cirro_compatible()</span></code></a></li>
<li><a class="reference internal" href="#sis.util.map_to_ubyte"><code class="docutils literal notranslate"><span class="pre">map_to_ubyte()</span></code></a></li>
<li><a class="reference internal" href="#sis.util.package_for_10x"><code class="docutils literal notranslate"><span class="pre">package_for_10x()</span></code></a></li>
<li><a class="reference internal" href="#sis.util.plot_cbg_centroids"><code class="docutils literal notranslate"><span class="pre">plot_cbg_centroids()</span></code></a></li>
<li><a class="reference internal" href="#sis.util.plot_genes"><code class="docutils literal notranslate"><span class="pre">plot_genes()</span></code></a></li>
<li><a class="reference internal" href="#sis.util.poly_to_geojson"><code class="docutils literal notranslate"><span class="pre">poly_to_geojson()</span></code></a></li>
<li><a class="reference internal" href="#sis.util.rainbow_wheel"><code class="docutils literal notranslate"><span class="pre">rainbow_wheel()</span></code></a></li>
<li><a class="reference internal" href="#sis.util.reduce_expression"><code class="docutils literal notranslate"><span class="pre">reduce_expression()</span></code></a></li>
<li><a class="reference internal" href="#sis.util.show_cells_and_transcripts"><code class="docutils literal notranslate"><span class="pre">show_cells_and_transcripts()</span></code></a></li>
<li><a class="reference internal" href="#sis.util.show_float_rgb"><code class="docutils literal notranslate"><span class="pre">show_float_rgb()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-sis">Module contents</a></li>
</ul>
</li>
</ul>
</div>
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Spots-in-Spaaaaaace!</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">sis package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/sis.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="sis-package">
<h1>sis package<a class="headerlink" href="#sis-package" title="Link to this heading"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading"></a></h2>
</section>
<section id="module-sis.celltype_mapping">
<span id="sis-celltype-mapping-module"></span><h2>sis.celltype_mapping module<a class="headerlink" href="#module-sis.celltype_mapping" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="sis.celltype_mapping.CKMapping">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.celltype_mapping.</span></span><span class="sig-name descname"><span class="pre">CKMapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sp_data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="s"><span class="pre">'AnnData'</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mapping_result_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#CKMapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.CKMapping" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#sis.celltype_mapping.CellTypeMapping" title="sis.celltype_mapping.CellTypeMapping"><code class="xref py py-class docutils literal notranslate"><span class="pre">CellTypeMapping</span></code></a></p>
<p>Currently for CKmapping mapping occurs on HPC through a separate script. This class loads in results for analysis</p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.CKMapping.plot_best_mapping">
<span class="sig-name descname"><span class="pre">plot_best_mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">level</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#CKMapping.plot_best_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.CKMapping.plot_best_mapping" title="Link to this definition"></a></dt>
<dd><p>Plot the best mapping at particular heirarchy level. Optionally provide a dictionary with keys specifying group
names and keys a list of labels to plot separately for easier visualization.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.CKMapping.plot_best_mapping_corr">
<span class="sig-name descname"><span class="pre">plot_best_mapping_corr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">level</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#CKMapping.plot_best_mapping_corr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.CKMapping.plot_best_mapping_corr" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.CKMapping.plot_best_mapping_probability">
<span class="sig-name descname"><span class="pre">plot_best_mapping_probability</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">level</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#CKMapping.plot_best_mapping_probability"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.CKMapping.plot_best_mapping_probability" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.CKMapping.plot_mapping_performance">
<span class="sig-name descname"><span class="pre">plot_mapping_performance</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#CKMapping.plot_mapping_performance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.CKMapping.plot_mapping_performance" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.celltype_mapping.CellTypeMapping">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.celltype_mapping.</span></span><span class="sig-name descname"><span class="pre">CellTypeMapping</span></span><a class="reference internal" href="_modules/sis/celltype_mapping.html#CellTypeMapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.CellTypeMapping" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Container class for cell type mapping methods</p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.CellTypeMapping.load_from_timestamp">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_from_timestamp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">directory</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timestamp</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'mapping'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meta_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#CellTypeMapping.load_from_timestamp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.CellTypeMapping.load_from_timestamp" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.CellTypeMapping.mapping">
<span class="sig-name descname"><span class="pre">mapping</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#CellTypeMapping.mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.CellTypeMapping.mapping" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.celltype_mapping.CellTypeMapping.ordered_labels">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ordered_labels</span></span><a class="headerlink" href="#sis.celltype_mapping.CellTypeMapping.ordered_labels" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.CellTypeMapping.qc_mapping">
<span class="sig-name descname"><span class="pre">qc_mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">qc_params</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#CellTypeMapping.qc_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.CellTypeMapping.qc_mapping" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.CellTypeMapping.reset_qc">
<span class="sig-name descname"><span class="pre">reset_qc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#CellTypeMapping.reset_qc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.CellTypeMapping.reset_qc" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.CellTypeMapping.save_mapping">
<span class="sig-name descname"><span class="pre">save_mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">save_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'mapping'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#CellTypeMapping.save_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.CellTypeMapping.save_mapping" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.CellTypeMapping.spatial_umap">
<span class="sig-name descname"><span class="pre">spatial_umap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">attr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'ad_sp'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umap_args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#CellTypeMapping.spatial_umap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.CellTypeMapping.spatial_umap" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.celltype_mapping.ScrattchMapping">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.celltype_mapping.</span></span><span class="sig-name descname"><span class="pre">ScrattchMapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sp_data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="s"><span class="pre">'AnnData'</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxonomy_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#ScrattchMapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.ScrattchMapping" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#sis.celltype_mapping.CellTypeMapping" title="sis.celltype_mapping.CellTypeMapping"><code class="xref py py-class docutils literal notranslate"><span class="pre">CellTypeMapping</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.ScrattchMapping.get_confusion_matrix">
<span class="sig-name descname"><span class="pre">get_confusion_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pivot_cols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#ScrattchMapping.get_confusion_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.ScrattchMapping.get_confusion_matrix" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.ScrattchMapping.load_scrattch_mapping_results">
<span class="sig-name descname"><span class="pre">load_scrattch_mapping_results</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#ScrattchMapping.load_scrattch_mapping_results"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.ScrattchMapping.load_scrattch_mapping_results" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.ScrattchMapping.load_taxonomy_anndata">
<span class="sig-name descname"><span class="pre">load_taxonomy_anndata</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#ScrattchMapping.load_taxonomy_anndata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.ScrattchMapping.load_taxonomy_anndata" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.ScrattchMapping.make_mapping_anndata">
<span class="sig-name descname"><span class="pre">make_mapping_anndata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">save_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ad_sp_layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">training_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_qc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#ScrattchMapping.make_mapping_anndata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.ScrattchMapping.make_mapping_anndata" title="Link to this definition"></a></dt>
<dd><p>save_path: where to save out mapping anndata object to load into R script
ad_sp_layer: scrattch_mapping requires log normalized data, identify where in the ad_sp object</p>
<blockquote>
<div><p>this data resides. If None it will be assumed the data is ad_sp.X otherwise identify
key for ad_sp.layers</p>
</div></blockquote>
<p>training_genes: list of genes to use for mapping, if None will use all genes in ad_sp.var_names
cell_qc: column to use in ad_sp.obs to filter cells for mapping. Values in column must be boolean. If None all cells will be used</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.ScrattchMapping.plot_best_mapping">
<span class="sig-name descname"><span class="pre">plot_best_mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">level</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qc_pass</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#ScrattchMapping.plot_best_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.ScrattchMapping.plot_best_mapping" title="Link to this definition"></a></dt>
<dd><p>Plot the best mapping at particular heirarchy level. Optionally provide a dictionary with keys specifying group
names and keys a list of labels to plot separately for easier visualization.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.ScrattchMapping.plot_best_mapping_corr">
<span class="sig-name descname"><span class="pre">plot_best_mapping_corr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">level</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qc_pass</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#ScrattchMapping.plot_best_mapping_corr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.ScrattchMapping.plot_best_mapping_corr" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.ScrattchMapping.plot_class_proportions">
<span class="sig-name descname"><span class="pre">plot_class_proportions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">level</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qc_pass</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#ScrattchMapping.plot_class_proportions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.ScrattchMapping.plot_class_proportions" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.ScrattchMapping.run_on_hpc">
<span class="sig-name descname"><span class="pre">run_on_hpc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ad_map_args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hpc_args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">docker</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'singularity</span> <span class="pre">exec</span> <span class="pre">--cleanenv</span> <span class="pre">docker://bicore/scrattch_mapping:latest'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">r_script</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'scrattch_mapping.R'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#ScrattchMapping.run_on_hpc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.ScrattchMapping.run_on_hpc" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.celltype_mapping.TangramMapping">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.celltype_mapping.</span></span><span class="sig-name descname"><span class="pre">TangramMapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sc_data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="s"><span class="pre">'AnnData'</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sp_data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="s"><span class="pre">'AnnData'</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#TangramMapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.TangramMapping" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#sis.celltype_mapping.CellTypeMapping" title="sis.celltype_mapping.CellTypeMapping"><code class="xref py py-class docutils literal notranslate"><span class="pre">CellTypeMapping</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.TangramMapping.compare_spatial_gene_exp">
<span class="sig-name descname"><span class="pre">compare_spatial_gene_exp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['cux2',</span> <span class="pre">'rorb',</span> <span class="pre">'fezf2',</span> <span class="pre">'lhx6',</span> <span class="pre">'pvalb',</span> <span class="pre">'grik1',</span> <span class="pre">'lamp5',</span> <span class="pre">'gfap',</span> <span class="pre">'opalin',</span> <span class="pre">'mog']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#TangramMapping.compare_spatial_gene_exp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.TangramMapping.compare_spatial_gene_exp" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.TangramMapping.evaluate_mapping">
<span class="sig-name descname"><span class="pre">evaluate_mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">level</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#TangramMapping.evaluate_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.TangramMapping.evaluate_mapping" title="Link to this definition"></a></dt>
<dd><p>Run a standard set of mapping evaluations</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.TangramMapping.get_discrete_cell_mapping">
<span class="sig-name descname"><span class="pre">get_discrete_cell_mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">levels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['class_label',</span> <span class="pre">'neighborhood_label',</span> <span class="pre">'subclass_label',</span> <span class="pre">'supertype_label',</span> <span class="pre">'cluster_label']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#TangramMapping.get_discrete_cell_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.TangramMapping.get_discrete_cell_mapping" title="Link to this definition"></a></dt>
<dd><p>Return mapped cell type that has the highest probability for each cell. Optionally set a lower probability threshold.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.TangramMapping.mapping">
<span class="sig-name descname"><span class="pre">mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">device</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'cpu'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'clusters'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_label</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'subclass'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#TangramMapping.mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.TangramMapping.mapping" title="Link to this definition"></a></dt>
<dd><p>main mapping function using tangram.map_cells_to_space. Default to cluster mode at subclass level.
Pass other argumaents to tangram.map_cells_to_space via args</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.TangramMapping.plot_cell_probability">
<span class="sig-name descname"><span class="pre">plot_cell_probability</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ncols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#TangramMapping.plot_cell_probability"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.TangramMapping.plot_cell_probability" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.TangramMapping.plot_discrete_mapping">
<span class="sig-name descname"><span class="pre">plot_discrete_mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">level</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#TangramMapping.plot_discrete_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.TangramMapping.plot_discrete_mapping" title="Link to this definition"></a></dt>
<dd><p>Plot the discrete mapping produced from get_discrete_cell_mapping at hierarchy level. Optionally provide a dictionary with keys specifying group
names and keys a list of labels to plot separately for easier visualization. If level is None defaults to self.meta[cluster_label]</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.TangramMapping.plot_discrete_mapping_probability">
<span class="sig-name descname"><span class="pre">plot_discrete_mapping_probability</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">groups</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">level</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#TangramMapping.plot_discrete_mapping_probability"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.TangramMapping.plot_discrete_mapping_probability" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.TangramMapping.project_genes">
<span class="sig-name descname"><span class="pre">project_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#TangramMapping.project_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.TangramMapping.project_genes" title="Link to this definition"></a></dt>
<dd><p>Project gene expression from RNAseq data to spatial data</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.celltype_mapping.TangramMapping.set_training_genes">
<span class="sig-name descname"><span class="pre">set_training_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">training_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/celltype_mapping.html#TangramMapping.set_training_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.celltype_mapping.TangramMapping.set_training_genes" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

</section>
<section id="module-sis.expression_dataset">
<span id="sis-expression-dataset-module"></span><h2>sis.expression_dataset module<a class="headerlink" href="#module-sis.expression_dataset" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="sis.expression_dataset.ExpressionDataset">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.expression_dataset.</span></span><span class="sig-name descname"><span class="pre">ExpressionDataset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">expression_data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expression_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_cells</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_label</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#ExpressionDataset"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.ExpressionDataset" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Encapsulates gene expression data from scRNAseq</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>expression_data</strong> (<em>AnnData</em>)  <dl class="simple">
<dt>AnnData with following attributes:</dt><dd><p>X: cell x gene count
obs: cell annotations with at least class, subclass, cluster annotations
var: gene names</p>
</dd>
</dl>
</p></li>
<li><p><strong>expression_type</strong> (<em>str</em>)  Indicates the kind of data in <em>expression_data.X</em>. Options are raw (raw expression counts), cpm, logcpm, binary</p></li>
<li><p><strong>region</strong> (<em>string</em>)  Region of the brain this data comes from</p></li>
<li><p><strong>save_path</strong> (<em>string</em>)  Directory to save date for later recall</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.ExpressionDataset.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#ExpressionDataset.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.ExpressionDataset.copy" title="Link to this definition"></a></dt>
<dd><p>Create a copy of this ExpressionDataset with some arguments changed.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.expression_dataset.ExpressionDataset.genes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">genes</span></span><a class="headerlink" href="#sis.expression_dataset.ExpressionDataset.genes" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.ExpressionDataset.load_arrow">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_arrow</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expression_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expression_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_n_samples</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#ExpressionDataset.load_arrow"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.ExpressionDataset.load_arrow" title="Link to this definition"></a></dt>
<dd><p>Return dataset loaded from 3 arrow files</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_file</strong> (<em>str</em>)  Name of arrow file containing a column gene with the same length and order as rows in expression_file.</p></li>
<li><p><strong>annotation_file</strong> (<em>str</em>)  Name of arrow file containing columns sample_id, cluster, subclass, and class.</p></li>
<li><p><strong>expression_file</strong> (<em>str</em>)  Name of arrow file containing one gene per row, one cell per column.</p></li>
<li><p><strong>expression_type</strong> (<em>str</em>)  Type of expression data provided in <em>expression_file</em> (see ExpressionDataset.__init__(expression_type))</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.ExpressionDataset.load_from_timestamp">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_from_timestamp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">directory</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timestamp</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#ExpressionDataset.load_from_timestamp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.ExpressionDataset.load_from_timestamp" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.ExpressionDataset.normalize">
<span class="sig-name descname"><span class="pre">normalize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">expression_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#ExpressionDataset.normalize"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.ExpressionDataset.normalize" title="Link to this definition"></a></dt>
<dd><p>Return a normalized copy of this dataset.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>expression_type</strong> (<em>str</em>)  Type of data to return. Options are cpm, logcpm, and binary</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.ExpressionDataset.select_genes">
<span class="sig-name descname"><span class="pre">select_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#sis.expression_dataset.ExpressionDataset" title="sis.expression_dataset.ExpressionDataset"><span class="pre">ExpressionDataset</span></a></span></span><a class="reference internal" href="_modules/sis/expression_dataset.html#ExpressionDataset.select_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.ExpressionDataset.select_genes" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.ExpressionDataset.subsample_by_type">
<span class="sig-name descname"><span class="pre">subsample_by_type</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target_cells</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#ExpressionDataset.subsample_by_type"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.ExpressionDataset.subsample_by_type" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.expression_dataset.GeneBasisMethod">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.expression_dataset.</span></span><span class="sig-name descname"><span class="pre">GeneBasisMethod</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exp_data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.expression_dataset.ExpressionDataset" title="sis.expression_dataset.ExpressionDataset"><span class="pre">ExpressionDataset</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">run_local</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#GeneBasisMethod"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.GeneBasisMethod" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#sis.expression_dataset.GenePanelMethod" title="sis.expression_dataset.GenePanelMethod"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenePanelMethod</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.GeneBasisMethod.celltype_mapping">
<span class="sig-name descname"><span class="pre">celltype_mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_panel</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.expression_dataset.GenePanelSelection" title="sis.expression_dataset.GenePanelSelection"><span class="pre">GenePanelSelection</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">levels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['class',</span> <span class="pre">'subclass',</span> <span class="pre">'cluster']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#GeneBasisMethod.celltype_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.GeneBasisMethod.celltype_mapping" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.GeneBasisMethod.df_to_sce">
<span class="sig-name descname"><span class="pre">df_to_sce</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#GeneBasisMethod.df_to_sce"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.GeneBasisMethod.df_to_sce" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.GeneBasisMethod.filter_genes">
<span class="sig-name descname"><span class="pre">filter_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">starting_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hvgs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#GeneBasisMethod.filter_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.GeneBasisMethod.filter_genes" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.GeneBasisMethod.load_gene_panel">
<span class="sig-name descname"><span class="pre">load_gene_panel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#GeneBasisMethod.load_gene_panel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.GeneBasisMethod.load_gene_panel" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.GeneBasisMethod.neighborhood_score">
<span class="sig-name descname"><span class="pre">neighborhood_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_panel</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.expression_dataset.GenePanelSelection" title="sis.expression_dataset.GenePanelSelection"><span class="pre">GenePanelSelection</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#GeneBasisMethod.neighborhood_score"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.GeneBasisMethod.neighborhood_score" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.GeneBasisMethod.panel_eval">
<span class="sig-name descname"><span class="pre">panel_eval</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_panel</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.expression_dataset.GenePanelSelection" title="sis.expression_dataset.GenePanelSelection"><span class="pre">GenePanelSelection</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">levels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['class',</span> <span class="pre">'subclass',</span> <span class="pre">'cluster']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#GeneBasisMethod.panel_eval"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.GeneBasisMethod.panel_eval" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.GeneBasisMethod.raw_to_sce">
<span class="sig-name descname"><span class="pre">raw_to_sce</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#GeneBasisMethod.raw_to_sce"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.GeneBasisMethod.raw_to_sce" title="Link to this definition"></a></dt>
<dd><p>Used in geneBasis to convert csv data files into SingleCellExperiment
object used by geneBasis to select and evaluate panel</p>
<section id="parameters">
<h3>Parameters:<a class="headerlink" href="#parameters" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>args<span class="classifier">dict</span></dt><dd><p>Optional arguments to pass to geneBasis.raw_to_sce</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.GeneBasisMethod.run_on_hpc">
<span class="sig-name descname"><span class="pre">run_on_hpc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">docker</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'/allen/programs/celltypes/workgroups/rnaseqanalysis/bicore/singularity/genebasis.sif'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">r_script</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'geneBasis.R'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hpc_args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">panel_args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#GeneBasisMethod.run_on_hpc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.GeneBasisMethod.run_on_hpc" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.GeneBasisMethod.save_to_csv">
<span class="sig-name descname"><span class="pre">save_to_csv</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#GeneBasisMethod.save_to_csv"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.GeneBasisMethod.save_to_csv" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.GeneBasisMethod.select_gene_panel">
<span class="sig-name descname"><span class="pre">select_gene_panel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'gene_panel_selection'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">panel_args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hpc_args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">docker</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'/allen/programs/celltypes/workgroups/rnaseqanalysis/bicore/singularity/genebasis.sif'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">r_script</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'geneBasis.R'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#GeneBasisMethod.select_gene_panel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.GeneBasisMethod.select_gene_panel" title="Link to this definition"></a></dt>
<dd><section id="paramters">
<h3>Paramters<a class="headerlink" href="#paramters" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>size<span class="classifier">int</span></dt><dd><p>Size of gene panel to select</p>
</dd>
<dt>data<span class="classifier">ExpressionDataset</span></dt><dd><p>Gene expression dataset from which to derive panel</p>
</dd>
<dt>panel_args<span class="classifier">dict | {}</span></dt><dd><p>Optional arguments to pass to geneBasis.gene_search</p>
</dd>
<dt>hpc_args<span class="classifier">dict | {}</span></dt><dd><p>Optionally run on HPC, must at least specify <cite>job_path</cite></p>
</dd>
</dl>
</section>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.expression_dataset.GenePanelMethod">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.expression_dataset.</span></span><span class="sig-name descname"><span class="pre">GenePanelMethod</span></span><a class="reference internal" href="_modules/sis/expression_dataset.html#GenePanelMethod"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.GenePanelMethod" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.GenePanelMethod.evaluate_gene_panel">
<span class="sig-name descname"><span class="pre">evaluate_gene_panel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">selection</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.expression_dataset.GenePanelSelection" title="sis.expression_dataset.GenePanelSelection"><span class="pre">GenePanelSelection</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#GenePanelMethod.evaluate_gene_panel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.GenePanelMethod.evaluate_gene_panel" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.GenePanelMethod.select_gene_panel">
<span class="sig-name descname"><span class="pre">select_gene_panel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.expression_dataset.ExpressionDataset" title="sis.expression_dataset.ExpressionDataset"><span class="pre">ExpressionDataset</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#sis.expression_dataset.GenePanelSelection" title="sis.expression_dataset.GenePanelSelection"><span class="pre">GenePanelSelection</span></a></span></span><a class="reference internal" href="_modules/sis/expression_dataset.html#GenePanelMethod.select_gene_panel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.GenePanelMethod.select_gene_panel" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.expression_dataset.GenePanelSelection">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.expression_dataset.</span></span><span class="sig-name descname"><span class="pre">GenePanelSelection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exp_data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.expression_dataset.ExpressionDataset" title="sis.expression_dataset.ExpressionDataset"><span class="pre">ExpressionDataset</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_panel</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.expression_dataset.GenePanelMethod" title="sis.expression_dataset.GenePanelMethod"><span class="pre">GenePanelMethod</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#GenePanelSelection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.GenePanelSelection" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Contains a gene panel selection as well as information about how the selection was performed.</p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.GenePanelSelection.eval_method">
<span class="sig-name descname"><span class="pre">eval_method</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.celltype_mapping.CellTypeMapping" title="sis.celltype_mapping.CellTypeMapping"><span class="pre">CellTypeMapping</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#GenePanelSelection.eval_method"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.GenePanelSelection.eval_method" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.GenePanelSelection.expression_dataset">
<span class="sig-name descname"><span class="pre">expression_dataset</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#GenePanelSelection.expression_dataset"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.GenePanelSelection.expression_dataset" title="Link to this definition"></a></dt>
<dd><p>Return a new ExpressionDataset containing only the genes selected in this panel.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.GenePanelSelection.get_confusion_matrix">
<span class="sig-name descname"><span class="pre">get_confusion_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#GenePanelSelection.get_confusion_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.GenePanelSelection.get_confusion_matrix" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.GenePanelSelection.get_fraction_mapped">
<span class="sig-name descname"><span class="pre">get_fraction_mapped</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#GenePanelSelection.get_fraction_mapped"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.GenePanelSelection.get_fraction_mapped" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.GenePanelSelection.load_from_timestamp">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_from_timestamp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">directory</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timestamp</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'gene_panel_selection'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#GenePanelSelection.load_from_timestamp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.GenePanelSelection.load_from_timestamp" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.GenePanelSelection.report">
<span class="sig-name descname"><span class="pre">report</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#GenePanelSelection.report"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.GenePanelSelection.report" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.GenePanelSelection.save_selection">
<span class="sig-name descname"><span class="pre">save_selection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">save_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'gene_panel_selection'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#GenePanelSelection.save_selection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.GenePanelSelection.save_selection" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.expression_dataset.PROPOSEMethod">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.expression_dataset.</span></span><span class="sig-name descname"><span class="pre">PROPOSEMethod</span></span><a class="reference internal" href="_modules/sis/expression_dataset.html#PROPOSEMethod"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.PROPOSEMethod" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#sis.expression_dataset.GenePanelMethod" title="sis.expression_dataset.GenePanelMethod"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenePanelMethod</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.PROPOSEMethod.select_gene_panel">
<span class="sig-name descname"><span class="pre">select_gene_panel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.expression_dataset.ExpressionDataset" title="sis.expression_dataset.ExpressionDataset"><span class="pre">ExpressionDataset</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_classes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cluster'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cuda_device</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_fraction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_fraction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pre_eliminate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#sis.expression_dataset.GenePanelSelection" title="sis.expression_dataset.GenePanelSelection"><span class="pre">GenePanelSelection</span></a></span></span><a class="reference internal" href="_modules/sis/expression_dataset.html#PROPOSEMethod.select_gene_panel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.PROPOSEMethod.select_gene_panel" title="Link to this definition"></a></dt>
<dd><section id="id1">
<h3>Paramters<a class="headerlink" href="#id1" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>size<span class="classifier">int</span></dt><dd><p>Size of gene panel to select</p>
</dd>
<dt>data<span class="classifier">ExpressionDataset</span></dt><dd><p>Gene expression dataset from which to derive panel</p>
</dd>
<dt>use_classes<span class="classifier">bool</span></dt><dd><p>If True, optimize for predicting classes rather than gene expression</p>
</dd>
<dt>annotation_column<span class="classifier">str</span></dt><dd><p>Name of annotation column that holds classes (when use_clases is True)</p>
</dd>
<dt>cuda_device<span class="classifier">int</span></dt><dd><p>Index of cuda GPU device to use</p>
</dd>
<dt>train_fraction<span class="classifier">float</span></dt><dd><p>Fraction of expression data to use for training (samples not used for training and testing are used for validation)</p>
</dd>
<dt>test_fraction<span class="classifier">float</span></dt><dd><p>Fraction of expression data to hold out for testing</p>
</dd>
<dt>pre_eliminate<span class="classifier">int</span></dt><dd><p>Maximum size of gene set to process with model</p>
</dd>
</dl>
</section>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.expression_dataset.ScranMethod">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.expression_dataset.</span></span><span class="sig-name descname"><span class="pre">ScranMethod</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exp_data=&lt;class</span> <span class="pre">'sis.expression_dataset.ExpressionDataset'&gt;</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#ScranMethod"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.ScranMethod" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#sis.expression_dataset.GenePanelMethod" title="sis.expression_dataset.GenePanelMethod"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenePanelMethod</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.ScranMethod.select_gene_panel">
<span class="sig-name descname"><span class="pre">select_gene_panel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'hvg_selection'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#ScranMethod.select_gene_panel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.ScranMethod.select_gene_panel" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.expression_dataset.SeuratMethod">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.expression_dataset.</span></span><span class="sig-name descname"><span class="pre">SeuratMethod</span></span><a class="reference internal" href="_modules/sis/expression_dataset.html#SeuratMethod"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.SeuratMethod" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#sis.expression_dataset.GenePanelMethod" title="sis.expression_dataset.GenePanelMethod"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenePanelMethod</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.SeuratMethod.select_gene_panel">
<span class="sig-name descname"><span class="pre">select_gene_panel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.expression_dataset.ExpressionDataset" title="sis.expression_dataset.ExpressionDataset"><span class="pre">ExpressionDataset</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">flavor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'seurat_v3'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#sis.expression_dataset.GenePanelSelection" title="sis.expression_dataset.GenePanelSelection"><span class="pre">GenePanelSelection</span></a></span></span><a class="reference internal" href="_modules/sis/expression_dataset.html#SeuratMethod.select_gene_panel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.SeuratMethod.select_gene_panel" title="Link to this definition"></a></dt>
<dd><section id="id2">
<h3>Paramters<a class="headerlink" href="#id2" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>size<span class="classifier">int</span></dt><dd><p>Size of gene panel to select</p>
</dd>
<dt>data<span class="classifier">ExpressionDataset</span></dt><dd><p>Gene expression dataset from which to derive panel</p>
</dd>
</dl>
</section>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.expression_dataset.gene_basis_panel_eval">
<span class="sig-prename descclassname"><span class="pre">sis.expression_dataset.</span></span><span class="sig-name descname"><span class="pre">gene_basis_panel_eval</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_panel</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.expression_dataset.GenePanelSelection" title="sis.expression_dataset.GenePanelSelection"><span class="pre">GenePanelSelection</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">levels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['class',</span> <span class="pre">'subclass',</span> <span class="pre">'cluster']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'gene_basis_evaluation'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#gene_basis_panel_eval"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.gene_basis_panel_eval" title="Link to this definition"></a></dt>
<dd><p>Peform gene panel evaluation using geneBasis.
Peforms two evaluations at various cell group levels:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>Neighborhood score</p></li>
<li><p>Confusion matrix / fraction mapped correctly</p></li>
</ol>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_panel</strong> (<a class="reference internal" href="#sis.expression_dataset.GenePanelSelection" title="sis.expression_dataset.GenePanelSelection"><em>GenePanelSelection</em></a>)  </p></li>
<li><p><strong>levels</strong> (<em>list</em><em> | </em><em>[</em><em>'class'</em><em>, </em><em>'subclass'</em><em>, </em><em>'cluster'</em><em>]</em>)  List of cell group levels to perform confusion matrix. Defaults to [class, subclass, cluster]</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.expression_dataset.importr">
<span class="sig-prename descclassname"><span class="pre">sis.expression_dataset.</span></span><span class="sig-name descname"><span class="pre">importr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lib</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#importr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.importr" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.expression_dataset.load_mfishtools_fraction_mapped">
<span class="sig-prename descclassname"><span class="pre">sis.expression_dataset.</span></span><span class="sig-name descname"><span class="pre">load_mfishtools_fraction_mapped</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gps</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.expression_dataset.GenePanelSelection" title="sis.expression_dataset.GenePanelSelection"><span class="pre">GenePanelSelection</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#load_mfishtools_fraction_mapped"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.load_mfishtools_fraction_mapped" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.expression_dataset.load_panel_eval">
<span class="sig-prename descclassname"><span class="pre">sis.expression_dataset.</span></span><span class="sig-name descname"><span class="pre">load_panel_eval</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#load_panel_eval"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.load_panel_eval" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.expression_dataset.mFISHtoolsMethod">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.expression_dataset.</span></span><span class="sig-name descname"><span class="pre">mFISHtoolsMethod</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exp_data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.expression_dataset.ExpressionDataset" title="sis.expression_dataset.ExpressionDataset"><span class="pre">ExpressionDataset</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#mFISHtoolsMethod"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.mFISHtoolsMethod" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#sis.expression_dataset.GenePanelMethod" title="sis.expression_dataset.GenePanelMethod"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenePanelMethod</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.mFISHtoolsMethod.build_panel_ad">
<span class="sig-name descname"><span class="pre">build_panel_ad</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_label</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">panel_args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#mFISHtoolsMethod.build_panel_ad"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.mFISHtoolsMethod.build_panel_ad" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.mFISHtoolsMethod.filter_genes">
<span class="sig-name descname"><span class="pre">filter_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">starting_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_binary_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">3000</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#mFISHtoolsMethod.filter_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.mFISHtoolsMethod.filter_genes" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.mFISHtoolsMethod.load_gene_panel">
<span class="sig-name descname"><span class="pre">load_gene_panel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#mFISHtoolsMethod.load_gene_panel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.mFISHtoolsMethod.load_gene_panel" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.mFISHtoolsMethod.run_on_hpc">
<span class="sig-name descname"><span class="pre">run_on_hpc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">docker</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'singularity</span> <span class="pre">exec</span> <span class="pre">--cleanenv</span> <span class="pre">docker://bicore/scrattch_mapping:latest'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">r_script</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'mFISHTools.R'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hpc_args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#mFISHtoolsMethod.run_on_hpc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.mFISHtoolsMethod.run_on_hpc" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.expression_dataset.mFISHtoolsMethod.select_gene_panel">
<span class="sig-name descname"><span class="pre">select_gene_panel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_label</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">panel_args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hpc_args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">docker</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'singularity</span> <span class="pre">exec</span> <span class="pre">--cleanenv</span> <span class="pre">docker://bicore/scrattch_mapping:latest'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">r_script</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'mFISHTools.R'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#mFISHtoolsMethod.select_gene_panel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.mFISHtoolsMethod.select_gene_panel" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.expression_dataset.marker_gene_eval">
<span class="sig-prename descclassname"><span class="pre">sis.expression_dataset.</span></span><span class="sig-name descname"><span class="pre">marker_gene_eval</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_panel</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.expression_dataset.GenePanelSelection" title="sis.expression_dataset.GenePanelSelection"><span class="pre">GenePanelSelection</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_gene_files</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cluster'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gene'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#marker_gene_eval"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.marker_gene_eval" title="Link to this definition"></a></dt>
<dd><p>Evaluates how many known marker genes are present in the gene panel</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_panel</strong> (<a class="reference internal" href="#sis.expression_dataset.GenePanelSelection" title="sis.expression_dataset.GenePanelSelection"><em>GenePanelSelection</em></a>)  GenePanelSelection with list of genes to perform evaluation on</p></li>
<li><p><strong>marker_gene_files</strong> (<em>dictionary</em>)  Dictionary where keys are the level (cluster, subclass, class) of the marker genes
and values are csv files with at one column designating the cell group (cluster_col)
and another column specifying a marker gene for that group (gene_col)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.expression_dataset.norm_confusion_matrix">
<span class="sig-prename descclassname"><span class="pre">sis.expression_dataset.</span></span><span class="sig-name descname"><span class="pre">norm_confusion_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">confusion_matrix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#norm_confusion_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.norm_confusion_matrix" title="Link to this definition"></a></dt>
<dd><p>Normalize confusion matrix to fraction of mapped cells</p>
<section id="id3">
<h3>Parameters:<a class="headerlink" href="#id3" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>confusion_matrix: pandas.Dataframe or pandas.MultiIndex</dt><dd><p>Dataframe or MultiIndex of pivot tables with known cell types
on the columns and mapped cell types on the rows with elements
containing counts. Marginal counts in a row call All also
required to calculate the fraction of correctly mapped cells.
Included when calling pandas.DataFrame.pivot_table(margins=True)</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.expression_dataset.run_mfishtools_fraction_mapped">
<span class="sig-prename descclassname"><span class="pre">sis.expression_dataset.</span></span><span class="sig-name descname"><span class="pre">run_mfishtools_fraction_mapped</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gps</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.expression_dataset.GenePanelSelection" title="sis.expression_dataset.GenePanelSelection"><span class="pre">GenePanelSelection</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_label</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Cluster'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hpc_args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/expression_dataset.html#run_mfishtools_fraction_mapped"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.expression_dataset.run_mfishtools_fraction_mapped" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-sis.hpc">
<span id="sis-hpc-module"></span><h2>sis.hpc module<a class="headerlink" href="#module-sis.hpc" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="sis.hpc.JobState">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.hpc.</span></span><span class="sig-name descname"><span class="pre">JobState</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">state_code</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">state</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/hpc.html#JobState"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.hpc.JobState" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="sis.hpc.JobState.state_codes">
<span class="sig-name descname"><span class="pre">state_codes</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'BF':</span> <span class="pre">('BOOT_FAIL',</span> <span class="pre">'Job</span> <span class="pre">terminated</span> <span class="pre">due</span> <span class="pre">to</span> <span class="pre">launch</span> <span class="pre">failure,</span> <span class="pre">typically</span> <span class="pre">due</span> <span class="pre">to</span> <span class="pre">a</span> <span class="pre">hardware</span> <span class="pre">failure</span> <span class="pre">(e.g.</span> <span class="pre">unable</span> <span class="pre">to</span> <span class="pre">boot</span> <span class="pre">the</span> <span class="pre">node</span> <span class="pre">or</span> <span class="pre">block</span> <span class="pre">and</span> <span class="pre">the</span> <span class="pre">job</span> <span class="pre">can</span> <span class="pre">not</span> <span class="pre">be</span> <span class="pre">requeued).'),</span> <span class="pre">'CA':</span> <span class="pre">('CANCELLED',</span> <span class="pre">'Job</span> <span class="pre">was</span> <span class="pre">explicitly</span> <span class="pre">cancelled</span> <span class="pre">by</span> <span class="pre">the</span> <span class="pre">user</span> <span class="pre">or</span> <span class="pre">system</span> <span class="pre">administrator.</span>&#160; <span class="pre">The</span> <span class="pre">job</span> <span class="pre">may</span> <span class="pre">or</span> <span class="pre">may</span> <span class="pre">not</span> <span class="pre">have</span> <span class="pre">been</span> <span class="pre">initiated.'),</span> <span class="pre">'CD':</span> <span class="pre">('COMPLETED',</span> <span class="pre">'Job</span> <span class="pre">has</span> <span class="pre">terminated</span> <span class="pre">all</span> <span class="pre">processes</span> <span class="pre">on</span> <span class="pre">all</span> <span class="pre">nodes</span> <span class="pre">with</span> <span class="pre">an</span> <span class="pre">exit</span> <span class="pre">code</span> <span class="pre">of</span> <span class="pre">zero.'),</span> <span class="pre">'CF':</span> <span class="pre">('CONFIGURING',</span> <span class="pre">'Job</span> <span class="pre">has</span> <span class="pre">been</span> <span class="pre">allocated</span> <span class="pre">resources,</span> <span class="pre">but</span> <span class="pre">are</span> <span class="pre">waiting</span> <span class="pre">for</span> <span class="pre">them</span> <span class="pre">to</span> <span class="pre">become</span> <span class="pre">ready</span> <span class="pre">for</span> <span class="pre">use</span> <span class="pre">(e.g.</span> <span class="pre">booting).'),</span> <span class="pre">'CG':</span> <span class="pre">('COMPLETING',</span> <span class="pre">'Job</span> <span class="pre">is</span> <span class="pre">in</span> <span class="pre">the</span> <span class="pre">process</span> <span class="pre">of</span> <span class="pre">completing.</span> <span class="pre">Some</span> <span class="pre">processes</span> <span class="pre">on</span> <span class="pre">some</span> <span class="pre">nodes</span> <span class="pre">may</span> <span class="pre">still</span> <span class="pre">be</span> <span class="pre">active.'),</span> <span class="pre">'DL':</span> <span class="pre">('DEADLINE',</span> <span class="pre">'Job</span> <span class="pre">terminated</span> <span class="pre">on</span> <span class="pre">deadline.'),</span> <span class="pre">'F':</span> <span class="pre">('FAILED',</span> <span class="pre">'Job</span> <span class="pre">terminated</span> <span class="pre">with</span> <span class="pre">non-zero</span> <span class="pre">exit</span> <span class="pre">code</span> <span class="pre">or</span> <span class="pre">other</span> <span class="pre">failure</span> <span class="pre">condition.'),</span> <span class="pre">'NF':</span> <span class="pre">('NODE_FAIL',</span> <span class="pre">'Job</span> <span class="pre">terminated</span> <span class="pre">due</span> <span class="pre">to</span> <span class="pre">failure</span> <span class="pre">of</span> <span class="pre">one</span> <span class="pre">or</span> <span class="pre">more</span> <span class="pre">allocated</span> <span class="pre">nodes.'),</span> <span class="pre">'NO':</span> <span class="pre">('NO_INFO',</span> <span class="pre">'Squeue</span> <span class="pre">returned</span> <span class="pre">no</span> <span class="pre">information</span> <span class="pre">about</span> <span class="pre">the</span> <span class="pre">requested</span> <span class="pre">job</span> <span class="pre">ID'),</span> <span class="pre">'OOM':</span> <span class="pre">('OUT_OF_MEMORY',</span> <span class="pre">'Job</span> <span class="pre">experienced</span> <span class="pre">out</span> <span class="pre">of</span> <span class="pre">memory</span> <span class="pre">error.'),</span> <span class="pre">'PD':</span> <span class="pre">('PENDING',</span> <span class="pre">'Job</span> <span class="pre">is</span> <span class="pre">awaiting</span> <span class="pre">resource</span> <span class="pre">allocation.'),</span> <span class="pre">'PR':</span> <span class="pre">('PREEMPTED',</span> <span class="pre">'Job</span> <span class="pre">terminated</span> <span class="pre">due</span> <span class="pre">to</span> <span class="pre">preemption.'),</span> <span class="pre">'R':</span> <span class="pre">('RUNNING',</span> <span class="pre">'Job</span> <span class="pre">currently</span> <span class="pre">has</span> <span class="pre">an</span> <span class="pre">allocation.'),</span> <span class="pre">'RD':</span> <span class="pre">('RESV_DEL_HOLD',</span> <span class="pre">'Job</span> <span class="pre">is</span> <span class="pre">being</span> <span class="pre">held</span> <span class="pre">after</span> <span class="pre">requested</span> <span class="pre">reservation</span> <span class="pre">was</span> <span class="pre">deleted.'),</span> <span class="pre">'RF':</span> <span class="pre">('REQUEUE_FED',</span> <span class="pre">'Job</span> <span class="pre">is</span> <span class="pre">being</span> <span class="pre">requeued</span> <span class="pre">by</span> <span class="pre">a</span> <span class="pre">federation.'),</span> <span class="pre">'RH':</span> <span class="pre">('REQUEUE_HOLD',</span> <span class="pre">'Held</span> <span class="pre">job</span> <span class="pre">is</span> <span class="pre">being</span> <span class="pre">requeued.'),</span> <span class="pre">'RQ':</span> <span class="pre">('REQUEUED',</span> <span class="pre">'Completing</span> <span class="pre">job</span> <span class="pre">is</span> <span class="pre">being</span> <span class="pre">requeued.'),</span> <span class="pre">'RS':</span> <span class="pre">('RESIZING',</span> <span class="pre">'Job</span> <span class="pre">is</span> <span class="pre">about</span> <span class="pre">to</span> <span class="pre">change</span> <span class="pre">size.'),</span> <span class="pre">'RV':</span> <span class="pre">('REVOKED',</span> <span class="pre">'Sibling</span> <span class="pre">was</span> <span class="pre">removed</span> <span class="pre">from</span> <span class="pre">cluster</span> <span class="pre">due</span> <span class="pre">to</span> <span class="pre">other</span> <span class="pre">cluster</span> <span class="pre">starting</span> <span class="pre">the</span> <span class="pre">job.'),</span> <span class="pre">'S':</span> <span class="pre">('SUSPENDED',</span> <span class="pre">'Job</span> <span class="pre">has</span> <span class="pre">an</span> <span class="pre">allocation,</span> <span class="pre">but</span> <span class="pre">execution</span> <span class="pre">has</span> <span class="pre">been</span> <span class="pre">suspended</span> <span class="pre">and</span> <span class="pre">CPUs</span> <span class="pre">have</span> <span class="pre">been</span> <span class="pre">released</span> <span class="pre">for</span> <span class="pre">other</span> <span class="pre">jobs.'),</span> <span class="pre">'SE':</span> <span class="pre">('SPECIAL_EXIT',</span> <span class="pre">'The</span> <span class="pre">job</span> <span class="pre">was</span> <span class="pre">requeued</span> <span class="pre">in</span> <span class="pre">a</span> <span class="pre">special</span> <span class="pre">state.</span> <span class="pre">This</span> <span class="pre">state</span> <span class="pre">can</span> <span class="pre">be</span> <span class="pre">set</span> <span class="pre">by</span> <span class="pre">users,</span> <span class="pre">typically</span> <span class="pre">in</span> <span class="pre">EpilogSlurmctld,</span> <span class="pre">if</span> <span class="pre">the</span> <span class="pre">job</span> <span class="pre">has</span> <span class="pre">terminated</span> <span class="pre">with</span> <span class="pre">a</span> <span class="pre">particular</span> <span class="pre">exit</span> <span class="pre">value.'),</span> <span class="pre">'SI':</span> <span class="pre">('SIGNALING',</span> <span class="pre">'Job</span> <span class="pre">is</span> <span class="pre">being</span> <span class="pre">signaled.'),</span> <span class="pre">'SO':</span> <span class="pre">('STAGE_OUT',</span> <span class="pre">'Job</span> <span class="pre">is</span> <span class="pre">staging</span> <span class="pre">out</span> <span class="pre">files.'),</span> <span class="pre">'ST':</span> <span class="pre">('STOPPED',</span> <span class="pre">'Job</span> <span class="pre">has</span> <span class="pre">an</span> <span class="pre">allocation,</span> <span class="pre">but</span> <span class="pre">execution</span> <span class="pre">has</span> <span class="pre">been</span> <span class="pre">stopped</span> <span class="pre">with</span> <span class="pre">SIGSTOP</span> <span class="pre">signal.</span>&#160; <span class="pre">CPUS</span> <span class="pre">have</span> <span class="pre">been</span> <span class="pre">retained</span> <span class="pre">by</span> <span class="pre">this</span> <span class="pre">job.'),</span> <span class="pre">'TO':</span> <span class="pre">('TIMEOUT',</span> <span class="pre">'Job</span> <span class="pre">terminated</span> <span class="pre">upon</span> <span class="pre">reaching</span> <span class="pre">its</span> <span class="pre">time</span> <span class="pre">limit.')}</span></em><a class="headerlink" href="#sis.hpc.JobState.state_codes" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sis.hpc.JobState.state_descriptions">
<span class="sig-name descname"><span class="pre">state_descriptions</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'BOOT_FAIL':</span> <span class="pre">'Job</span> <span class="pre">terminated</span> <span class="pre">due</span> <span class="pre">to</span> <span class="pre">launch</span> <span class="pre">failure,</span> <span class="pre">typically</span> <span class="pre">due</span> <span class="pre">to</span> <span class="pre">a</span> <span class="pre">hardware</span> <span class="pre">failure</span> <span class="pre">(e.g.</span> <span class="pre">unable</span> <span class="pre">to</span> <span class="pre">boot</span> <span class="pre">the</span> <span class="pre">node</span> <span class="pre">or</span> <span class="pre">block</span> <span class="pre">and</span> <span class="pre">the</span> <span class="pre">job</span> <span class="pre">can</span> <span class="pre">not</span> <span class="pre">be</span> <span class="pre">requeued).',</span> <span class="pre">'CANCELLED':</span> <span class="pre">'Job</span> <span class="pre">was</span> <span class="pre">explicitly</span> <span class="pre">cancelled</span> <span class="pre">by</span> <span class="pre">the</span> <span class="pre">user</span> <span class="pre">or</span> <span class="pre">system</span> <span class="pre">administrator.</span>&#160; <span class="pre">The</span> <span class="pre">job</span> <span class="pre">may</span> <span class="pre">or</span> <span class="pre">may</span> <span class="pre">not</span> <span class="pre">have</span> <span class="pre">been</span> <span class="pre">initiated.',</span> <span class="pre">'COMPLETED':</span> <span class="pre">'Job</span> <span class="pre">has</span> <span class="pre">terminated</span> <span class="pre">all</span> <span class="pre">processes</span> <span class="pre">on</span> <span class="pre">all</span> <span class="pre">nodes</span> <span class="pre">with</span> <span class="pre">an</span> <span class="pre">exit</span> <span class="pre">code</span> <span class="pre">of</span> <span class="pre">zero.',</span> <span class="pre">'COMPLETING':</span> <span class="pre">'Job</span> <span class="pre">is</span> <span class="pre">in</span> <span class="pre">the</span> <span class="pre">process</span> <span class="pre">of</span> <span class="pre">completing.</span> <span class="pre">Some</span> <span class="pre">processes</span> <span class="pre">on</span> <span class="pre">some</span> <span class="pre">nodes</span> <span class="pre">may</span> <span class="pre">still</span> <span class="pre">be</span> <span class="pre">active.',</span> <span class="pre">'CONFIGURING':</span> <span class="pre">'Job</span> <span class="pre">has</span> <span class="pre">been</span> <span class="pre">allocated</span> <span class="pre">resources,</span> <span class="pre">but</span> <span class="pre">are</span> <span class="pre">waiting</span> <span class="pre">for</span> <span class="pre">them</span> <span class="pre">to</span> <span class="pre">become</span> <span class="pre">ready</span> <span class="pre">for</span> <span class="pre">use</span> <span class="pre">(e.g.</span> <span class="pre">booting).',</span> <span class="pre">'DEADLINE':</span> <span class="pre">'Job</span> <span class="pre">terminated</span> <span class="pre">on</span> <span class="pre">deadline.',</span> <span class="pre">'FAILED':</span> <span class="pre">'Job</span> <span class="pre">terminated</span> <span class="pre">with</span> <span class="pre">non-zero</span> <span class="pre">exit</span> <span class="pre">code</span> <span class="pre">or</span> <span class="pre">other</span> <span class="pre">failure</span> <span class="pre">condition.',</span> <span class="pre">'NODE_FAIL':</span> <span class="pre">'Job</span> <span class="pre">terminated</span> <span class="pre">due</span> <span class="pre">to</span> <span class="pre">failure</span> <span class="pre">of</span> <span class="pre">one</span> <span class="pre">or</span> <span class="pre">more</span> <span class="pre">allocated</span> <span class="pre">nodes.',</span> <span class="pre">'NO_INFO':</span> <span class="pre">'Squeue</span> <span class="pre">returned</span> <span class="pre">no</span> <span class="pre">information</span> <span class="pre">about</span> <span class="pre">the</span> <span class="pre">requested</span> <span class="pre">job</span> <span class="pre">ID',</span> <span class="pre">'OUT_OF_MEMORY':</span> <span class="pre">'Job</span> <span class="pre">experienced</span> <span class="pre">out</span> <span class="pre">of</span> <span class="pre">memory</span> <span class="pre">error.',</span> <span class="pre">'PENDING':</span> <span class="pre">'Job</span> <span class="pre">is</span> <span class="pre">awaiting</span> <span class="pre">resource</span> <span class="pre">allocation.',</span> <span class="pre">'PREEMPTED':</span> <span class="pre">'Job</span> <span class="pre">terminated</span> <span class="pre">due</span> <span class="pre">to</span> <span class="pre">preemption.',</span> <span class="pre">'REQUEUED':</span> <span class="pre">'Completing</span> <span class="pre">job</span> <span class="pre">is</span> <span class="pre">being</span> <span class="pre">requeued.',</span> <span class="pre">'REQUEUE_FED':</span> <span class="pre">'Job</span> <span class="pre">is</span> <span class="pre">being</span> <span class="pre">requeued</span> <span class="pre">by</span> <span class="pre">a</span> <span class="pre">federation.',</span> <span class="pre">'REQUEUE_HOLD':</span> <span class="pre">'Held</span> <span class="pre">job</span> <span class="pre">is</span> <span class="pre">being</span> <span class="pre">requeued.',</span> <span class="pre">'RESIZING':</span> <span class="pre">'Job</span> <span class="pre">is</span> <span class="pre">about</span> <span class="pre">to</span> <span class="pre">change</span> <span class="pre">size.',</span> <span class="pre">'RESV_DEL_HOLD':</span> <span class="pre">'Job</span> <span class="pre">is</span> <span class="pre">being</span> <span class="pre">held</span> <span class="pre">after</span> <span class="pre">requested</span> <span class="pre">reservation</span> <span class="pre">was</span> <span class="pre">deleted.',</span> <span class="pre">'REVOKED':</span> <span class="pre">'Sibling</span> <span class="pre">was</span> <span class="pre">removed</span> <span class="pre">from</span> <span class="pre">cluster</span> <span class="pre">due</span> <span class="pre">to</span> <span class="pre">other</span> <span class="pre">cluster</span> <span class="pre">starting</span> <span class="pre">the</span> <span class="pre">job.',</span> <span class="pre">'RUNNING':</span> <span class="pre">'Job</span> <span class="pre">currently</span> <span class="pre">has</span> <span class="pre">an</span> <span class="pre">allocation.',</span> <span class="pre">'SIGNALING':</span> <span class="pre">'Job</span> <span class="pre">is</span> <span class="pre">being</span> <span class="pre">signaled.',</span> <span class="pre">'SPECIAL_EXIT':</span> <span class="pre">'The</span> <span class="pre">job</span> <span class="pre">was</span> <span class="pre">requeued</span> <span class="pre">in</span> <span class="pre">a</span> <span class="pre">special</span> <span class="pre">state.</span> <span class="pre">This</span> <span class="pre">state</span> <span class="pre">can</span> <span class="pre">be</span> <span class="pre">set</span> <span class="pre">by</span> <span class="pre">users,</span> <span class="pre">typically</span> <span class="pre">in</span> <span class="pre">EpilogSlurmctld,</span> <span class="pre">if</span> <span class="pre">the</span> <span class="pre">job</span> <span class="pre">has</span> <span class="pre">terminated</span> <span class="pre">with</span> <span class="pre">a</span> <span class="pre">particular</span> <span class="pre">exit</span> <span class="pre">value.',</span> <span class="pre">'STAGE_OUT':</span> <span class="pre">'Job</span> <span class="pre">is</span> <span class="pre">staging</span> <span class="pre">out</span> <span class="pre">files.',</span> <span class="pre">'STOPPED':</span> <span class="pre">'Job</span> <span class="pre">has</span> <span class="pre">an</span> <span class="pre">allocation,</span> <span class="pre">but</span> <span class="pre">execution</span> <span class="pre">has</span> <span class="pre">been</span> <span class="pre">stopped</span> <span class="pre">with</span> <span class="pre">SIGSTOP</span> <span class="pre">signal.</span>&#160; <span class="pre">CPUS</span> <span class="pre">have</span> <span class="pre">been</span> <span class="pre">retained</span> <span class="pre">by</span> <span class="pre">this</span> <span class="pre">job.',</span> <span class="pre">'SUSPENDED':</span> <span class="pre">'Job</span> <span class="pre">has</span> <span class="pre">an</span> <span class="pre">allocation,</span> <span class="pre">but</span> <span class="pre">execution</span> <span class="pre">has</span> <span class="pre">been</span> <span class="pre">suspended</span> <span class="pre">and</span> <span class="pre">CPUs</span> <span class="pre">have</span> <span class="pre">been</span> <span class="pre">released</span> <span class="pre">for</span> <span class="pre">other</span> <span class="pre">jobs.',</span> <span class="pre">'TIMEOUT':</span> <span class="pre">'Job</span> <span class="pre">terminated</span> <span class="pre">upon</span> <span class="pre">reaching</span> <span class="pre">its</span> <span class="pre">time</span> <span class="pre">limit.'}</span></em><a class="headerlink" href="#sis.hpc.JobState.state_descriptions" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sis.hpc.JobState.state_to_code">
<span class="sig-name descname"><span class="pre">state_to_code</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'BOOT_FAIL':</span> <span class="pre">'BF',</span> <span class="pre">'CANCELLED':</span> <span class="pre">'CA',</span> <span class="pre">'COMPLETED':</span> <span class="pre">'CD',</span> <span class="pre">'COMPLETING':</span> <span class="pre">'CG',</span> <span class="pre">'CONFIGURING':</span> <span class="pre">'CF',</span> <span class="pre">'DEADLINE':</span> <span class="pre">'DL',</span> <span class="pre">'FAILED':</span> <span class="pre">'F',</span> <span class="pre">'NODE_FAIL':</span> <span class="pre">'NF',</span> <span class="pre">'NO_INFO':</span> <span class="pre">'NO',</span> <span class="pre">'OUT_OF_MEMORY':</span> <span class="pre">'OOM',</span> <span class="pre">'PENDING':</span> <span class="pre">'PD',</span> <span class="pre">'PREEMPTED':</span> <span class="pre">'PR',</span> <span class="pre">'REQUEUED':</span> <span class="pre">'RQ',</span> <span class="pre">'REQUEUE_FED':</span> <span class="pre">'RF',</span> <span class="pre">'REQUEUE_HOLD':</span> <span class="pre">'RH',</span> <span class="pre">'RESIZING':</span> <span class="pre">'RS',</span> <span class="pre">'RESV_DEL_HOLD':</span> <span class="pre">'RD',</span> <span class="pre">'REVOKED':</span> <span class="pre">'RV',</span> <span class="pre">'RUNNING':</span> <span class="pre">'R',</span> <span class="pre">'SIGNALING':</span> <span class="pre">'SI',</span> <span class="pre">'SPECIAL_EXIT':</span> <span class="pre">'SE',</span> <span class="pre">'STAGE_OUT':</span> <span class="pre">'SO',</span> <span class="pre">'STOPPED':</span> <span class="pre">'ST',</span> <span class="pre">'SUSPENDED':</span> <span class="pre">'S',</span> <span class="pre">'TIMEOUT':</span> <span class="pre">'TO'}</span></em><a class="headerlink" href="#sis.hpc.JobState.state_to_code" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.hpc.SlurmJob">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.hpc.</span></span><span class="sig-name descname"><span class="pre">SlurmJob</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sbatch_output</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">host</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">array_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_array</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/hpc.html#SlurmJob"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.hpc.SlurmJob" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.hpc.SlurmJob.cancel">
<span class="sig-name descname"><span class="pre">cancel</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/hpc.html#SlurmJob.cancel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.hpc.SlurmJob.cancel" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.hpc.SlurmJob.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#sis.hpc.SlurmJob.error" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.hpc.SlurmJob.error_file">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error_file</span></span><a class="headerlink" href="#sis.hpc.SlurmJob.error_file" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.hpc.SlurmJob.is_done">
<span class="sig-name descname"><span class="pre">is_done</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/hpc.html#SlurmJob.is_done"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.hpc.SlurmJob.is_done" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.hpc.SlurmJob.output">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">output</span></span><a class="headerlink" href="#sis.hpc.SlurmJob.output" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.hpc.SlurmJob.output_file">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">output_file</span></span><a class="headerlink" href="#sis.hpc.SlurmJob.output_file" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.hpc.SlurmJob.state">
<span class="sig-name descname"><span class="pre">state</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/hpc.html#SlurmJob.state"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.hpc.SlurmJob.state" title="Link to this definition"></a></dt>
<dd><p>Return the state code for this job, or a dictionary of state codes for array jobs.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.hpc.SlurmJobArray">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.hpc.</span></span><span class="sig-name descname"><span class="pre">SlurmJobArray</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sbatch_output</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">host</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/hpc.html#SlurmJobArray"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.hpc.SlurmJobArray" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#sis.hpc.SlurmJob" title="sis.hpc.SlurmJob"><code class="xref py py-class docutils literal notranslate"><span class="pre">SlurmJob</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.hpc.SlurmJobArray.cancel">
<span class="sig-name descname"><span class="pre">cancel</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/hpc.html#SlurmJobArray.cancel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.hpc.SlurmJobArray.cancel" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.hpc.SlurmJobArray.finished_jobs">
<span class="sig-name descname"><span class="pre">finished_jobs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/hpc.html#SlurmJobArray.finished_jobs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.hpc.SlurmJobArray.finished_jobs" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.hpc.SlurmJobArray.is_done">
<span class="sig-name descname"><span class="pre">is_done</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/hpc.html#SlurmJobArray.is_done"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.hpc.SlurmJobArray.is_done" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.hpc.SlurmJobArray.state">
<span class="sig-name descname"><span class="pre">state</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/hpc.html#SlurmJobArray.state"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.hpc.SlurmJobArray.state" title="Link to this definition"></a></dt>
<dd><p>Return the state code for this job, or a dictionary of state codes for array jobs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.hpc.SlurmJobArray.state_counts">
<span class="sig-name descname"><span class="pre">state_counts</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/hpc.html#SlurmJobArray.state_counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.hpc.SlurmJobArray.state_counts" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.hpc.SlurmJobArray.unfinished_jobs">
<span class="sig-name descname"><span class="pre">unfinished_jobs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/hpc.html#SlurmJobArray.unfinished_jobs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.hpc.SlurmJobArray.unfinished_jobs" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.hpc.SlurmJobArray.wait_iter">
<span class="sig-name descname"><span class="pre">wait_iter</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/hpc.html#SlurmJobArray.wait_iter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.hpc.SlurmJobArray.wait_iter" title="Link to this definition"></a></dt>
<dd><p>Iterator that returns jobs as they finish</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.hpc.close_ssh_connections">
<span class="sig-prename descclassname"><span class="pre">sis.hpc.</span></span><span class="sig-name descname"><span class="pre">close_ssh_connections</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/hpc.html#close_ssh_connections"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.hpc.close_ssh_connections" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.hpc.hpc_worker">
<span class="sig-prename descclassname"><span class="pre">sis.hpc.</span></span><span class="sig-name descname"><span class="pre">hpc_worker</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/hpc.html#hpc_worker"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.hpc.hpc_worker" title="Link to this definition"></a></dt>
<dd><p>Invoked when running <cite>python -m sis.hpc</cite></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.hpc.run">
<span class="sig-prename descclassname"><span class="pre">sis.hpc.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">host</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/hpc.html#run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.hpc.run" title="Link to this definition"></a></dt>
<dd><p>Run a command on any host. If the host is other than localhost,
then the command is run over ssh.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.hpc.run_slurm">
<span class="sig-prename descclassname"><span class="pre">sis.hpc.</span></span><span class="sig-name descname"><span class="pre">run_slurm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hpc_host</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">partition</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nodes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ntasks</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">array</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mincpus</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gpus_per_node</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mem</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">command</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mail_user</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mail_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'END,FAIL'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">error</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/hpc.html#run_slurm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.hpc.run_slurm" title="Link to this definition"></a></dt>
<dd><p>Call sbatch (locally or over ssh) and return a SlurmJob instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>hpc_host</strong> (<em>str</em>)  Host to call sbatch from. If not localhost, then commands are executed
on <em>hpc_host</em> via ssh.</p></li>
<li><p><strong>job_path</strong> (<em>str</em>)  Location to write slurm script file and output/error logs.
Must be accessible from <em>hpc_host</em>.</p></li>
<li><p><strong>partition</strong> (<em>str</em>)  SLURM partition to allocate job</p></li>
<li><p><strong>job_name</strong> (<em>str</em>)  SLURM job name. This is also used to set default file names for the sbatch script,
output log, and error log.</p></li>
<li><p><strong>nodes</strong> (<em>int</em>)  Number of nodes to request</p></li>
<li><p><strong>ntasks</strong> (<em>int</em>)  Number of tasks to launch per job.
(Use multiple tasks when all must be scheduled to start simultaneously)</p></li>
<li><p><strong>array</strong> (<em>str</em><em> | </em><em>None</em>)  Use array=A-B where A and B are the first and last index to start multiple,
independent jobs. The output/error file names may include %a to reference the
array index, and the command may use $SLURM_ARRAY_TASK_ID.</p></li>
<li><p><strong>mincpus</strong> (<em>int</em><em> | </em><em>None</em>)  Minimum CPUs per node</p></li>
<li><p><strong>gpus_per_node</strong> (<em>int</em><em> | </em><em>str</em><em> | </em><em>None</em>)  GPUs per node</p></li>
<li><p><strong>mem</strong> (<em>str</em>)  Maximum memory to allocate per job (e.g. 100G)</p></li>
<li><p><strong>time</strong> (<em>str</em>)  Maximum job time (e.g. 4:00:00)</p></li>
<li><p><strong>command</strong> (<em>str</em>)  Command to run per job (e.g. srun mycommand $SLURM_ARRAY_TASK_ID)</p></li>
<li><p><strong>mail_user</strong> (<em>str</em><em> | </em><em>None</em>)  Optional email address to notify upon completion</p></li>
<li><p><strong>mail_type</strong> (<em>str</em>)  Events on which to send email (default=END,FAIL)</p></li>
<li><p><strong>output</strong> (<em>str</em><em> | </em><em>None</em>)  Optional location of file to store command output (default is in job_path
with automatically chosen name)</p></li>
<li><p><strong>error</strong> (<em>str</em><em> | </em><em>None</em>)  Optional location of file to store error output (default is in job_path
with automatically chosen name)</p></li>
<li><p><strong>job_file</strong> (<em>str</em><em> | </em><em>None</em>)  Optional location to store sbatch script (default is in job_path
with automatically chosen name)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.hpc.run_slurm_func">
<span class="sig-prename descclassname"><span class="pre">sis.hpc.</span></span><span class="sig-name descname"><span class="pre">run_slurm_func</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">run_spec</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conda_env</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/hpc.html#run_slurm_func"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.hpc.run_slurm_func" title="Link to this definition"></a></dt>
<dd><p>Run a function on SLURM.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>run_spec</strong> (<em>tuple</em><em> | </em><em>dict</em>)  Either a tuple containing (func, args, kwargs) to be called in SLURM jobs, or a dict
where the keys are array job IDs and the values are (func, args, kwargs) tuples.</p></li>
<li><p><strong>conda_env</strong> (<em>str</em>)  Conda envronment from which to run <em>func</em></p></li>
<li><p><strong>**kwds</strong>  All other keyword arguments are passed to run_slurm().</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.hpc.sacct">
<span class="sig-prename descclassname"><span class="pre">sis.hpc.</span></span><span class="sig-name descname"><span class="pre">sacct</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">host</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_duration</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/hpc.html#sacct"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.hpc.sacct" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.hpc.squeue">
<span class="sig-prename descclassname"><span class="pre">sis.hpc.</span></span><span class="sig-name descname"><span class="pre">squeue</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">host</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_duration</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/hpc.html#squeue"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.hpc.squeue" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-sis.image">
<span id="sis-image-module"></span><h2>sis.image module<a class="headerlink" href="#module-sis.image" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="sis.image.Image">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.image.</span></span><span class="sig-name descname"><span class="pre">Image</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transform</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.image.ImageTransform" title="sis.image.ImageTransform"><span class="pre">ImageTransform</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">channels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#Image"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.Image" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#sis.image.ImageBase" title="sis.image.ImageBase"><code class="xref py py-class docutils literal notranslate"><span class="pre">ImageBase</span></code></a></p>
<p>An Image defined by 4D numpy array (frames, rows, cols, channels).</p>
<p>Carries metadata about pixel transform and channel identity.</p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.image.Image.get_data">
<span class="sig-name descname"><span class="pre">get_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">channel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#Image.get_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.Image.get_data" title="Link to this definition"></a></dt>
<dd><p>Return array of image data.</p>
<p>If the image has multiple channels, then the name of the channel to return must be given.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.image.Image.get_sub_data">
<span class="sig-name descname"><span class="pre">get_sub_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frames</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">channel</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#Image.get_sub_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.Image.get_sub_data" title="Link to this definition"></a></dt>
<dd><p>Get image data for a subregion</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>frames</strong> (<em>tuple</em>)  (first_frame, last_frame)</p></li>
<li><p><strong>rows</strong> (<em>tuple</em>)  (first_row, last_row+1)</p></li>
<li><p><strong>cols</strong> (<em>tuple</em>)  (first_col, last_col+1)</p></li>
<li><p><strong>channel</strong> (<em>str</em><em> | </em><em>None</em>)  Name of channel to return data from</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.image.Image.shape">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">shape</span></span><a class="headerlink" href="#sis.image.Image.shape" title="Link to this definition"></a></dt>
<dd><p>Return 4D shape (frames, rows, columns, channels)</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.image.ImageBase">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.image.</span></span><span class="sig-name descname"><span class="pre">ImageBase</span></span><a class="reference internal" href="_modules/sis/image.html#ImageBase"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageBase" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.image.ImageBase.bounds">
<span class="sig-name descname"><span class="pre">bounds</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#ImageBase.bounds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageBase.bounds" title="Link to this definition"></a></dt>
<dd><p>Return (xlim, ylim) bounds of the image in spot coordinates</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.image.ImageBase.get_channel">
<span class="sig-name descname"><span class="pre">get_channel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">channel</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#ImageBase.get_channel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageBase.get_channel" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.image.ImageBase.get_data">
<span class="sig-name descname"><span class="pre">get_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">channel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#ImageBase.get_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageBase.get_data" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.image.ImageBase.get_frame">
<span class="sig-name descname"><span class="pre">get_frame</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frame</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#ImageBase.get_frame"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageBase.get_frame" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.image.ImageBase.get_pixel_subregion">
<span class="sig-name descname"><span class="pre">get_pixel_subregion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#ImageBase.get_pixel_subregion"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageBase.get_pixel_subregion" title="Link to this definition"></a></dt>
<dd><p>Return a view of this image limited to the region [:, rowmin:rowmax, colmin:colmax]</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.image.ImageBase.get_subregion">
<span class="sig-name descname"><span class="pre">get_subregion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">incl_end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#ImageBase.get_subregion"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageBase.get_subregion" title="Link to this definition"></a></dt>
<dd><p>Return a view of this image limited to the region [(xmin, xmax), (ymin, ymax)]</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.image.ImageBase.shape">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">shape</span></span><a class="headerlink" href="#sis.image.ImageBase.shape" title="Link to this definition"></a></dt>
<dd><p>Return 4D shape (frames, rows, columns, channels)</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.image.ImageBase.show">
<span class="sig-name descname"><span class="pre">show</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frame</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">channel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#ImageBase.show"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageBase.show" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.image.ImageFile">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.image.</span></span><span class="sig-name descname"><span class="pre">ImageFile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transform</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.image.ImageTransform" title="sis.image.ImageTransform"><span class="pre">ImageTransform</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">axes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">channels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#ImageFile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageFile" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#sis.image.ImageBase" title="sis.image.ImageBase"><code class="xref py py-class docutils literal notranslate"><span class="pre">ImageBase</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.image.ImageFile.get_data">
<span class="sig-name descname"><span class="pre">get_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">channel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#ImageFile.get_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageFile.get_data" title="Link to this definition"></a></dt>
<dd><p>Return array of image data.</p>
<p>If the image has multiple channels, then the name of the channel to return must be given.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.image.ImageFile.get_sub_data">
<span class="sig-name descname"><span class="pre">get_sub_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frames</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">channel</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#ImageFile.get_sub_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageFile.get_sub_data" title="Link to this definition"></a></dt>
<dd><p>Get image data for a subregion</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>frames</strong> (<em>tuple</em>)  (first_frame, last_frame)</p></li>
<li><p><strong>rows</strong> (<em>tuple</em>)  (first_row, last_row+1)</p></li>
<li><p><strong>cols</strong> (<em>tuple</em>)  (first_col, last_col+1)</p></li>
<li><p><strong>channel</strong> (<em>str</em><em> | </em><em>None</em>)  Name of channel to return data from</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.image.ImageFile.load_merscope">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_merscope</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">image_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transform_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">channel</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#ImageFile.load_merscope"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageFile.load_merscope" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.image.ImageFile.shape">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">shape</span></span><a class="headerlink" href="#sis.image.ImageFile.shape" title="Link to this definition"></a></dt>
<dd><p>Return 4D shape (frames, rows, columns, channels)</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.image.ImageStack">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.image.</span></span><span class="sig-name descname"><span class="pre">ImageStack</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">images</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#ImageStack"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageStack" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#sis.image.ImageBase" title="sis.image.ImageBase"><code class="xref py py-class docutils literal notranslate"><span class="pre">ImageBase</span></code></a></p>
<p>A stack of Image z-planes</p>
<p>Assumes images are all the same shape and evenly spaced along the z axis.</p>
<dl class="py property">
<dt class="sig sig-object py" id="sis.image.ImageStack.channels">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">channels</span></span><a class="headerlink" href="#sis.image.ImageStack.channels" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.image.ImageStack.get_data">
<span class="sig-name descname"><span class="pre">get_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">channel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#ImageStack.get_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageStack.get_data" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.image.ImageStack.get_sub_data">
<span class="sig-name descname"><span class="pre">get_sub_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frames</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">channel</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#ImageStack.get_sub_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageStack.get_sub_data" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.image.ImageStack.load_merscope_stacks">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_merscope_stacks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#ImageStack.load_merscope_stacks"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageStack.load_merscope_stacks" title="Link to this definition"></a></dt>
<dd><p>Read standard merscope image mosaic format, returning multiple image stacks</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.image.ImageStack.shape">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">shape</span></span><a class="headerlink" href="#sis.image.ImageStack.shape" title="Link to this definition"></a></dt>
<dd><p>Return 4D shape (frames, rows, columns, channels)</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.image.ImageStack.transform">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">transform</span></span><a class="headerlink" href="#sis.image.ImageStack.transform" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.image.ImageTransform">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.image.</span></span><span class="sig-name descname"><span class="pre">ImageTransform</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matrix</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#ImageTransform"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageTransform" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Transfomation mapping between (x, y) spot coordinates and (row, col) image pixels</p>
<p><em>matrix</em> is a numpy array of shape (2, 3) containing the affine transformation matrix that maps from spots to pixels</p>
<dl class="py property">
<dt class="sig sig-object py" id="sis.image.ImageTransform.inverse_matrix">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">inverse_matrix</span></span><a class="headerlink" href="#sis.image.ImageTransform.inverse_matrix" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.image.ImageTransform.map_from_pixels">
<span class="sig-name descname"><span class="pre">map_from_pixels</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pixels</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#ImageTransform.map_from_pixels"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageTransform.map_from_pixels" title="Link to this definition"></a></dt>
<dd><p>Map (row, col) image pixels to positions (x, y).</p>
<p>Points must be an array of shape (N, 2).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.image.ImageTransform.map_to_pixels">
<span class="sig-name descname"><span class="pre">map_to_pixels</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">points</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#ImageTransform.map_to_pixels"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageTransform.map_to_pixels" title="Link to this definition"></a></dt>
<dd><p>Map (x, y) positions to image pixels (row, col).</p>
<p>Points must be an array of shape (N, 2).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.image.ImageTransform.translated">
<span class="sig-name descname"><span class="pre">translated</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">offset</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#ImageTransform.translated"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageTransform.translated" title="Link to this definition"></a></dt>
<dd><p>Return a new transform that is translated by <em>offset</em> (where offset is expressed in pixels)</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.image.ImageView">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.image.</span></span><span class="sig-name descname"><span class="pre">ImageView</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">image</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frames</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">channels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#ImageView"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageView" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#sis.image.ImageBase" title="sis.image.ImageBase"><code class="xref py py-class docutils literal notranslate"><span class="pre">ImageBase</span></code></a></p>
<p>Represents a subset of data from an Image (a rectangular subregion or subset of channels)</p>
<dl class="py property">
<dt class="sig sig-object py" id="sis.image.ImageView.channels">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">channels</span></span><a class="headerlink" href="#sis.image.ImageView.channels" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.image.ImageView.get_data">
<span class="sig-name descname"><span class="pre">get_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">channel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#ImageView.get_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageView.get_data" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.image.ImageView.get_sub_data">
<span class="sig-name descname"><span class="pre">get_sub_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frames</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rows</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">channel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/image.html#ImageView.get_sub_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.image.ImageView.get_sub_data" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.image.ImageView.name">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">name</span></span><a class="headerlink" href="#sis.image.ImageView.name" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.image.ImageView.shape">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">shape</span></span><a class="headerlink" href="#sis.image.ImageView.shape" title="Link to this definition"></a></dt>
<dd><p>Return 4D shape (frames, rows, columns, channels)</p>
</dd></dl>

</dd></dl>

</section>
<section id="module-sis.optional_import">
<span id="sis-optional-import-module"></span><h2>sis.optional_import module<a class="headerlink" href="#module-sis.optional_import" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="sis.optional_import.OptionalImportError">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.optional_import.</span></span><span class="sig-name descname"><span class="pre">OptionalImportError</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exc</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/optional_import.html#OptionalImportError"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.optional_import.OptionalImportError" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Dummy object that just re-raises an ImportError when it is accessed.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.optional_import.optional_import">
<span class="sig-prename descclassname"><span class="pre">sis.optional_import.</span></span><span class="sig-name descname"><span class="pre">optional_import</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">module</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">package</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/optional_import.html#optional_import"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.optional_import.optional_import" title="Link to this definition"></a></dt>
<dd><p>Try importing a module, but if that fails, wait until the first time it is
accessed before raising the ImportError.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>module</strong> (<em>str</em>)  Name of module to import (or import from)</p></li>
<li><p><strong>names</strong> (<em>str</em><em> | </em><em>list</em><em> of </em><em>str</em><em> | </em><em>None</em>)  Optional name or list of names to import from <em>module</em>. If string, then
the single imported object is returned. If list, then a list
of imported objects is returned.</p></li>
<li><p><strong>package</strong> (<em>str</em><em> | </em><em>None</em>)  Optional base package from which relative <em>module</em> names are constructed.
This argument is required if the <em>module</em> is relative (begins with <cite>.</cite>).</p></li>
<li><p><strong>Examples::</strong>  <p># import numba
numba = optional_import(numba)</p>
<p># from numpy import array, zeros
array, zeros = optional_import(numpy, names=[array, zeros])</p>
<p># from ..mypackage import myname
myname = optional_import(..mypackage, names=myname, package=__name__)</p>
</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-sis.qc">
<span id="sis-qc-module"></span><h2>sis.qc module<a class="headerlink" href="#module-sis.qc" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="sis.qc.calc_n_blanks">
<span class="sig-prename descclassname"><span class="pre">sis.qc.</span></span><span class="sig-name descname"><span class="pre">calc_n_blanks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_by_gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">DataFrame</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/qc.html#calc_n_blanks"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.qc.calc_n_blanks" title="Link to this definition"></a></dt>
<dd><p>Calculate number of blanks per cell.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.qc.calc_n_genes">
<span class="sig-prename descclassname"><span class="pre">sis.qc.</span></span><span class="sig-name descname"><span class="pre">calc_n_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_by_gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">DataFrame</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/qc.html#calc_n_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.qc.calc_n_genes" title="Link to this definition"></a></dt>
<dd><p>Calculate number of genes per cell.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.qc.calc_n_transcripts">
<span class="sig-prename descclassname"><span class="pre">sis.qc.</span></span><span class="sig-name descname"><span class="pre">calc_n_transcripts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_by_gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">DataFrame</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/qc.html#calc_n_transcripts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.qc.calc_n_transcripts" title="Link to this definition"></a></dt>
<dd><p>Calculate number of detected transcripts per cell.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.qc.get_counts_from_cbg">
<span class="sig-prename descclassname"><span class="pre">sis.qc.</span></span><span class="sig-name descname"><span class="pre">get_counts_from_cbg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_by_gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">DataFrame</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/qc.html#get_counts_from_cbg"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.qc.get_counts_from_cbg" title="Link to this definition"></a></dt>
<dd><p>Get counts array from a cell by gene table in either an AnnData object
or pandas DataFrame.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.qc.run_doublet_detection">
<span class="sig-prename descclassname"><span class="pre">sis.qc.</span></span><span class="sig-name descname"><span class="pre">run_doublet_detection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_by_gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method_kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/qc.html#run_doublet_detection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.qc.run_doublet_detection" title="Link to this definition"></a></dt>
<dd><p>Wrapper function to run a doublet detection algorithm on the HPC.
Saves results into a csv file and doublet detection metadata into a json.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.qc.run_solo_doublet_detection">
<span class="sig-prename descclassname"><span class="pre">sis.qc.</span></span><span class="sig-name descname"><span class="pre">run_solo_doublet_detection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_by_gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/qc.html#run_solo_doublet_detection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.qc.run_solo_doublet_detection" title="Link to this definition"></a></dt>
<dd><p>Detect doublets using SOLO. Returns results in a pandas DataFrame.</p>
</dd></dl>

</section>
<section id="module-sis.segmentation">
<span id="sis-segmentation-module"></span><h2>sis.segmentation module<a class="headerlink" href="#module-sis.segmentation" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="sis.segmentation.BaysorSegmentationMethod">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.segmentation.</span></span><span class="sig-name descname"><span class="pre">BaysorSegmentationMethod</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">options</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#BaysorSegmentationMethod"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.BaysorSegmentationMethod" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#sis.segmentation.SegmentationMethod" title="sis.segmentation.SegmentationMethod"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentationMethod</span></code></a></p>
<p>Implements 3D Baysor segmentation on SpotTable</p>
<p>Requires a Baysor binary to be executable on the system.</p>
<dl>
<dt>options = {</dt><dd><p>region: ((xmin, xmax), (ymin, ymax)),  # or None for whole table
baysor_bin: /path/to/Baysor,
baysor_output_path: /path/to/baysor/data,
use_prior_segmentation: True,
no_gene_names: False,          # if true, remove gene names
cell_dia: 10,                  # um
z_plane_thickness: 1.5,        # um
baysor_options: {</p>
<blockquote>
<div><p>scale-std: 25%,
prior-segmentation-confidence: None,
n-clusters: None,
no-ncv-estimation: True,</p>
</div></blockquote>
<p>}</p>
</dd>
</dl>
<p>}</p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.BaysorSegmentationMethod.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spot_table</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#BaysorSegmentationMethod.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.BaysorSegmentationMethod.run" title="Link to this definition"></a></dt>
<dd><p>Run segmentation on spot_table and return a Segmentation object.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.segmentation.BaysorSegmentationResult">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.segmentation.</span></span><span class="sig-name descname"><span class="pre">BaysorSegmentationResult</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.segmentation.SegmentationMethod" title="sis.segmentation.SegmentationMethod"><span class="pre">SegmentationMethod</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_spot_table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable"><span class="pre">SpotTable</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">baysor_command</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">baysor_output</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#BaysorSegmentationResult"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.BaysorSegmentationResult" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#sis.segmentation.SegmentationResult" title="sis.segmentation.SegmentationResult"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentationResult</span></code></a></p>
<dl class="py property">
<dt class="sig sig-object py" id="sis.segmentation.BaysorSegmentationResult.cell_ids">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">cell_ids</span></span><a class="headerlink" href="#sis.segmentation.BaysorSegmentationResult.cell_ids" title="Link to this definition"></a></dt>
<dd><p>Array of segmented cell IDs for each spot in the table</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.segmentation.CellposeSegmentationMethod">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.segmentation.</span></span><span class="sig-name descname"><span class="pre">CellposeSegmentationMethod</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">options</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#CellposeSegmentationMethod"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.CellposeSegmentationMethod" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#sis.segmentation.SegmentationMethod" title="sis.segmentation.SegmentationMethod"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentationMethod</span></code></a></p>
<p>Implements 2D or 3D cellpose segmentation on SpotTable</p>
<p>Will automatically segment from images attached to the SpotTable or
generate an image from total mRNA.</p>
<dl>
<dt>options = {</dt><dd><p>region: ((xmin, xmax), (ymin, ymax)),  # or None for whole table
cellpose_model: cyto2,
images: {</p>
<blockquote>
<div><p>nuclei: DAPI,
cyto: total_mrna,</p>
</div></blockquote>
<p>},
px_size: 0.108,          # um / px
cell_dia: 10,            # um
z_plane_thickness: 1.5,  # um
cellpose_options: {</p>
<blockquote>
<div><p>gpu: True,
batch_size: 8,</p>
</div></blockquote>
<p>},
dilate: 0,  # um - dilate segmentation after cellpose has finished</p>
</dd>
</dl>
<p>}</p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.CellposeSegmentationMethod.get_total_mrna_image">
<span class="sig-name descname"><span class="pre">get_total_mrna_image</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spot_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_shape</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_transform</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.image.ImageTransform" title="sis.image.ImageTransform"><span class="pre">ImageTransform</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_planes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frame</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gauss_kernel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(1,</span> <span class="pre">3,</span> <span class="pre">3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">median_kernel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(2,</span> <span class="pre">10,</span> <span class="pre">10)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#CellposeSegmentationMethod.get_total_mrna_image"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.CellposeSegmentationMethod.get_total_mrna_image" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.CellposeSegmentationMethod.map_spots_to_img_px">
<span class="sig-name descname"><span class="pre">map_spots_to_img_px</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spot_table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable"><span class="pre">SpotTable</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">image</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.image.Image" title="sis.image.Image"><span class="pre">Image</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_transform</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.image.ImageTransform" title="sis.image.ImageTransform"><span class="pre">ImageTransform</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_shape</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#CellposeSegmentationMethod.map_spots_to_img_px"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.CellposeSegmentationMethod.map_spots_to_img_px" title="Link to this definition"></a></dt>
<dd><p>Map spot table (x, y, z) positions to image (frame, row, col).</p>
<p>Optionally, provide the <em>image_transform</em> and <em>image_shape</em> instead of <em>image</em>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.CellposeSegmentationMethod.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spot_table</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#CellposeSegmentationMethod.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.CellposeSegmentationMethod.run" title="Link to this definition"></a></dt>
<dd><p>Run segmentation on spot_table and return a Segmentation object.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.segmentation.CellposeSegmentationResult">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.segmentation.</span></span><span class="sig-name descname"><span class="pre">CellposeSegmentationResult</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.segmentation.SegmentationMethod" title="sis.segmentation.SegmentationMethod"><span class="pre">SegmentationMethod</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_spot_table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable"><span class="pre">SpotTable</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">cellpose_output</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_transform</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.image.ImageTransform" title="sis.image.ImageTransform"><span class="pre">ImageTransform</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#CellposeSegmentationResult"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.CellposeSegmentationResult" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#sis.segmentation.SegmentationResult" title="sis.segmentation.SegmentationResult"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentationResult</span></code></a></p>
<dl class="py property">
<dt class="sig sig-object py" id="sis.segmentation.CellposeSegmentationResult.cell_ids">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">cell_ids</span></span><a class="headerlink" href="#sis.segmentation.CellposeSegmentationResult.cell_ids" title="Link to this definition"></a></dt>
<dd><p>Array of segmented cell IDs for each spot in the table</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.segmentation.CellposeSegmentationResult.mask_image">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mask_image</span></span><a class="headerlink" href="#sis.segmentation.CellposeSegmentationResult.mask_image" title="Link to this definition"></a></dt>
<dd><p>Return Image instance containing mask data and pixel transform</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.CellposeSegmentationResult.show">
<span class="sig-name descname"><span class="pre">show</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#CellposeSegmentationResult.show"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.CellposeSegmentationResult.show" title="Link to this definition"></a></dt>
<dd><p>Display segmentation result in matplotlib axes</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.segmentation.MerscopeSegmentationRun">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.segmentation.</span></span><span class="sig-name descname"><span class="pre">MerscopeSegmentationRun</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dt_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dt_cache</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subrgn</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seg_method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.segmentation.SegmentationMethod" title="sis.segmentation.SegmentationMethod"><span class="pre">SegmentationMethod</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">seg_opts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polygon_opts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seg_hpc_opts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polygon_hpc_opts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hpc_opts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#MerscopeSegmentationRun"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.MerscopeSegmentationRun" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#sis.segmentation.SegmentationRun" title="sis.segmentation.SegmentationRun"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentationRun</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.MerscopeSegmentationRun.get_load_args">
<span class="sig-name descname"><span class="pre">get_load_args</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#MerscopeSegmentationRun.get_load_args"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.MerscopeSegmentationRun.get_load_args" title="Link to this definition"></a></dt>
<dd><p>Get args to pass to loading function (e.g. when submitting jobs to hpc).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.MerscopeSegmentationRun.get_load_func">
<span class="sig-name descname"><span class="pre">get_load_func</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#MerscopeSegmentationRun.get_load_func"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.MerscopeSegmentationRun.get_load_func" title="Link to this definition"></a></dt>
<dd><p>Get the function to load a spot table.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationMethod">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.segmentation.</span></span><span class="sig-name descname"><span class="pre">SegmentationMethod</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationMethod"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationMethod" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Base class defining segmentation methods.</p>
<p>Subclasses should initialize with a dictionary of options, then calling
run(spot_table) will execute the segmentation method and return a SegmentationResult.</p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationMethod.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spot_table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable"><span class="pre">SpotTable</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationMethod.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationMethod.run" title="Link to this definition"></a></dt>
<dd><p>Run segmentation on spot_table and return a Segmentation object.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationResult">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.segmentation.</span></span><span class="sig-name descname"><span class="pre">SegmentationResult</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.segmentation.SegmentationMethod" title="sis.segmentation.SegmentationMethod"><span class="pre">SegmentationMethod</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_spot_table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable"><span class="pre">SpotTable</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationResult"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationResult" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Represents a segmentation of SpotTable datamethod, options, results</p>
<dl class="py property">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationResult.cell_ids">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">cell_ids</span></span><a class="headerlink" href="#sis.segmentation.SegmentationResult.cell_ids" title="Link to this definition"></a></dt>
<dd><p>Array of segmented cell IDs for each spot in the table</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationResult.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationResult.save"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationResult.save" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationResult.spot_table">
<span class="sig-name descname"><span class="pre">spot_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">min_spots</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationResult.spot_table"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationResult.spot_table" title="Link to this definition"></a></dt>
<dd><p>Return a new SpotTable with cell_ids determined by the segmentation.</p>
<p>if min_spots is given, then it specifies the threshold below which cells will be discarded</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.segmentation.</span></span><span class="sig-name descname"><span class="pre">SegmentationRun</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dt_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dt_cache</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subrgn</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seg_method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.segmentation.SegmentationMethod" title="sis.segmentation.SegmentationMethod"><span class="pre">SegmentationMethod</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">seg_opts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polygon_opts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seg_hpc_opts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polygon_hpc_opts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hpc_opts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Base class for running segmentation on a whole section (or subregion).
When the class is initialized, it creates the segmentation output directory
and sets paths for intermediate files.</p>
<p>Each step has defined input and output files. Steps can be run individually
by calling their respective methods or in a defined sequence by calling the
run() method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dt_file</strong> (<em>str</em><em> or </em><em>Path</em>)  Path to the detected transcripts file.</p></li>
<li><p><strong>image_path</strong> (<em>str</em><em> or </em><em>Path</em>)  Path to the images.</p></li>
<li><p><strong>output_dir</strong> (<em>str</em><em> or </em><em>Path</em>)  Where to save output files.</p></li>
<li><p><strong>dt_cache</strong> (<em>str</em><em> or </em><em>Path</em><em>, </em><em>optional</em>)  Path to the detected transcripts cache file. Used for faster loading.</p></li>
<li><p><strong>subrgn</strong> (<em>str</em><em> or </em><em>tuple</em>)  The subregion to segment. Set to a string, e.g. DAPI, to segment the
full region bounded by an associated image channel. To segment a
smaller region, set to a tuple corresponding to a bounding box.</p></li>
<li><p><strong>seg_method</strong> (<a class="reference internal" href="#sis.segmentation.SegmentationMethod" title="sis.segmentation.SegmentationMethod"><em>SegmentationMethod</em></a>)  The segmentation method to use. Must be found in sis.segmentation.</p></li>
<li><p><strong>seg_opts</strong> (<em>dict</em>)  Options to pass to seg_method.</p></li>
<li><p><strong>polygon_opts</strong> (<em>dict</em><em>, </em><em>optional</em>)  Options to pass to for cell polygon generation. Currently supports save_file_extension and alpha_inv_coeff.
Default is None, which sets save_file_extension to geojson and alpha_inv_coeff to 4/3.</p></li>
<li><p><strong>seg_hpc_opts</strong> (<em>dict</em><em>, </em><em>None</em><em>, </em><em>optional</em>)  Options to use for segmenting tiles on the hpc. Default is None</p></li>
<li><p><strong>polygon_hpc_opts</strong> (<em>dict</em><em>, </em><em>None</em><em>, </em><em>optional</em>)  Options to use for calculating cell polygons on the hpc. Default is None</p></li>
<li><p><strong>hpc_opts</strong> (<em>dict</em><em>, </em><em>None</em><em>, </em><em>optional</em>)  Options to use for both segmenting tiles and calculating cell polygons on the hpc (can be used in place of submitting both seg_hpc_opts and polygon_hpc_opts).
Default is None</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.clean_up">
<span class="sig-name descname"><span class="pre">clean_up</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all_ints'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.clean_up"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.clean_up" title="Link to this definition"></a></dt>
<dd><p>Clean up intermediate files after segmentation and polygon generation is complete</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>mode</strong> (<em>string</em><em>, </em><em>optional</em>)  Can be all_ints, seg_ints, polygon_ints, or none depending on desired clean up. Defaults to all_ints.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.create_cell_by_gene">
<span class="sig-name descname"><span class="pre">create_cell_by_gene</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">remove_bg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_prod_cids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.create_cell_by_gene"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.create_cell_by_gene" title="Link to this definition"></a></dt>
<dd><p>Create and save a cell by gene file in Anndata format using
the attached spot table.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.from_json">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_json</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">json_file</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.from_json"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.from_json" title="Link to this definition"></a></dt>
<dd><p>Load a run from a json file</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.from_spatial_dataset">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_spatial_dataset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sp_dataset</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subrgn</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seg_method</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seg_opts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hpc_opts</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.from_spatial_dataset"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.from_spatial_dataset" title="Link to this definition"></a></dt>
<dd><p>Alternate constructor to load from a SpatialDataset</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.get_load_args">
<em class="property"><span class="pre">abstract</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_load_args</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.get_load_args"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.get_load_args" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.get_load_func">
<em class="property"><span class="pre">abstract</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_load_func</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.get_load_func"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.get_load_func" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.get_polygon_run_spec">
<span class="sig-name descname"><span class="pre">get_polygon_run_spec</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.get_polygon_run_spec"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.get_polygon_run_spec" title="Link to this definition"></a></dt>
<dd><p>Generates a cell polygon run spec for running cell polygon jobs on the hpc</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.get_seg_run_spec">
<span class="sig-name descname"><span class="pre">get_seg_run_spec</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">regions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">result_files</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.get_seg_run_spec"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.get_seg_run_spec" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.load_cbg">
<span class="sig-name descname"><span class="pre">load_cbg</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.load_cbg"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.load_cbg" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.load_cell_ids">
<span class="sig-name descname"><span class="pre">load_cell_ids</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.load_cell_ids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.load_cell_ids" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.load_metadata">
<span class="sig-name descname"><span class="pre">load_metadata</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.load_metadata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.load_metadata" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.load_regions">
<span class="sig-name descname"><span class="pre">load_regions</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.load_regions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.load_regions" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.load_results">
<span class="sig-name descname"><span class="pre">load_results</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.load_results"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.load_results" title="Link to this definition"></a></dt>
<dd><p>Load the results of a finished segmentation.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.load_run_spec">
<span class="sig-name descname"><span class="pre">load_run_spec</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">run_spec_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.load_run_spec"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.load_run_spec" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.load_spot_table">
<span class="sig-name descname"><span class="pre">load_spot_table</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.load_spot_table"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.load_spot_table" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.merge_cell_polygons">
<span class="sig-name descname"><span class="pre">merge_cell_polygons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">run_spec</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.merge_cell_polygons"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.merge_cell_polygons" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.merge_segmented_tiles">
<span class="sig-name descname"><span class="pre">merge_segmented_tiles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">run_spec</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.merge_segmented_tiles"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.merge_segmented_tiles" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.resume">
<span class="sig-name descname"><span class="pre">resume</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.resume"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.resume" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">use_prod_cids</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clean_up</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'all_ints'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.run" title="Link to this definition"></a></dt>
<dd><p>Run all steps to perform tiled segmentation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>use_prod_cids</strong> (<em>bool</em><em>, </em><em>optional</em>)  If True, generate production cell ids and use them to index cells
in the cell by gene table. Default True.</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>)  The string to prepend to all production cell ids.</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em>, </em><em>optional</em>)  The string to append to all production cell ids.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>)  Whether to allow overwriting of output files. Default False.</p></li>
<li><p><strong>clean_up</strong> (<em>str</em><em>, </em><em>bool</em><em>, </em><em>None</em><em>, </em><em>optional</em>)  Whether or not to clean up intermediate files after segmentation
Accepts: all_ints, seg_ints, polygon_ints, none, True, False, None
Default: cleans up all intermediate files.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>SpotTable</em>  The segmented spot table.</p></li>
<li><p><em>AnnData</em>  The cell by gene table.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.save_cbg">
<span class="sig-name descname"><span class="pre">save_cbg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_by_gene</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.save_cbg"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.save_cbg" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.save_cell_ids">
<span class="sig-name descname"><span class="pre">save_cell_ids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.save_cell_ids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.save_cell_ids" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.save_metadata">
<span class="sig-name descname"><span class="pre">save_metadata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.save_metadata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.save_metadata" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.save_regions">
<span class="sig-name descname"><span class="pre">save_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">regions</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.save_regions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.save_regions" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.save_run_spec">
<span class="sig-name descname"><span class="pre">save_run_spec</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">run_spec</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">run_spec_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.save_run_spec"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.save_run_spec" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.set_intermediate_file_paths">
<span class="sig-name descname"><span class="pre">set_intermediate_file_paths</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.set_intermediate_file_paths"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.set_intermediate_file_paths" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.submit_jobs">
<span class="sig-name descname"><span class="pre">submit_jobs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">job_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">run_spec</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.submit_jobs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.submit_jobs" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.tile_seg_region">
<span class="sig-name descname"><span class="pre">tile_seg_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_tile_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.tile_seg_region"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.tile_seg_region" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.track_job_progress">
<span class="sig-name descname"><span class="pre">track_job_progress</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">jobs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.track_job_progress"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.track_job_progress" title="Link to this definition"></a></dt>
<dd><p>Track progress of hpc jobs. until all jobs have ended</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.SegmentationRun.update_metadata">
<span class="sig-name descname"><span class="pre">update_metadata</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#SegmentationRun.update_metadata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.SegmentationRun.update_metadata" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.segmentation.StereoSeqSegmentationRun">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.segmentation.</span></span><span class="sig-name descname"><span class="pre">StereoSeqSegmentationRun</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dt_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dt_cache</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subrgn</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seg_method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.segmentation.SegmentationMethod" title="sis.segmentation.SegmentationMethod"><span class="pre">SegmentationMethod</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">seg_opts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polygon_opts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seg_hpc_opts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polygon_hpc_opts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hpc_opts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#StereoSeqSegmentationRun"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.StereoSeqSegmentationRun" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#sis.segmentation.SegmentationRun" title="sis.segmentation.SegmentationRun"><code class="xref py py-class docutils literal notranslate"><span class="pre">SegmentationRun</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.StereoSeqSegmentationRun.get_load_args">
<span class="sig-name descname"><span class="pre">get_load_args</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#StereoSeqSegmentationRun.get_load_args"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.StereoSeqSegmentationRun.get_load_args" title="Link to this definition"></a></dt>
<dd><p>Get args to pass to loading function (e.g. when submitting jobs to hpc).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.segmentation.StereoSeqSegmentationRun.get_load_func">
<span class="sig-name descname"><span class="pre">get_load_func</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#StereoSeqSegmentationRun.get_load_func"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.StereoSeqSegmentationRun.get_load_func" title="Link to this definition"></a></dt>
<dd><p>Get the function to load a spot table.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.segmentation.dilate_labels">
<span class="sig-prename descclassname"><span class="pre">sis.segmentation.</span></span><span class="sig-name descname"><span class="pre">dilate_labels</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">img</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">radius</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#dilate_labels"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.dilate_labels" title="Link to this definition"></a></dt>
<dd><p>Dilate labeled regions of an image.</p>
<p>Given an image with 0 in the background and objects labeled with different integer values (such
as a cell segmentation mask), return a new image with objects expanded by <em>radius</em>.</p>
<p>(Credit: <a class="reference external" href="https://stackoverflow.com/a/70261747">https://stackoverflow.com/a/70261747</a>)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.segmentation.load_baysor_result">
<span class="sig-prename descclassname"><span class="pre">sis.segmentation.</span></span><span class="sig-name descname"><span class="pre">load_baysor_result</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">result_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_noise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_no_cell</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">brl_output</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#load_baysor_result"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.load_baysor_result" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.segmentation.merge_segmentation_results">
<span class="sig-prename descclassname"><span class="pre">sis.segmentation.</span></span><span class="sig-name descname"><span class="pre">merge_segmentation_results</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spot_table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable"><span class="pre">SpotTable</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">run_spec</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tiles</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable"><span class="pre">SpotTable</span></a><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#merge_segmentation_results"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.merge_segmentation_results" title="Link to this definition"></a></dt>
<dd><p>Merge results of a tiled segmentation, updating spot_table.cell_ids in place.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spot_table</strong> (<a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable"><em>SpotTable</em></a>)  The entire spot table to which to assign cell_ids.</p></li>
<li><p><strong>run_spec</strong> (<em>dict</em>)  The specifications provided to segment each tile on the hpc.</p></li>
<li><p><strong>tiles</strong> (<em>list</em><em>[</em><a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable"><em>SpotTable</em></a><em>]</em><em>|</em><em>None</em>)  The individual tiles that were segmented.
If not provided, will be generated from spot_table and run_spec.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>cell_ids</em>  The cell_ids assigned to each spot in spot_table.</p></li>
<li><p><em>merge_results</em>  Information about conflicts collected during tile merging.</p></li>
<li><p><em>skipped</em>  Indices of tiles skipped during segmentation.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.segmentation.run_segmentation">
<span class="sig-prename descclassname"><span class="pre">sis.segmentation.</span></span><span class="sig-name descname"><span class="pre">run_segmentation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">load_func</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">load_args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subregion</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method_class</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method_args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">result_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_id_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/segmentation.html#run_segmentation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.segmentation.run_segmentation" title="Link to this definition"></a></dt>
<dd><p>Load a spot table, run segmentation (possibly on a subregion), and save the SegmentationResult.</p>
</dd></dl>

</section>
<section id="module-sis.spatial_dataset">
<span id="sis-spatial-dataset-module"></span><h2>sis.spatial_dataset module<a class="headerlink" href="#module-sis.spatial_dataset" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="sis.spatial_dataset.MERSCOPESection">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.spatial_dataset.</span></span><span class="sig-name descname"><span class="pre">MERSCOPESection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">barcode</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#MERSCOPESection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.MERSCOPESection" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#sis.spatial_dataset.SpatialDataset" title="sis.spatial_dataset.SpatialDataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">SpatialDataset</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.MERSCOPESection.calc_cell_qc_metrics">
<span class="sig-name descname"><span class="pre">calc_cell_qc_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_by_gene</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#MERSCOPESection.calc_cell_qc_metrics"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.MERSCOPESection.calc_cell_qc_metrics" title="Link to this definition"></a></dt>
<dd><p>Calculate basic qc parameters and add them to the cell by gene
table.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.MERSCOPESection.check_doublet_detection_status">
<span class="sig-name descname"><span class="pre">check_doublet_detection_status</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#MERSCOPESection.check_doublet_detection_status"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.MERSCOPESection.check_doublet_detection_status" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.MERSCOPESection.get_section_data_paths">
<span class="sig-name descname"><span class="pre">get_section_data_paths</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#MERSCOPESection.get_section_data_paths"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.MERSCOPESection.get_section_data_paths" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.MERSCOPESection.get_section_metadata">
<span class="sig-name descname"><span class="pre">get_section_metadata</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#MERSCOPESection.get_section_metadata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.MERSCOPESection.get_section_metadata" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.MERSCOPESection.load_doublet_anndata">
<span class="sig-name descname"><span class="pre">load_doublet_anndata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">timestamp</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#MERSCOPESection.load_doublet_anndata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.MERSCOPESection.load_doublet_anndata" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.MERSCOPESection.load_segmentation_results">
<span class="sig-name descname"><span class="pre">load_segmentation_results</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">timestamp</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#MERSCOPESection.load_segmentation_results"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.MERSCOPESection.load_segmentation_results" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.MERSCOPESection.load_spottable">
<span class="sig-name descname"><span class="pre">load_spottable</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#MERSCOPESection.load_spottable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.MERSCOPESection.load_spottable" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.MERSCOPESection.make_anndata">
<span class="sig-name descname"><span class="pre">make_anndata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_by_gene</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sp_anndata.h5ad'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#MERSCOPESection.make_anndata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.MERSCOPESection.make_anndata" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.MERSCOPESection.make_cirro">
<span class="sig-name descname"><span class="pre">make_cirro</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ct_map</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#MERSCOPESection.make_cirro"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.MERSCOPESection.make_cirro" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.MERSCOPESection.run_doublet_detection_on_section">
<span class="sig-name descname"><span class="pre">run_doublet_detection_on_section</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_by_gene</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method_kwargs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hpc_args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_col</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#MERSCOPESection.run_doublet_detection_on_section"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.MERSCOPESection.run_doublet_detection_on_section" title="Link to this definition"></a></dt>
<dd><p>Submits a job to run doublet detection on a GPU node on the HPC.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.MERSCOPESection.run_mapping_on_section">
<span class="sig-name descname"><span class="pre">run_mapping_on_section</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxonomy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hpc_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#MERSCOPESection.run_mapping_on_section"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.MERSCOPESection.run_mapping_on_section" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.MERSCOPESection.run_qc_filtering_on_section">
<span class="sig-name descname"><span class="pre">run_qc_filtering_on_section</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_by_gene</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qc_params</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#MERSCOPESection.run_qc_filtering_on_section"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.MERSCOPESection.run_qc_filtering_on_section" title="Link to this definition"></a></dt>
<dd><p>Filter cells based on defined qc params. Keys of qc_params must be
in the cell_by_gene.obs.</p>
<p>Note that the cell_by_gene table is not subset; instead, a column
cell_qc_pass is created that indicates whether cells passed or failed
any qc thresholds.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.MERSCOPESection.run_segmentation_on_section">
<span class="sig-name descname"><span class="pre">run_segmentation_on_section</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">subrgn</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seg_method</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seg_opts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hpc_opts</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#MERSCOPESection.run_segmentation_on_section"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.MERSCOPESection.run_segmentation_on_section" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.spatial_dataset.SpatialDataset">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.spatial_dataset.</span></span><span class="sig-name descname"><span class="pre">SpatialDataset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">barcode</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#SpatialDataset"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.SpatialDataset" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.SpatialDataset.check_segmentation_status">
<span class="sig-name descname"><span class="pre">check_segmentation_status</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#SpatialDataset.check_segmentation_status"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.SpatialDataset.check_segmentation_status" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.SpatialDataset.get_analysis_status">
<span class="sig-name descname"><span class="pre">get_analysis_status</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#SpatialDataset.get_analysis_status"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.SpatialDataset.get_analysis_status" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.SpatialDataset.get_mappings">
<span class="sig-name descname"><span class="pre">get_mappings</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">print_output</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#SpatialDataset.get_mappings"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.SpatialDataset.get_mappings" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.SpatialDataset.load_from_barcode">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_from_barcode</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">barcode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dataset_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.spatial_dataset.SpatialDataset" title="sis.spatial_dataset.SpatialDataset"><span class="pre">SpatialDataset</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'spatial_dataset'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#SpatialDataset.load_from_barcode"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.SpatialDataset.load_from_barcode" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.SpatialDataset.save_dataset">
<span class="sig-name descname"><span class="pre">save_dataset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'spatial_dataset'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#SpatialDataset.save_dataset"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.SpatialDataset.save_dataset" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.SpatialDataset.show_gene_transcripts">
<span class="sig-name descname"><span class="pre">show_gene_transcripts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subregions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spot_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_fig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#SpatialDataset.show_gene_transcripts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.SpatialDataset.show_gene_transcripts" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.SpatialDataset.spatial_corr_to_bulk">
<span class="sig-name descname"><span class="pre">spatial_corr_to_bulk</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spot_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_fig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#SpatialDataset.spatial_corr_to_bulk"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.SpatialDataset.spatial_corr_to_bulk" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.SpatialDataset.view_mapping_results">
<span class="sig-name descname"><span class="pre">view_mapping_results</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ct_map</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ct_level</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#SpatialDataset.view_mapping_results"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.SpatialDataset.view_mapping_results" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.spatial_dataset.StereoSeqSection">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.spatial_dataset.</span></span><span class="sig-name descname"><span class="pre">StereoSeqSection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">barcode</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#StereoSeqSection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.StereoSeqSection" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#sis.spatial_dataset.SpatialDataset" title="sis.spatial_dataset.SpatialDataset"><code class="xref py py-class docutils literal notranslate"><span class="pre">SpatialDataset</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.StereoSeqSection.evaluate_qc_metric">
<span class="sig-name descname"><span class="pre">evaluate_qc_metric</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">qc_metric</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qc</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#StereoSeqSection.evaluate_qc_metric"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.StereoSeqSection.evaluate_qc_metric" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.StereoSeqSection.get_section_data_paths">
<span class="sig-name descname"><span class="pre">get_section_data_paths</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#StereoSeqSection.get_section_data_paths"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.StereoSeqSection.get_section_data_paths" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.StereoSeqSection.get_section_metadata">
<span class="sig-name descname"><span class="pre">get_section_metadata</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#StereoSeqSection.get_section_metadata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.StereoSeqSection.get_section_metadata" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.StereoSeqSection.load_spottable">
<span class="sig-name descname"><span class="pre">load_spottable</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#StereoSeqSection.load_spottable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.StereoSeqSection.load_spottable" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.StereoSeqSection.qc_bin200">
<span class="sig-name descname"><span class="pre">qc_bin200</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">median_transcript_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">18000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base_transcript_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">7000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_fig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#StereoSeqSection.qc_bin200"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.StereoSeqSection.qc_bin200" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.StereoSeqSection.qc_corr_to_pair">
<span class="sig-name descname"><span class="pre">qc_corr_to_pair</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">corr_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#StereoSeqSection.qc_corr_to_pair"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.StereoSeqSection.qc_corr_to_pair" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.StereoSeqSection.qc_gene_expression">
<span class="sig-name descname"><span class="pre">qc_gene_expression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spot_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_fig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#StereoSeqSection.qc_gene_expression"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.StereoSeqSection.qc_gene_expression" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.StereoSeqSection.qc_mapping_results">
<span class="sig-name descname"><span class="pre">qc_mapping_results</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ct_map</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_thresh</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">map_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.6</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#StereoSeqSection.qc_mapping_results"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.StereoSeqSection.qc_mapping_results" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.StereoSeqSection.qc_spatial_corr_to_bulk">
<span class="sig-name descname"><span class="pre">qc_spatial_corr_to_bulk</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.6</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kwargs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#StereoSeqSection.qc_spatial_corr_to_bulk"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.StereoSeqSection.qc_spatial_corr_to_bulk" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.StereoSeqSection.qc_widget">
<span class="sig-name descname"><span class="pre">qc_widget</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metric</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#StereoSeqSection.qc_widget"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.StereoSeqSection.qc_widget" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.StereoSeqSection.run_mapping_on_section">
<span class="sig-name descname"><span class="pre">run_mapping_on_section</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">binsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxonomy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">training_genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hvgs'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hpc_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#StereoSeqSection.run_mapping_on_section"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.StereoSeqSection.run_mapping_on_section" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.StereoSeqSection.set_partner_chips">
<span class="sig-name descname"><span class="pre">set_partner_chips</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">barcodes</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#StereoSeqSection.set_partner_chips"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.StereoSeqSection.set_partner_chips" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="sis.spatial_dataset.StereoSeqSectionCollection">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.spatial_dataset.</span></span><span class="sig-name descname"><span class="pre">StereoSeqSectionCollection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">barcode_list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#StereoSeqSectionCollection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.StereoSeqSectionCollection" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#sis.spatial_dataset.StereoSeqSection" title="sis.spatial_dataset.StereoSeqSection"><code class="xref py py-class docutils literal notranslate"><span class="pre">StereoSeqSection</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="sis.spatial_dataset.StereoSeqSectionCollection.get_counts">
<span class="sig-name descname"><span class="pre">get_counts</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spatial_dataset.html#StereoSeqSectionCollection.get_counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spatial_dataset.StereoSeqSectionCollection.get_counts" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

</section>
<section id="module-sis.spot_table">
<span id="sis-spot-table-module"></span><h2>sis.spot_table module<a class="headerlink" href="#module-sis.spot_table" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.spot_table.</span></span><span class="sig-name descname"><span class="pre">SpotTable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pos</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_ids</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_id_to_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_ids</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">production_cell_ids</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pcid_to_cid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cid_to_pcid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_polygons</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent_table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable"><span class="pre">SpotTable</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent_inds</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent_region</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">images</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Represents a spatial transcriptomics spot table.</p>
<ul class="simple">
<li><p>Contains x, y, z, gene per spot</p></li>
<li><p>May be a subset of another table, in which case indices are tracked between tables</p></li>
<li><p>May contain cell IDs loaded from segmentation</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pos</strong> (<em>array</em>)  Array of shape (N, 3) giving the pos of each detected transcript.</p></li>
<li><p><strong>gene_names</strong> (<em>array</em><em> | </em><em>None</em>)  Array of shape (N,) giving the name of the gene detected in each transcript.
Must specify either <em>gene_names</em> or <em>gene_ids</em>, not both.</p></li>
<li><p><strong>gene_ids</strong> (<em>array</em><em> | </em><em>None</em>)  Array of shape (N,) describing the gene detected in each transcript, as an index into <em>gene_id_to_name</em>.
Must specify either <em>gene_names</em> or <em>gene_ids</em>, not both.</p></li>
<li><p><strong>gene_id_to_name</strong> (<em>ndarray</em><em> | </em><em>None</em>)  Array mapping from values in <em>gene_ids</em> to string names.</p></li>
<li><p><strong>cell_ids</strong> (<em>array</em><em> | </em><em>None</em>)  Optional array of cell IDs per spot</p></li>
<li><p><strong>parent_table</strong> (<a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable"><em>SpotTable</em></a><em> | </em><em>None</em>)  Indicates that this table is a subset of a parent SpotTable.</p></li>
<li><p><strong>parent_inds</strong> (<em>array</em><em> | </em><em>None</em>)  Indices used to select the subset of spots in this table from the parent spots.</p></li>
<li><p><strong>parent_region</strong> (<em>tuple</em><em> | </em><em>None</em>)  X,Y boundaries ((xmin, xmax), (ymin, ymax)) used to select this table from the parent table.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.add_image">
<span class="sig-name descname"><span class="pre">add_image</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">image</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.add_image"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.add_image" title="Link to this definition"></a></dt>
<dd><p>Attach an image to this dataset</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.binned_expression_counts">
<span class="sig-name descname"><span class="pre">binned_expression_counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">binsize</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.binned_expression_counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.binned_expression_counts" title="Link to this definition"></a></dt>
<dd><p>Return an array of spatially binned gene expression counts with shape (n_x_bins, n_y_bins, n_genes)</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.bounds">
<span class="sig-name descname"><span class="pre">bounds</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.bounds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.bounds" title="Link to this definition"></a></dt>
<dd><p>Return ((xmin, xmax), (ymin, ymax)) giving the boundaries of data included in this table.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.calculate_cell_features">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">calculate_cell_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_polygon</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">z_plane_thickness</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.calculate_cell_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.calculate_cell_features" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.calculate_cell_polygons">
<span class="sig-name descname"><span class="pre">calculate_cell_polygons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alpha_inv</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">separate_z_planes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cells_to_run</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_inv_coeff</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disable_tqdm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.calculate_cell_polygons"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.calculate_cell_polygons" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.calculate_optimal_polygon">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">calculate_optimal_polygon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">xy_pos</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_inv</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_inv_coeff</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.calculate_optimal_polygon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.calculate_optimal_polygon" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.cell_bounds">
<span class="sig-name descname"><span class="pre">cell_bounds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.cell_bounds"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.cell_bounds" title="Link to this definition"></a></dt>
<dd><p>Return xmin, xmax, ymin, ymax for <em>cell_id</em></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.cell_by_gene_anndata">
<span class="sig-name descname"><span class="pre">cell_by_gene_anndata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x_format</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_dtype</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'uint16'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.cell_by_gene_anndata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.cell_by_gene_anndata" title="Link to this definition"></a></dt>
<dd><p>Return a cell x gene table in AnnData format with cell centroids in x,y obs columns.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.cell_by_gene_dense_matrix">
<span class="sig-name descname"><span class="pre">cell_by_gene_dense_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dtype</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'uint16'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.cell_by_gene_dense_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.cell_by_gene_dense_matrix" title="Link to this definition"></a></dt>
<dd><p>Return a numpy array containing a cell-by-gene table derived from this spot table.
Also return cell_ids and gene_ids used to construct the matrix</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.cell_by_gene_sparse_matrix">
<span class="sig-name descname"><span class="pre">cell_by_gene_sparse_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dtype</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'uint16'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.cell_by_gene_sparse_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.cell_by_gene_sparse_matrix" title="Link to this definition"></a></dt>
<dd><p>Return cell-by-gene data in a scipy.sparse.csr_matrix
Also return cell_ids and gene_ids used to construct the matrix</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.cell_centroids">
<span class="sig-name descname"><span class="pre">cell_centroids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">use_production_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.cell_centroids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.cell_centroids" title="Link to this definition"></a></dt>
<dd><p>Return a Pandas DataFrame of cell centroids calculated mean of the x,y,z coordinates for each cell.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.cell_ids">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">cell_ids</span></span><a class="headerlink" href="#sis.spot_table.SpotTable.cell_ids" title="Link to this definition"></a></dt>
<dd><p>An array of cell IDs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.cell_ids_changed">
<span class="sig-name descname"><span class="pre">cell_ids_changed</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.cell_ids_changed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.cell_ids_changed" title="Link to this definition"></a></dt>
<dd><p>Call when self.cell_ids has been modified to invalidate caches.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.cell_indices">
<span class="sig-name descname"><span class="pre">cell_indices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_ids</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.cell_indices"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.cell_indices" title="Link to this definition"></a></dt>
<dd><p>Return indices giving table location of all spots with <em>cell_ids</em></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.cell_indices_within_padding">
<span class="sig-name descname"><span class="pre">cell_indices_within_padding</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">padding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.cell_indices_within_padding"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.cell_indices_within_padding" title="Link to this definition"></a></dt>
<dd><p>Return spot indices all cells that do not come within <em>padding</em> of the parent_region.</p>
<p>This is used to exclude cells near the edge of a tile, where the segmentation becomes unreliable.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.cell_mask">
<span class="sig-name descname"><span class="pre">cell_mask</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_ids</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.cell_mask"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.cell_mask" title="Link to this definition"></a></dt>
<dd><p>Return a mask selecting spots that belong to cells in <em>cell_ids</em></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.cell_palette">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">cell_palette</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cells</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.cell_palette"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.cell_palette" title="Link to this definition"></a></dt>
<dd><p>Generate a color palette suitable for distinguisging individual cells</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.cell_polygon">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">cell_polygon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_points_array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_inv</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.cell_polygon"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.cell_polygon" title="Link to this definition"></a></dt>
<dd><p>get 2D alpha shape from a set of points, modified slightly from
<a class="reference external" href="http://blog.thehumangeo.com/2014/05/12/drawing-boundaries-in-python/">http://blog.thehumangeo.com/2014/05/12/drawing-boundaries-in-python/</a></p>
<p>args:
cell_points  numpy array of  point locations with expected columns [x,y].
alpha_inv parameter that sets the radius filter on the Delaunay triangulation.</p>
<blockquote>
<div><p>traditionally alpha is defined as 1/radius,
and here the function input is inverted for slightly more intuitive use</p>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.cell_scatter_plot">
<span class="sig-name descname"><span class="pre">cell_scatter_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.cell_scatter_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.cell_scatter_plot" title="Link to this definition"></a></dt>
<dd><p>Scatter plot of spots colored by cell ID</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.cells_inside_region">
<span class="sig-name descname"><span class="pre">cells_inside_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">xlim</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylim</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.cells_inside_region"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.cells_inside_region" title="Link to this definition"></a></dt>
<dd><p>Return IDs of cells that are entirely inside xlim, ylim.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.convert_cell_id">
<span class="sig-name descname"><span class="pre">convert_cell_id</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.convert_cell_id"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.convert_cell_id" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">deep</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.copy" title="Link to this definition"></a></dt>
<dd><p>Return a copy of self, optionally with some attributes replaced.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.dataframe">
<span class="sig-name descname"><span class="pre">dataframe</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.dataframe"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.dataframe" title="Link to this definition"></a></dt>
<dd><p>Return a dataframe containing the specified columns.</p>
<p>By default, columns are x, y, z, gene_ids, and cell_ids (if available).
Also available: gene_names</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.filter_cells">
<span class="sig-name descname"><span class="pre">filter_cells</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">real_cells</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_spot_count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.filter_cells"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.filter_cells" title="Link to this definition"></a></dt>
<dd><p>Return a filtered spot table containing only cells matching the filter criteria.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>real_cells</strong> (<em>bool</em><em>|</em><em>None</em>)  If True, include only spots that are segmented to a cell (cell IDs &gt; 0)</p></li>
<li><p><strong>min_spot_count</strong> (<em>int</em><em>|</em><em>None</em>)  Include only spots that are segmented to cells with a minimum number of spots</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.gene_indices">
<span class="sig-name descname"><span class="pre">gene_indices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.gene_indices"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.gene_indices" title="Link to this definition"></a></dt>
<dd><p>Return an array of indices where the detected transcript is in either gene_names or gene_ids</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.gene_names">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gene_names</span></span><a class="headerlink" href="#sis.spot_table.SpotTable.gene_names" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.generate_production_cell_ids">
<span class="sig-name descname"><span class="pre">generate_production_cell_ids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.generate_production_cell_ids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.generate_production_cell_ids" title="Link to this definition"></a></dt>
<dd><p>Generates cell ids which count up from 1 to the total cell count rather than jumping between integers.
Production cell ids are of type string to allow for concatenating a prefix and/or suffix to the id</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>prefix</strong> (<em>str</em>)  String to prepend to all production cell ids</p></li>
<li><p><strong>suffix</strong> (<em>str</em>)  String to postpend to all production cell ids</p></li>
<li><p><strong>Sets</strong>  </p></li>
<li><p><strong>----------</strong>  </p></li>
<li><p><strong>self.production_cell_ids</strong>  </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.get_cell_features">
<span class="sig-name descname"><span class="pre">get_cell_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">z_plane_thickness</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_production_ids</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_both_ids</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disable_tqdm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.get_cell_features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.get_cell_features" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.get_genes">
<span class="sig-name descname"><span class="pre">get_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.get_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.get_genes" title="Link to this definition"></a></dt>
<dd><p>Return a subtable containing only the genes specified by either gene_names or gene_ids.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.get_geojson_collection">
<span class="sig-name descname"><span class="pre">get_geojson_collection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">use_production_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.get_geojson_collection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.get_geojson_collection" title="Link to this definition"></a></dt>
<dd><p>Create a geojson feature/geometry collection from the cell polygons</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.get_image">
<span class="sig-name descname"><span class="pre">get_image</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">channel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frame</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.get_image"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.get_image" title="Link to this definition"></a></dt>
<dd><p>Return the image with the given name or channel name</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.get_subregion">
<span class="sig-name descname"><span class="pre">get_subregion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">xlim</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylim</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">incl_end</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.get_subregion"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.get_subregion" title="Link to this definition"></a></dt>
<dd><p>Return a SpotTable including the subset of this table inside the region xlim, ylim</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.grid_tiles">
<span class="sig-name descname"><span class="pre">grid_tiles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">max_tile_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">incl_end</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.grid_tiles"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.grid_tiles" title="Link to this definition"></a></dt>
<dd><p>Return a grid of overlapping tiles with equal size, where the width and height
must be less than max_tile_size.</p>
<p>See also: split_tiles</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.load_baysor">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_baysor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.load_baysor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.load_baysor" title="Link to this definition"></a></dt>
<dd><p>Return a new SpotTable loaded from a baysor result file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.load_cell_polygons">
<span class="sig-name descname"><span class="pre">load_cell_polygons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">load_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reset_cache</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disable_tqdm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.load_cell_polygons"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.load_cell_polygons" title="Link to this definition"></a></dt>
<dd><p>load cell polygons from a geojson feature collection file</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.load_json">
<span class="sig-name descname"><span class="pre">load_json</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">json_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.load_json"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.load_json" title="Link to this definition"></a></dt>
<dd><p>Return a subset of this table as described by a json file (previously saved with save_json)</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.load_merscope">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_merscope</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">csv_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_rows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.load_merscope"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.load_merscope" title="Link to this definition"></a></dt>
<dd><p>Load MERSCOPE data from a detected transcripts CSV file.</p>
<p>CSV reading is slow, so optionally cache the result to a .npz file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.load_npz">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_npz</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">npz_file</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.load_npz"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.load_npz" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.load_pickle">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_pickle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.load_pickle"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.load_pickle" title="Link to this definition"></a></dt>
<dd><p>Return a new SpotTable loaded a SpotTable pickle stored on disk</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.load_stereoseq">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_stereoseq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gem_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gem_cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">tuple</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">(('gene',</span> <span class="pre">0),</span> <span class="pre">('x',</span> <span class="pre">1),</span> <span class="pre">('y',</span> <span class="pre">2),</span> <span class="pre">('MIDcounts',</span> <span class="pre">3))</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skiprows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_rows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_channel</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.load_stereoseq"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.load_stereoseq" title="Link to this definition"></a></dt>
<dd><p>Load StereoSeq data from gem file. This can be slow so optionally cache the result to a .npz file.
1/19/2023: New StereoSeq data has cell_ids, add optional cell_cols to add to SpotTable. Also has a flag
for whether the spot was in the main cell or the extended cell (in_cell)
1/3/2024: new cellbin.gem doesnt have <cite>in_cell</cite>, lets not worry about it and just specify the column
that the cell ID is in</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.map_gene_ids_to_names">
<span class="sig-name descname"><span class="pre">map_gene_ids_to_names</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ids</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.map_gene_ids_to_names"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.map_gene_ids_to_names" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.map_gene_names_to_ids">
<span class="sig-name descname"><span class="pre">map_gene_names_to_ids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">names</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.map_gene_names_to_ids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.map_gene_names_to_ids" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.map_indices_from_parent">
<span class="sig-name descname"><span class="pre">map_indices_from_parent</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">inds</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.map_indices_from_parent"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.map_indices_from_parent" title="Link to this definition"></a></dt>
<dd><p>Given an array of indices into the parent SpotTable, return a new array of indices that
select the same spots in this SpotTable.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.map_indices_to_parent">
<span class="sig-name descname"><span class="pre">map_indices_to_parent</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">inds</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.map_indices_to_parent"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.map_indices_to_parent" title="Link to this definition"></a></dt>
<dd><p>Given an array of indices into this SpotTable, return a new array of indices that
select the same spots in the parent SpotTable.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.map_mask_to_parent">
<span class="sig-name descname"><span class="pre">map_mask_to_parent</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mask</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.map_mask_to_parent"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.map_mask_to_parent" title="Link to this definition"></a></dt>
<dd><p>Given a boolean mask that selects spots from this SpotTable, return a new boolean mask
that selects the same spots from the parent SpotTable.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.merge_cells">
<span class="sig-name descname"><span class="pre">merge_cells</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">padding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.merge_cells"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.merge_cells" title="Link to this definition"></a></dt>
<dd><p>Merge cell IDs from SpotTable <em>other</em> into self.</p>
<p>Returns a structure describing merge conflicts.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.plot_rect">
<span class="sig-name descname"><span class="pre">plot_rect</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.plot_rect"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.plot_rect" title="Link to this definition"></a></dt>
<dd><p>Plot the bounding region of this table as a rectangle.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.production_cell_ids">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">production_cell_ids</span></span><a class="headerlink" href="#sis.spot_table.SpotTable.production_cell_ids" title="Link to this definition"></a></dt>
<dd><p>An array of production cell IDs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.reduced_expression_map">
<span class="sig-name descname"><span class="pre">reduced_expression_map</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">binsize</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umap_args=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umap_ax=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm=&lt;function</span> <span class="pre">log_plus_1&gt;</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.reduced_expression_map"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.reduced_expression_map" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.save_cell_polygons">
<span class="sig-name descname"><span class="pre">save_cell_polygons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">save_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_production_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.save_cell_polygons"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.save_cell_polygons" title="Link to this definition"></a></dt>
<dd><p>save a geojson geometry collection from the cell polygons</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.save_csv">
<span class="sig-name descname"><span class="pre">save_csv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">columns</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.save_csv"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.save_csv" title="Link to this definition"></a></dt>
<dd><p>Save a CSV file with columns x, y, z, gene_id, [gene_name, cell_id].</p>
<p>Optionally, use the <em>columns</em> argument to specify which columns to write.
By default, the cell ID column is only present if cell IDs are available.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.save_json">
<span class="sig-name descname"><span class="pre">save_json</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">json_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.save_json"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.save_json" title="Link to this definition"></a></dt>
<dd><p>Save parent_region and parent_inds to json.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.save_npz">
<span class="sig-name descname"><span class="pre">save_npz</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">npz_file</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.save_npz"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.save_npz" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.scatter_plot">
<span class="sig-name descname"><span class="pre">scatter_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'x'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'y'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gene_ids'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">z_slice</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.scatter_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.scatter_plot" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.set_cell_ids_from_tiles">
<span class="sig-name descname"><span class="pre">set_cell_ids_from_tiles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tiles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">padding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.set_cell_ids_from_tiles"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.set_cell_ids_from_tiles" title="Link to this definition"></a></dt>
<dd><p>Overwrite all cell IDs by merging from <em>tiles</em>, which may be a list of SpotTable
or SegmentationResult instances.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.show_binned_heatmap">
<span class="sig-name descname"><span class="pre">show_binned_heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.show_binned_heatmap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.show_binned_heatmap" title="Link to this definition"></a></dt>
<dd><p>Show an image of binned spot positions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.show_image">
<span class="sig-name descname"><span class="pre">show_image</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">channel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">z_index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">z_pos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.show_image"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.show_image" title="Link to this definition"></a></dt>
<dd><p>Show a channel / z plane from an image</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.show_subregion_images">
<span class="sig-name descname"><span class="pre">show_subregion_images</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">subtables</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.show_subregion_images"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.show_subregion_images" title="Link to this definition"></a></dt>
<dd><p>Show a spot table image and successive subregions</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.split_tiles">
<span class="sig-name descname"><span class="pre">split_tiles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">max_spots_per_tile</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_tile_width</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">incl_end</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.split_tiles"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.split_tiles" title="Link to this definition"></a></dt>
<dd><p>Return a list of SpotTables that tile this one.</p>
<p>This table will be split into rows of equal height, and each row will be split into
columns with roughly the same number of spots (less than <em>max_spots_per_tile</em>).</p>
<p>see also: grid_tiles</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>max_spots_per_tile</strong> (<em>int</em><em> | </em><em>None</em>)  Maximum number of spots to include in each tile.</p></li>
<li><p><strong>target_tile_width</strong> (<em>float</em><em> | </em><em>None</em>)  Automatically select max_spots_per_tile to get an approximate tile width</p></li>
<li><p><strong>overlap</strong> (<em>float</em>)  Distance to overlap tiles</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.unique_cell_ids">
<span class="sig-name descname"><span class="pre">unique_cell_ids</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#SpotTable.unique_cell_ids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.SpotTable.unique_cell_ids" title="Link to this definition"></a></dt>
<dd><p>numpy array of unique cell ids (excluding background)</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.x">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">x</span></span><a class="headerlink" href="#sis.spot_table.SpotTable.x" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.y">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">y</span></span><a class="headerlink" href="#sis.spot_table.SpotTable.y" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.z">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">z</span></span><a class="headerlink" href="#sis.spot_table.SpotTable.z" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.spot_table.run_cell_polygon_calculation">
<span class="sig-prename descclassname"><span class="pre">sis.spot_table.</span></span><span class="sig-name descname"><span class="pre">run_cell_polygon_calculation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">load_func</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">load_args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subregion</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">tuple</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_id_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_subset_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">result_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_inv_coeff</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/spot_table.html#run_cell_polygon_calculation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.spot_table.run_cell_polygon_calculation" title="Link to this definition"></a></dt>
<dd><p>Load a spot table, calculate the cell polygons (possibly on a subset of cells), and save the result.</p>
</dd></dl>

</section>
<section id="module-sis.util">
<span id="sis-util-module"></span><h2>sis.util module<a class="headerlink" href="#module-sis.util" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="sis.util.example_function">
<span class="sig-prename descclassname"><span class="pre">sis.util.</span></span><span class="sig-name descname"><span class="pre">example_function</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/util.html#example_function"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.util.example_function" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.util.load_config">
<span class="sig-prename descclassname"><span class="pre">sis.util.</span></span><span class="sig-name descname"><span class="pre">load_config</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">configfile</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/util.html#load_config"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.util.load_config" title="Link to this definition"></a></dt>
<dd><p>loads configuration yaml file for spatial analysis. If no file is found, returns empty dict</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>configfile</strong>  path to yaml file. If None, will look for spatial_config.yml in the directory above this file</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>dict of configuration parameters</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.util.log_plus_1">
<span class="sig-prename descclassname"><span class="pre">sis.util.</span></span><span class="sig-name descname"><span class="pre">log_plus_1</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/util.html#log_plus_1"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.util.log_plus_1" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.util.make_cirro_compatible">
<span class="sig-prename descclassname"><span class="pre">sis.util.</span></span><span class="sig-name descname"><span class="pre">make_cirro_compatible</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_by_gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/util.html#make_cirro_compatible"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.util.make_cirro_compatible" title="Link to this definition"></a></dt>
<dd><p>Make an AnnData object compatible with Cirrocumulus visualization tool.</p>
<p>Copy cell spatial coordinates to obsm, as expected by Cirrocumulus
(<a class="reference external" href="https://cirrocumulus.readthedocs.io">https://cirrocumulus.readthedocs.io</a>). Also, generate UMAP from .X as an
additional visualization option.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>cell_by_gene</strong>  AnnData object output by
sis.spot_table.cell_by_gene_anndata()</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>copy of cell_by_gene with cirrocumulus-compatible fields</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>cell_by_gene_cirro</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.util.map_to_ubyte">
<span class="sig-prename descclassname"><span class="pre">sis.util.</span></span><span class="sig-name descname"><span class="pre">map_to_ubyte</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/util.html#map_to_ubyte"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.util.map_to_ubyte" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.util.package_for_10x">
<span class="sig-prename descclassname"><span class="pre">sis.util.</span></span><span class="sig-name descname"><span class="pre">package_for_10x</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">anndata_object</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_directory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_id_var_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dry_run</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exist_ok</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_category</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Supertype'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/util.html#package_for_10x"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.util.package_for_10x" title="Link to this definition"></a></dt>
<dd><p>takes a reference dataset as anndata object and writes a reference file that can be uploaded
to 10x as a reference for  Xenium  gene panel selection
see <a class="reference external" href="https://www.10xgenomics.com/support/in-situ-gene-expression/documentation/steps/panel-design/xenium-panel-getting-started#input-ref-anno">https://www.10xgenomics.com/support/in-situ-gene-expression/documentation/steps/panel-design/xenium-panel-getting-started#input-ref-anno</a></p>
<p>keyword arguments dry_run and exist_ok may help you avoid overwriting something important</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.util.plot_cbg_centroids">
<span class="sig-prename descclassname"><span class="pre">sis.util.</span></span><span class="sig-name descname"><span class="pre">plot_cbg_centroids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_by_gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'center_x'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'center_y'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/util.html#plot_cbg_centroids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.util.plot_cbg_centroids" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.util.plot_genes">
<span class="sig-prename descclassname"><span class="pre">sis.util.</span></span><span class="sig-name descname"><span class="pre">plot_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spottable</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">highlight_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subsample</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[15,</span> <span class="pre">15]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transpose_plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fontsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_background</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[0.65,</span> <span class="pre">0.65,</span> <span class="pre">0.65]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">markersize_background</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">markersize_highlight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">incoming_ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/util.html#plot_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.util.plot_genes" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.util.poly_to_geojson">
<span class="sig-prename descclassname"><span class="pre">sis.util.</span></span><span class="sig-name descname"><span class="pre">poly_to_geojson</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">polygon</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/util.html#poly_to_geojson"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.util.poly_to_geojson" title="Link to this definition"></a></dt>
<dd><p>turns a single shapely Polygon into a geojson polygon
:param polygon shapely.Polygon:</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>geojson polygon</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.util.rainbow_wheel">
<span class="sig-prename descclassname"><span class="pre">sis.util.</span></span><span class="sig-name descname"><span class="pre">rainbow_wheel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">points</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">center</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">radius</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">center_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/util.html#rainbow_wheel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.util.rainbow_wheel" title="Link to this definition"></a></dt>
<dd><p>Given an Nx2 array of point locations, return an Nx3 array of RGB
colors derived from a rainbow color wheel centered over the mean point location.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.util.reduce_expression">
<span class="sig-prename descclassname"><span class="pre">sis.util.</span></span><span class="sig-name descname"><span class="pre">reduce_expression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umap_args</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/util.html#reduce_expression"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.util.reduce_expression" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.util.show_cells_and_transcripts">
<span class="sig-prename descclassname"><span class="pre">sis.util.</span></span><span class="sig-name descname"><span class="pre">show_cells_and_transcripts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spottable</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">anndata_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">segmentation_geopandas</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes_to_highlight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_annotation_category</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'supertype_scANVI_leiden'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_annotation_values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_annotation_values_background</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_annotation_colormapping</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_blanks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loaded_image_array</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loaded_image_extent</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">initial_figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[20,</span> <span class="pre">20]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fontsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Greys'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">selected_cell_outline_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_patches</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skip_genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/util.html#show_cells_and_transcripts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.util.show_cells_and_transcripts" title="Link to this definition"></a></dt>
<dd><p>Parameters:</p>
<p>loaded_image_array</p>
<p>will take a 2D numpy array and use it for background. otherwise a maximum projection of image data from the SpotTable is created and used.</p>
<p>cell_annotation_category</p>
<p>column in anndata_obj.obs to get annotation information from</p>
<p>cell_annotation_values</p>
<p>values in <cite>cell_annotation_category</cite> to show</p>
<p><a href="#id4"><span class="problematic" id="id5">**</span></a>kwargs are passed to <cite>plot_genes</cite></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="sis.util.show_float_rgb">
<span class="sig-prename descclassname"><span class="pre">sis.util.</span></span><span class="sig-name descname"><span class="pre">show_float_rgb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extent</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/sis/util.html#show_float_rgb"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#sis.util.show_float_rgb" title="Link to this definition"></a></dt>
<dd><p>Show a color image given a WxHx3 array of floats.
Each channel is normalized independently.</p>
</dd></dl>

</section>
<section id="module-sis">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-sis" title="Link to this heading"></a></h2>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Allen Spatial Analysis Working Group.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>