

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>spot_table &mdash; Spots in Space  documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=5929fcd5"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="image" href="image.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            Spots in Space
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">spot_table</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#spottable">SpotTable</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sis.spot_table.SpotTable"><code class="docutils literal notranslate"><span class="pre">SpotTable</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.pos"><code class="docutils literal notranslate"><span class="pre">SpotTable.pos</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.parent_table"><code class="docutils literal notranslate"><span class="pre">SpotTable.parent_table</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.parent_inds"><code class="docutils literal notranslate"><span class="pre">SpotTable.parent_inds</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.parent_region"><code class="docutils literal notranslate"><span class="pre">SpotTable.parent_region</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.gene_ids"><code class="docutils literal notranslate"><span class="pre">SpotTable.gene_ids</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.gene_id_to_name"><code class="docutils literal notranslate"><span class="pre">SpotTable.gene_id_to_name</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.gene_name_to_id"><code class="docutils literal notranslate"><span class="pre">SpotTable.gene_name_to_id</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.images"><code class="docutils literal notranslate"><span class="pre">SpotTable.images</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable._gene_names"><code class="docutils literal notranslate"><span class="pre">SpotTable._gene_names</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.__init__"><code class="docutils literal notranslate"><span class="pre">SpotTable.__init__()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.dataframe"><code class="docutils literal notranslate"><span class="pre">SpotTable.dataframe()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.gene_names"><code class="docutils literal notranslate"><span class="pre">SpotTable.gene_names</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.x"><code class="docutils literal notranslate"><span class="pre">SpotTable.x</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.y"><code class="docutils literal notranslate"><span class="pre">SpotTable.y</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.z"><code class="docutils literal notranslate"><span class="pre">SpotTable.z</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.map_gene_names_to_ids"><code class="docutils literal notranslate"><span class="pre">SpotTable.map_gene_names_to_ids()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.map_gene_ids_to_names"><code class="docutils literal notranslate"><span class="pre">SpotTable.map_gene_ids_to_names()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.bounds"><code class="docutils literal notranslate"><span class="pre">SpotTable.bounds()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.get_subregion"><code class="docutils literal notranslate"><span class="pre">SpotTable.get_subregion()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.get_genes"><code class="docutils literal notranslate"><span class="pre">SpotTable.get_genes()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.detect_z_planes"><code class="docutils literal notranslate"><span class="pre">SpotTable.detect_z_planes()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.z_plane_mask"><code class="docutils literal notranslate"><span class="pre">SpotTable.z_plane_mask()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.save_csv"><code class="docutils literal notranslate"><span class="pre">SpotTable.save_csv()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.save_json"><code class="docutils literal notranslate"><span class="pre">SpotTable.save_json()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.load_json"><code class="docutils literal notranslate"><span class="pre">SpotTable.load_json()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.load_pickle"><code class="docutils literal notranslate"><span class="pre">SpotTable.load_pickle()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.read_merscope_csv"><code class="docutils literal notranslate"><span class="pre">SpotTable.read_merscope_csv()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.load_merscope"><code class="docutils literal notranslate"><span class="pre">SpotTable.load_merscope()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.read_stereoseq_gem"><code class="docutils literal notranslate"><span class="pre">SpotTable.read_stereoseq_gem()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.load_stereoseq"><code class="docutils literal notranslate"><span class="pre">SpotTable.load_stereoseq()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.read_xenium_transcripts"><code class="docutils literal notranslate"><span class="pre">SpotTable.read_xenium_transcripts()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.load_xenium"><code class="docutils literal notranslate"><span class="pre">SpotTable.load_xenium()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.save_npz"><code class="docutils literal notranslate"><span class="pre">SpotTable.save_npz()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.load_npz"><code class="docutils literal notranslate"><span class="pre">SpotTable.load_npz()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.gene_indices"><code class="docutils literal notranslate"><span class="pre">SpotTable.gene_indices()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.map_indices_to_parent"><code class="docutils literal notranslate"><span class="pre">SpotTable.map_indices_to_parent()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.map_indices_from_parent"><code class="docutils literal notranslate"><span class="pre">SpotTable.map_indices_from_parent()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.map_mask_to_parent"><code class="docutils literal notranslate"><span class="pre">SpotTable.map_mask_to_parent()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.copy"><code class="docutils literal notranslate"><span class="pre">SpotTable.copy()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.split_tiles"><code class="docutils literal notranslate"><span class="pre">SpotTable.split_tiles()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.grid_tiles"><code class="docutils literal notranslate"><span class="pre">SpotTable.grid_tiles()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.plot_rect"><code class="docutils literal notranslate"><span class="pre">SpotTable.plot_rect()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.scatter_plot"><code class="docutils literal notranslate"><span class="pre">SpotTable.scatter_plot()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.binned_expression_counts"><code class="docutils literal notranslate"><span class="pre">SpotTable.binned_expression_counts()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.reduced_expression_map"><code class="docutils literal notranslate"><span class="pre">SpotTable.reduced_expression_map()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.show_binned_heatmap"><code class="docutils literal notranslate"><span class="pre">SpotTable.show_binned_heatmap()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.add_image"><code class="docutils literal notranslate"><span class="pre">SpotTable.add_image()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.get_image"><code class="docutils literal notranslate"><span class="pre">SpotTable.get_image()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SpotTable.show_image"><code class="docutils literal notranslate"><span class="pre">SpotTable.show_image()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#segmentedspottable">SegmentedSpotTable</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.spot_table"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.spot_table</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.cell_polygons"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.cell_polygons</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.seg_metadata"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.seg_metadata</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable._old_cell_ids"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable._old_cell_ids</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable._cell_ids"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable._cell_ids</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable._cell_labels"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable._cell_labels</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable._cl_to_cid"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable._cl_to_cid</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable._cid_to_cl"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable._cid_to_cl</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable._cell_index"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable._cell_index</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable._cell_bounds"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable._cell_bounds</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable._unique_cell_ids"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable._unique_cell_ids</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.__init__"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.__init__()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.__getattr__"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.__getattr__()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.cell_ids"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.cell_ids</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.cell_labels"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.cell_labels</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.convert_cell_id"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.convert_cell_id()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.convert_cell_label"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.convert_cell_label()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.cell_ids_changed"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.cell_ids_changed()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.unique_cell_ids"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.unique_cell_ids</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.generate_cell_labels"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.generate_cell_labels()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.filter_cells"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.filter_cells()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.cell_by_gene_dense_matrix"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.cell_by_gene_dense_matrix()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.cell_by_gene_sparse_matrix"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.cell_by_gene_sparse_matrix()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.cell_by_gene_anndata"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.cell_by_gene_anndata()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.cell_bounds"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.cell_bounds()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.cell_centroids"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.cell_centroids()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.cell_polygon"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.cell_polygon()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.calculate_optimal_polygon"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.calculate_optimal_polygon()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.calculate_cell_polygons"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.calculate_cell_polygons()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.calculate_cell_features"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.calculate_cell_features()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.calculate_all_cell_features"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.calculate_all_cell_features()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.get_geojson_collection"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.get_geojson_collection()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.save_cell_polygons"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.save_cell_polygons()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.load_cell_polygons"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.load_cell_polygons()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.cell_indices"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.cell_indices()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.cells_inside_region"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.cells_inside_region()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.cell_mask"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.cell_mask()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.cell_indices_within_padding"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.cell_indices_within_padding()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.merge_cells"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.merge_cells()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.set_cell_ids_from_tiles"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.set_cell_ids_from_tiles()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.load_merscope_cell_ids"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.load_merscope_cell_ids()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.load_merscope"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.load_merscope()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.load_xenium"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.load_xenium()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.load_stereoseq"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.load_stereoseq()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.load_npz"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.load_npz()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.save_npz"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.save_npz()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.dataframe"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.dataframe()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.cell_scatter_plot"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.cell_scatter_plot()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.scatter_plot"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.scatter_plot()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.cell_palette"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.cell_palette()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.copy"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.copy()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.get_subregion"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.get_subregion()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable.save_xenium_kit_cbg"><code class="docutils literal notranslate"><span class="pre">SegmentedSpotTable.save_xenium_kit_cbg()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="image.html">image</a></li>
<li class="toctree-l1"><a class="reference internal" href="segmentation.html">segmentation</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Spots in Space</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">spot_table</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/spot_table.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="spot-table">
<h1>spot_table<a class="headerlink" href="#spot-table" title="Link to this heading"></a></h1>
<section id="spottable">
<h2>SpotTable<a class="headerlink" href="#spottable" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.spot_table.</span></span><span class="sig-name descname"><span class="pre">SpotTable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pos</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_ids</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_id_to_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent_table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable"><span class="pre">SpotTable</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent_inds</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent_region</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">images</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="image.html#sis.image.ImageBase" title="sis.image.ImageBase"><span class="pre">ImageBase</span></a><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference internal" href="image.html#sis.image.ImageBase" title="sis.image.ImageBase"><span class="pre">ImageBase</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Represents a spatial transcriptomics spot table.</p>
<p>This class represents a spatial transcriptomics spot table, which contains information about the position of each detected transcript, the associated gene, and other optional attributes. It can be used to manipulate and analyze spatial transcriptomics data.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.pos">
<span class="sig-name descname"><span class="pre">pos</span></span><a class="headerlink" href="#sis.spot_table.SpotTable.pos" title="Link to this definition"></a></dt>
<dd><p>Array of shape (N, 3) giving the position of each detected transcript.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.parent_table">
<span class="sig-name descname"><span class="pre">parent_table</span></span><a class="headerlink" href="#sis.spot_table.SpotTable.parent_table" title="Link to this definition"></a></dt>
<dd><p>Indicates that this table is a subset of a parent SpotTable.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable">SpotTable</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.parent_inds">
<span class="sig-name descname"><span class="pre">parent_inds</span></span><a class="headerlink" href="#sis.spot_table.SpotTable.parent_inds" title="Link to this definition"></a></dt>
<dd><p>Indices used to select the subset of spots in this table from the parent spots.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.parent_region">
<span class="sig-name descname"><span class="pre">parent_region</span></span><a class="headerlink" href="#sis.spot_table.SpotTable.parent_region" title="Link to this definition"></a></dt>
<dd><p>X,Y boundaries ((xmin, xmax), (ymin, ymax)) used to select this table from the parent table.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.gene_ids">
<span class="sig-name descname"><span class="pre">gene_ids</span></span><a class="headerlink" href="#sis.spot_table.SpotTable.gene_ids" title="Link to this definition"></a></dt>
<dd><p>Array of shape (N,) describing the gene detected in each transcript, as an index into <em>gene_id_to_name</em>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.gene_id_to_name">
<span class="sig-name descname"><span class="pre">gene_id_to_name</span></span><a class="headerlink" href="#sis.spot_table.SpotTable.gene_id_to_name" title="Link to this definition"></a></dt>
<dd><p>Array mapping from values in <em>gene_ids</em> to string names.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.gene_name_to_id">
<span class="sig-name descname"><span class="pre">gene_name_to_id</span></span><a class="headerlink" href="#sis.spot_table.SpotTable.gene_name_to_id" title="Link to this definition"></a></dt>
<dd><p>Dictionary mapping from gene names to gene IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.images">
<span class="sig-name descname"><span class="pre">images</span></span><a class="headerlink" href="#sis.spot_table.SpotTable.images" title="Link to this definition"></a></dt>
<dd><p>Image(s) associated with the data (e.g. nuclei stain).</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[<a class="reference internal" href="image.html#sis.image.ImageBase" title="sis.image.ImageBase">ImageBase</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable._gene_names">
<span class="sig-name descname"><span class="pre">_gene_names</span></span><a class="headerlink" href="#sis.spot_table.SpotTable._gene_names" title="Link to this definition"></a></dt>
<dd><p>Cached array of gene names corresponding to the gene IDs in the SpotTable.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pos</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_ids</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_id_to_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent_table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable"><span class="pre">SpotTable</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent_inds</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent_region</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">images</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="image.html#sis.image.ImageBase" title="sis.image.ImageBase"><span class="pre">ImageBase</span></a><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference internal" href="image.html#sis.image.ImageBase" title="sis.image.ImageBase"><span class="pre">ImageBase</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.__init__" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pos</strong> (<em>numpy.ndarray</em>) – Array of shape (N, 3) giving the position of each detected transcript.</p></li>
<li><p><strong>gene_names</strong> (<em>numpy.ndarray</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Array of shape (N,) giving the name of the gene detected in each transcript. Must specify either <em>gene_names</em> or <em>gene_ids</em>, not both.</p></li>
<li><p><strong>gene_ids</strong> (<em>numpy.ndarray</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Array of shape (N,) describing the gene detected in each transcript, as an index into <em>gene_id_to_name</em>. Must specify either <em>gene_names</em> or <em>gene_ids</em>, not both.</p></li>
<li><p><strong>gene_id_to_name</strong> (<em>numpy.ndarray</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Array mapping from values in <em>gene_ids</em> to string names.</p></li>
<li><p><strong>parent_table</strong> (<a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable"><em>SpotTable</em></a><em> or </em><em>None</em><em>, </em><em>optional</em>) – Indicates that this table is a subset of a parent SpotTable.</p></li>
<li><p><strong>parent_inds</strong> (<em>numpy.ndarray</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Indices used to select the subset of spots in this table from the parent spots.</p></li>
<li><p><strong>parent_region</strong> (<em>tuple</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – X,Y boundaries ((xmin, xmax), (ymin, ymax)) used to select this table from the parent table.</p></li>
<li><p><strong>images</strong> (<em>list</em><em>[</em><a class="reference internal" href="image.html#sis.image.ImageBase" title="sis.image.ImageBase"><em>ImageBase</em></a><em>] or </em><a class="reference internal" href="image.html#sis.image.ImageBase" title="sis.image.ImageBase"><em>ImageBase</em></a><em> or </em><em>None</em><em>, </em><em>optional</em>) – Image(s) associated with the data (e.g. nuclei stain).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.dataframe">
<span class="sig-name descname"><span class="pre">dataframe</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['x',</span> <span class="pre">'y',</span> <span class="pre">'z',</span> <span class="pre">'gene_ids']</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.dataframe" title="Link to this definition"></a></dt>
<dd><p>Return a dataframe containing the specified columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>cols</strong> (<em>List</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – The columns to include in the dataframe. Default is [‘x’, ‘y’, ‘z’, ‘gene_ids’].
Additional available columns: ‘gene_names’.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.gene_names">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gene_names</span></span><a class="headerlink" href="#sis.spot_table.SpotTable.gene_names" title="Link to this definition"></a></dt>
<dd><p>Return an array of gene names corresponding to the gene IDs in the SpotTable.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>self._gene_names</strong></p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.x">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">x</span></span><a class="headerlink" href="#sis.spot_table.SpotTable.x" title="Link to this definition"></a></dt>
<dd><p>Return the x-coordinates of the spots in the SpotTable.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Array of x-coordinates of the spots in the SpotTable (self.pos[:, 0])</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.y">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">y</span></span><a class="headerlink" href="#sis.spot_table.SpotTable.y" title="Link to this definition"></a></dt>
<dd><p>Return the y-coordinates of the spots in the SpotTable.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><ul class="simple">
<li><p><em>None</em> – If no y-coordinates are available, returns None.</p></li>
<li><p><em>numpy.ndarray</em> – Array of y-coordinates of the spots in the SpotTable (self.pos[:, 1])</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.z">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">z</span></span><a class="headerlink" href="#sis.spot_table.SpotTable.z" title="Link to this definition"></a></dt>
<dd><p>Return the z-coordinates of the spots in the SpotTable.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><ul class="simple">
<li><p><em>None</em> – If no z-coordinates are available, returns None.</p></li>
<li><p><em>numpy.ndarray</em> – Array of z-coordinates of the spots in the SpotTable (self.pos[:, 2])</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.map_gene_names_to_ids">
<span class="sig-name descname"><span class="pre">map_gene_names_to_ids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">names</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.map_gene_names_to_ids" title="Link to this definition"></a></dt>
<dd><p>Map gene names to gene IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>names</strong> (<em>array</em>) – Array of gene names to be mapped to gene IDs.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>out</strong> – Array of gene IDs corresponding to the input gene names.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>array</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.map_gene_ids_to_names">
<span class="sig-name descname"><span class="pre">map_gene_ids_to_names</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ids</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.map_gene_ids_to_names" title="Link to this definition"></a></dt>
<dd><p>Map gene IDs to gene names.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ids</strong> (<em>array</em>) – Array of gene IDs to be mapped to gene names.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>out</strong> – Array of gene names corresponding to the input gene IDs.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>array</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.bounds">
<span class="sig-name descname"><span class="pre">bounds</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.bounds" title="Link to this definition"></a></dt>
<dd><p>Return ((xmin, xmax), (ymin, ymax)) giving the boundaries of data included in this table.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A tuple containing two tuples: (xmin, xmax) and (ymin, ymax) representing the boundaries of the data included in this table.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.get_subregion">
<span class="sig-name descname"><span class="pre">get_subregion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">xlim</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylim</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">incl_end</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.get_subregion" title="Link to this definition"></a></dt>
<dd><p>Return a SpotTable including the subset of this table inside the region xlim, ylim.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>xlim</strong> (<em>tuple</em>) – X-axis boundaries (xmin, xmax) of the region.</p></li>
<li><p><strong>ylim</strong> (<em>tuple</em>) – Y-axis boundaries (ymin, ymax) of the region.</p></li>
<li><p><strong>incl_end</strong> (<em>bool</em><em>, </em><em>optional</em>) – Flag indicating whether to include all pixels of the image that overlap with the region,
rather than just those inside the region (i.e. ceil vs floor). Default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>sub</strong> – A new SpotTable object containing the subset of spots and image inside the specified region.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable">SpotTable</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.get_genes">
<span class="sig-name descname"><span class="pre">get_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.get_genes" title="Link to this definition"></a></dt>
<dd><p>Return a subtable containing only the genes specified by either gene_names or gene_ids.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_names</strong> (<em>array</em><em> or </em><em>None</em>) – Array of gene names to include in the subtable.</p></li>
<li><p><strong>gene_ids</strong> (<em>array</em><em> or </em><em>None</em>) – Array of gene IDs to include in the subtable.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A new SpotTable object containing only the specified genes.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable">SpotTable</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.detect_z_planes">
<span class="sig-name descname"><span class="pre">detect_z_planes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">float_cut</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.detect_z_planes" title="Link to this definition"></a></dt>
<dd><p>Return a tuple containing the minimum and maximum z-planes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>float_cut</strong> (<em>float</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – If specified, only include z-planes that contain at least this fraction of spots.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple containing the minimum and maximum z-planes. The minimum z-plane is inclusive, while the maximum z-plane is exclusive.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.z_plane_mask">
<span class="sig-name descname"><span class="pre">z_plane_mask</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">z_planes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.z_plane_mask" title="Link to this definition"></a></dt>
<dd><p>Return a mask of the SpotTable containing only spots in the specified z_planes
Z-planes are specified as a tuple [min, max) to align with the output of detect_z_planes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>z_planes</strong> (<em>tuple</em>) – contains the min (inclusive) and max (exclusive) z_planes to keep</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>mask</strong> – A boolean mask indicating which spots are in the specified z-planes.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.save_csv">
<span class="sig-name descname"><span class="pre">save_csv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">columns</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.save_csv" title="Link to this definition"></a></dt>
<dd><p>Save a CSV file with columns x, y, z, gene_id, [gene_name, cell_id].</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> (<em>str</em>) – The name of the CSV file to be saved.</p></li>
<li><p><strong>columns</strong> (<em>list</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – A list of column names to be included in the CSV file.
If not provided, the default columns [‘x’, ‘y’, ‘z’, ‘gene_id’] will be used.
If cell IDs are available, the ‘cell_id’ column will also be included.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The cell ID column is only present if cell IDs are available.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.save_json">
<span class="sig-name descname"><span class="pre">save_json</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">json_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.save_json" title="Link to this definition"></a></dt>
<dd><p>Save the parent region and parent indices to a JSON file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>json_file</strong> (<em>str</em>) – The path to the JSON file.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.load_json">
<span class="sig-name descname"><span class="pre">load_json</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">json_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.load_json" title="Link to this definition"></a></dt>
<dd><p>Return a subset of this table as described by a json file (previously saved with save_json)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>json_file</strong> (<em>str</em>) – The path to the JSON file.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.load_pickle">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_pickle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.load_pickle" title="Link to this definition"></a></dt>
<dd><p>Return a new SpotTable loaded a SpotTable pickle stored on disk</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>file_name</strong> (<em>str</em>) – The path to the pickle file.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.read_merscope_csv">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">read_merscope_csv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">csv_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols_to_use</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('x',</span> <span class="pre">'y',</span> <span class="pre">'z',</span> <span class="pre">'gene')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_rows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.read_merscope_csv" title="Link to this definition"></a></dt>
<dd><p>Helper function to read a MERSCOPE csv file.
Intended to reduce duplicated code between SpotTable.load_merscope()
and SegmentedSpotTable.load_merscope().</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>csv_file</strong> (<em>str</em>) – Path to the MERSCOPE CSV file.</p></li>
<li><p><strong>cols_to_use</strong> (<em>tuple</em>) – Columns to use from the CSV file.</p></li>
<li><p><strong>max_rows</strong> (<em>int</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Maximum number of rows to read from the CSV file.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p><strong>(raw_data, pos, gene_names)</strong> –</p>
<ul class="simple">
<li><p>Raw data read from the CSV file.</p></li>
<li><p>Positions of the spots.</p></li>
<li><p>Gene names corresponding to the spots.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[numpy.ndarray, numpy.ndarray, numpy.ndarray]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.load_merscope">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_merscope</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">csv_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_rows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.load_merscope" title="Link to this definition"></a></dt>
<dd><p>Load MERSCOPE data from a detected transcripts CSV file. This is the
preferred method for resegmentation. If you want the original MERSCOPE
segmentation, use SegmentedSpotTable.load_merscope.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>csv_file</strong> (<em>str</em>) – Path to the detected transcripts file.</p></li>
<li><p><strong>cache_file</strong> (<em>str</em><em> or </em><em>None</em>) – Path to the detected transcripts cache file, which is an npz file
representing the raw SpotTable (without cell_ids). If passed, will
create a cache file if one does not already exists.</p></li>
<li><p><strong>image_path</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Path to directory containing a MERSCOPE image stack.</p></li>
<li><p><strong>max_rows</strong> (<em>int</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Maximum number of rows to load from the CSV file.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>SpotTable with data loaded from inputs</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable">sis.spot_table.SpotTable</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.read_stereoseq_gem">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">read_stereoseq_gem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gem_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gem_cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skiprows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_rows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.read_stereoseq_gem" title="Link to this definition"></a></dt>
<dd><p>Helper function to read raw data from a StereoSeq dataset.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gem_file</strong> (<em>str</em>) – Path to the GEM file.</p></li>
<li><p><strong>gem_cols</strong> (<em>dict</em><em> or </em><em>tuple</em>) – Dictionary or tuple specifying the columns to use from the GEM file.</p></li>
<li><p><strong>cell_col</strong> (<em>int</em><em> or </em><em>None</em>) – Column index for cell IDs. If None, cell IDs are not included.</p></li>
<li><p><strong>skiprows</strong> (<em>int</em><em> or </em><em>None</em>) – Number of rows to skip at the beginning of the GEM file.</p></li>
<li><p><strong>max_rows</strong> (<em>int</em><em> or </em><em>None</em>) – Maximum number of rows to read from the GEM file.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>(pos, gene_ids, gene_id_to_name, cell_ids)</strong> (<em>tuple[numpy.ndarray, numpy.ndarray, numpy.ndarray, numpy.ndarray]</em>) –</p>
<ul>
<li><p>Array of shape (N, 2) giving the position of each detected transcript.</p></li>
<li><p>Array of shape (N,) describing the gene detected in each transcript, as an index into <em>gene_id_to_name</em>.</p></li>
<li><p>Array mapping from values in <em>gene_ids</em> to string names.</p></li>
<li><p>Array of shape (N,) giving the cell ID for each detected transcript. Only included if <em>cell_col</em> is not None.</p></li>
</ul>
</li>
<li><p><strong>(pos, gene_ids, gene_id_to_name)</strong> (<em>tuple[numpy.ndarray, numpy.ndarray, numpy.ndarray]</em>) –</p>
<ul>
<li><p>Array of shape (N, 2) giving the position of each detected transcript.</p></li>
<li><p>Array of shape (N,) describing the gene detected in each transcript, as an index into <em>gene_id_to_name</em>.</p></li>
<li><p>Array mapping from values in <em>gene_ids</em> to string names.</p></li>
</ul>
</li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.load_stereoseq">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_stereoseq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gem_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gem_cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">tuple</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">(('gene',</span> <span class="pre">0),</span> <span class="pre">('x',</span> <span class="pre">1),</span> <span class="pre">('y',</span> <span class="pre">2),</span> <span class="pre">('MIDcounts',</span> <span class="pre">3))</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skiprows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_rows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_channel</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.load_stereoseq" title="Link to this definition"></a></dt>
<dd><p>Load StereoSeq data from gem file. This can be slow so optionally cache the result to a .npz file.
1/19/2023: New StereoSeq data has cell_ids, add optional cell_cols to add to SpotTable. Also has a flag
for whether the spot was in the main cell or the extended cell (in_cell)
1/3/2024: new cellbin.gem doesn’t have <cite>in_cell</cite>, let’s not worry about it and just specify the column
that the cell ID is in</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gem_file</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Path to the GEM file.</p></li>
<li><p><strong>cache_file</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Path to the detected transcripts cache file, which is an npz file
representing the raw SpotTable (without cell_ids). If passed, will
create a cache file if one does not already exists.</p></li>
<li><p><strong>gem_cols</strong> (<em>dict</em><em> or </em><em>tuple</em><em>, </em><em>optional</em>) – Dictionary or tuple specifying the columns to use from the GEM file.</p></li>
<li><p><strong>cell_col</strong> (<em>int</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Column index for cell IDs. If None, cell IDs are not included.</p></li>
<li><p><strong>skiprows</strong> (<em>int</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Number of rows to skip at the beginning of the GEM file.</p></li>
<li><p><strong>max_rows</strong> (<em>int</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Maximum number of rows to read from the GEM file.</p></li>
<li><p><strong>image_file</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Path to the image file. If None, no image is loaded.</p></li>
<li><p><strong>image_channel</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Channel name for the image. If None, no channel is specified.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>SpotTable with data loaded from inputs.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable">sis.spot_table.SpotTable</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.read_xenium_transcripts">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">read_xenium_transcripts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">transcript_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_rows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">z_depth</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">3.0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.read_xenium_transcripts" title="Link to this definition"></a></dt>
<dd><p>Helper function to read a Xenium transcripts file. (currently supports only csv and parquet)
Intended to reduce duplicated code between SpotTable.load_xenium()
and SegmentedSpotTable.load_xenium().</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>transcript_file</strong> (<em>str</em>) – Path to the detected transcripts file.</p></li>
<li><p><strong>max_rows</strong> (<em>int</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Maximum number of rows to read from the CSV file.</p></li>
<li><p><strong>z_depth</strong> (<em>float</em><em>, </em><em>optional</em>) – Depth (in um) of a imaging layer i.e. z-plane
Used to bin z-positiions into discrete planes</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p><strong>(pos, gene_names)</strong> –</p>
<ul class="simple">
<li><p>Array of shape (N, 3) giving the position of each detected transcript.</p></li>
<li><p>Array of shape (N,) giving the name of the gene detected in each transcript.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[numpy.ndarray, numpy.ndarray]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.load_xenium">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_xenium</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">transcript_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_rows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">z_depth</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">3.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pyramid_level</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_image</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.load_xenium" title="Link to this definition"></a></dt>
<dd><dl class="simple">
<dt>Load Xenium data from a detected transcripts CSV file.</dt><dd><p>This is the preferred method for resegmentation. If you want the original Xenium
segmentation, use SegmentedSpotTable.load_xenium.
CSV reading is slow, so optionally cache the result to a .npz file.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>transcript_file</strong> (<em>str</em>) – Path to the detected transcripts file.</p></li>
<li><p><strong>cache_file</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the detected transcripts cache file, which is an npz file
representing the raw SpotTable (without cell_ids). If passed, will
create a cache file if one does not already exists.</p></li>
<li><p><strong>image_path</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to directory containing a Xenium image stack.</p></li>
<li><p><strong>max_rows</strong> (<em>int</em><em>, </em><em>optional</em>) – Maximum number of rows to load from the CSV file.</p></li>
<li><p><strong>z_depth</strong> (<em>float</em><em>, </em><em>optional</em>) – Depth (in um) of a imaging layer i.e. z-plane
Used to bin z-positiions into discrete planes</p></li>
<li><p><strong>pyramid_level</strong> (<em>int</em><em>, </em><em>optional</em>) – Xenium images can have multiple resolutions stored in an image pyramid.
This parameter specifies which level of the pyramid to load.
If None, we default to highest resolution</p></li>
<li><p><strong>cache_image</strong> (<em>bool</em><em>, </em><em>optional</em>) – Xenium images are large and not memory mapped and thus we may want to keep them in memory or not.
The trade off is speed vs memory.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>SpotTable with data loaded from inputs.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable">sis.spot_table.SpotTable</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.save_npz">
<span class="sig-name descname"><span class="pre">save_npz</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">npz_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.save_npz" title="Link to this definition"></a></dt>
<dd><p>Save this SpotTable as an NPZ file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>npz_file</strong> (<em>str</em>) – Path to the NPZ file.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.load_npz">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_npz</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">npz_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">images</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="image.html#sis.image.ImageBase" title="sis.image.ImageBase"><span class="pre">ImageBase</span></a><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference internal" href="image.html#sis.image.ImageBase" title="sis.image.ImageBase"><span class="pre">ImageBase</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.load_npz" title="Link to this definition"></a></dt>
<dd><p>Load from an NPZ file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>npz_file</strong> (<em>str</em>) – Path to the npz file.</p></li>
<li><p><strong>images</strong> (<a class="reference internal" href="image.html#sis.image.ImageBase" title="sis.image.ImageBase"><em>ImageBase</em></a><em> or </em><em>list</em><em>[</em><a class="reference internal" href="image.html#sis.image.ImageBase" title="sis.image.ImageBase"><em>ImageBase</em></a><em>] or </em><em>None</em><em>, </em><em>optional</em>) – Image(s) to attach to the SpotTable. Must be loaded separately
since these cannot be stored in the NPZ file.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Spot table loaded from compressed NPZ file</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable">sis.spot_table.SpotTable</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.gene_indices">
<span class="sig-name descname"><span class="pre">gene_indices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.gene_indices" title="Link to this definition"></a></dt>
<dd><p>Return an array of indices where the detected transcript is in either gene_names or gene_ids</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_names</strong> (<em>array</em><em> or </em><em>None</em>) – Array of gene names to retrieve indices for. Specify either this or gene_ids.</p></li>
<li><p><strong>gene_ids</strong> (<em>array</em><em> or </em><em>None</em>) – Array of gene IDs to retrieve indices for.  Specify either this or gene_names.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>inds</strong> – Array of indices where the detected transcript is in either gene_names or gene_ids.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.map_indices_to_parent">
<span class="sig-name descname"><span class="pre">map_indices_to_parent</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">inds</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.map_indices_to_parent" title="Link to this definition"></a></dt>
<dd><p>Given an array of indices in this SpotTable, return a new array of indices that
select the same spots in the parent SpotTable.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>inds</strong> (<em>np.ndarray</em>) – Array of indices in SpotTable to be mapped to parent SpotTable.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Array of indices into the parent SpotTable.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.map_indices_from_parent">
<span class="sig-name descname"><span class="pre">map_indices_from_parent</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">inds</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.map_indices_from_parent" title="Link to this definition"></a></dt>
<dd><p>Given an array of indices into the parent SpotTable, return a new array of indices that
select the same spots in this SpotTable.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>inds</strong> (<em>np.ndarray</em>) – Array of indices in the parent SpotTable to be mapped to this SpotTable</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Array of indices into this SpotTable.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.map_mask_to_parent">
<span class="sig-name descname"><span class="pre">map_mask_to_parent</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mask</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.map_mask_to_parent" title="Link to this definition"></a></dt>
<dd><p>Given a boolean mask that selects spots from this SpotTable, return a new boolean mask
that selects the same spots from the parent SpotTable.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>mask</strong> (<em>np.ndarray</em>) – Boolean mask that selects spots from this SpotTable to be mapped to the parent SpotTable.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Boolean mask that selects the same spots from the parent SpotTable.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">deep</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.copy" title="Link to this definition"></a></dt>
<dd><p>Return a copy of self, optionally with some attributes replaced.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>deep</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, make a deep copy of the SpotTable.</p></li>
<li><p><strong>**kwds</strong> (<em>dict</em><em>[</em><em>str</em><em>, </em><em>Any</em><em>]</em><em>, </em><em>optional</em>) – Attributes to replace in the copied SpotTable.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A new SpotTable object with the same attributes as self, but with some attributes replaced.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable">sis.spot_table.SpotTable</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.split_tiles">
<span class="sig-name descname"><span class="pre">split_tiles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">max_spots_per_tile</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_tile_width</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">incl_end</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_transcripts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.split_tiles" title="Link to this definition"></a></dt>
<dd><p>Return a list of SpotTables that tile this one.</p>
<p>This table will be split into rows of equal height, and each row will be split into
columns with roughly the same number of spots (less than <em>max_spots_per_tile</em>).</p>
<p>see also: grid_tiles</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>max_spots_per_tile</strong> (<em>int</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Maximum number of spots to include in each tile.</p></li>
<li><p><strong>target_tile_width</strong> (<em>float</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Automatically select max_spots_per_tile to get an approximate tile width</p></li>
<li><p><strong>overlap</strong> (<em>float</em><em>, </em><em>optional</em>) – Distance to overlap tiles</p></li>
<li><p><strong>incl_end</strong> (<em>bool</em><em>, </em><em>optional</em>) – Include all pixels of the image that overlap with the region, rather than just those inside the region.</p></li>
<li><p><strong>min_transcripts</strong> (<em>int</em><em>, </em><em>optional</em>) – Minimum number of transcripts in a tile to be returned</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>filter_tiles</strong> – A list of SpotTables that tile this one.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[SpotTables]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.grid_tiles">
<span class="sig-name descname"><span class="pre">grid_tiles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">max_tile_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">incl_end</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_transcripts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.grid_tiles" title="Link to this definition"></a></dt>
<dd><p>Return a grid of overlapping tiles with equal size, where the width and height
must be less than max_tile_size.</p>
<p>See also: split_tiles</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>max_tile_size</strong> (<em>int</em>) – Maximum width and height of each tile.</p></li>
<li><p><strong>overlap</strong> (<em>float</em><em>, </em><em>optional</em>) – Distance to overlap tiles</p></li>
<li><p><strong>incl_end</strong> (<em>bool</em><em>, </em><em>optional</em>) – Include all pixels of the image that overlap with the region, rather than just those inside the region.</p></li>
<li><p><strong>min_transcripts</strong> (<em>int</em><em>, </em><em>optional</em>) – Minimum number of transcripts in a tile to be returned</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>filter_tiles</strong> – A list of SpotTables that tile this one.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[SpotTables]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.plot_rect">
<span class="sig-name descname"><span class="pre">plot_rect</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.plot_rect" title="Link to this definition"></a></dt>
<dd><p>Plot the bounding region of this table as a rectangle.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em>) – The axes to plot the rectangle on.</p></li>
<li><p><strong>color</strong> (<em>str</em>) – The color of the rectangle.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>rect</strong> – The rectangle object that was added to the axes.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>matplotlib.patches.Rectangle</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.scatter_plot">
<span class="sig-name descname"><span class="pre">scatter_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'x'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'y'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gene_ids'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">z_idx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">z_pos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">palette</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.scatter_plot" title="Link to this definition"></a></dt>
<dd><p>Plot a scatter plot of the spots in this table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em><em>, </em><em>optional</em>) – The axes to plot the scatter plot on.</p></li>
<li><p><strong>x</strong> (<em>str</em><em>, </em><em>optional</em>) – The attribute to use for the x-axis.</p></li>
<li><p><strong>y</strong> (<em>str</em><em>, </em><em>optional</em>) – The attribute to use for the y-axis.</p></li>
<li><p><strong>color</strong> (<em>str</em><em>, </em><em>optional</em>) – The attribute to use for the color.</p></li>
<li><p><strong>alpha</strong> (<em>float</em><em>, </em><em>optional</em>) – The alpha value for the points.</p></li>
<li><p><strong>size</strong> (<em>float</em><em>, </em><em>optional</em>) – The size of the points.</p></li>
<li><p><strong>z_idx</strong> (<em>int</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Index of the z-plane to plot in the sorted list of z-planes in the SpotTable.
If None and <em>z_pos</em> is None, plot all z-slices.</p></li>
<li><p><strong>z_pos</strong> (<em>int</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Coordinates of the z-plane to plot.
If None and <em>z_pos</em> is None, plot all z-slices.</p></li>
<li><p><strong>palette</strong> (<em>str</em><em> or </em><em>list</em><em>, </em><em>optional</em>) – The palette to use for the colors.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.binned_expression_counts">
<span class="sig-name descname"><span class="pre">binned_expression_counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">binsize</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.binned_expression_counts" title="Link to this definition"></a></dt>
<dd><p>Return an array of spatially binned gene expression counts with shape (n_x_bins, n_y_bins, n_genes)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>binsize</strong> (<em>float</em>) – Size of each bin in um.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.reduced_expression_map">
<span class="sig-name descname"><span class="pre">reduced_expression_map</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">binsize</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umap_args=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">umap_ax=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm=&lt;ufunc</span> <span class="pre">'log1p'&gt;</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.reduced_expression_map" title="Link to this definition"></a></dt>
<dd><p>Show a UMAP of the binned expression counts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>binsize</strong> (<em>float</em>) – Size of each bin in um.</p></li>
<li><p><strong>umap_args</strong> (<em>dict</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Arguments to pass to UMAP.</p></li>
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – The axes to plot the UMAP on.</p></li>
<li><p><strong>umap_ax</strong> (<em>matplotlib.axes.Axes</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – The axes to plot the UMAP on.</p></li>
<li><p><strong>norm</strong> (<em>callable</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Function to normalize the expression counts.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple containing the binned expression counts, the x and y bins, and the UMAP coordinates.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.show_binned_heatmap">
<span class="sig-name descname"><span class="pre">show_binned_heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.show_binned_heatmap" title="Link to this definition"></a></dt>
<dd><p>Show an image of binned spot positions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em>) – The axes to plot the heatmap on.</p></li>
<li><p><strong>image_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Height/width of the image in pixels.</p></li>
<li><p><strong>bin_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Size of each bin in um. Alternative to image_size.</p></li>
<li><p><strong>log</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to take the log of the binned counts.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.add_image">
<span class="sig-name descname"><span class="pre">add_image</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">image</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.add_image" title="Link to this definition"></a></dt>
<dd><p>Attach an image to this dataset</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>image</strong> (<a class="reference internal" href="image.html#sis.image.ImageBase" title="sis.image.ImageBase"><em>ImageBase</em></a>) – The image to attach.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.get_image">
<span class="sig-name descname"><span class="pre">get_image</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">channel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frames</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.get_image" title="Link to this definition"></a></dt>
<dd><p>Return the image with the given name or channel name</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – The name of the image to return.</p></li>
<li><p><strong>channel</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – The channel name of the image to return.</p></li>
<li><p><strong>frames</strong> (<em>tuple</em><em> or </em><em>int</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – A tuple of the first (inclusive) and last (exclusive) indices of the frames (specific z planes) used to create the image, e.g. frames=(2,5) would create an image from z planes 2, 3, and 4.
else, if an int is provided, that specific z plane is used.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>selected_img</strong> – The image with the given name or channel name.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="image.html#sis.image.ImageBase" title="sis.image.ImageBase">ImageBase</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SpotTable.show_image">
<span class="sig-name descname"><span class="pre">show_image</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">channel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frames</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SpotTable.show_image" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – The axes to plot the image on. If None, a new figure is created.</p></li>
<li><p><strong>name</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – The name of the image to return.</p></li>
<li><p><strong>channel</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – The channel name of the image to return.</p></li>
<li><p><strong>frames</strong> (<em>tuple</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – A tuple of the first (inclusive) and last (exclusive) indices of the frames (specific z planes) used to create the image, e.g. frames=(2,5) would create an image from z planes 2, 3, and 4.
else, if an int is provided, that specific z plane is used.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="segmentedspottable">
<h2>SegmentedSpotTable<a class="headerlink" href="#segmentedspottable" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">sis.spot_table.</span></span><span class="sig-name descname"><span class="pre">SegmentedSpotTable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spot_table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable"><span class="pre">SpotTable</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_ids</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_labels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cl_to_cid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cid_to_cl</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_polygons</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seg_metadata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Represents a collection of cells for a spatial transcriptomics dataset
that has been segmented.</p>
<ul class="simple">
<li><p>Contains a SpotTable</p></li>
<li><p>Must contain cell_ids loaded from segmentation</p></li>
<li><p>May contain cell_polygons</p></li>
<li><p>May contain segmentation metadata</p></li>
</ul>
<dl class="py attribute">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.spot_table">
<span class="sig-name descname"><span class="pre">spot_table</span></span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.spot_table" title="Link to this definition"></a></dt>
<dd><p>The spot table associated with the dataset, containing information about transcript locations and gene ids.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable">SpotTable</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.cell_polygons">
<span class="sig-name descname"><span class="pre">cell_polygons</span></span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.cell_polygons" title="Link to this definition"></a></dt>
<dd><p>Polygons associated with each cell_id. Used to approximate the shapes of cells and measurements such as volume.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.seg_metadata">
<span class="sig-name descname"><span class="pre">seg_metadata</span></span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.seg_metadata" title="Link to this definition"></a></dt>
<dd><p>Metadata about segmentation, e.g. method, parameters, options. Will be saved in the cell by gene anndata uns if created.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable._old_cell_ids">
<span class="sig-name descname"><span class="pre">_old_cell_ids</span></span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable._old_cell_ids" title="Link to this definition"></a></dt>
<dd><p>Array of old cell ids before the most recent change. Used to update polygons when cell_ids are changed</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable._cell_ids">
<span class="sig-name descname"><span class="pre">_cell_ids</span></span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable._cell_ids" title="Link to this definition"></a></dt>
<dd><p>Array of integer cell ids per spot</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable._cell_labels">
<span class="sig-name descname"><span class="pre">_cell_labels</span></span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable._cell_labels" title="Link to this definition"></a></dt>
<dd><p>Uniformly increasing cell ids stored as a string. Allows for user prefix and suffix</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable._cl_to_cid">
<span class="sig-name descname"><span class="pre">_cl_to_cid</span></span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable._cl_to_cid" title="Link to this definition"></a></dt>
<dd><p>Dictionary mapping cell labels to cell ids</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable._cid_to_cl">
<span class="sig-name descname"><span class="pre">_cid_to_cl</span></span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable._cid_to_cl" title="Link to this definition"></a></dt>
<dd><p>Dictionary mapping cell ids to cell labels</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable._cell_index">
<span class="sig-name descname"><span class="pre">_cell_index</span></span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable._cell_index" title="Link to this definition"></a></dt>
<dd><p>Cached information about which transcripts belong to which cells, used to speed up lookups.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable._cell_bounds">
<span class="sig-name descname"><span class="pre">_cell_bounds</span></span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable._cell_bounds" title="Link to this definition"></a></dt>
<dd><p>Cached information about cell bounds, used to speed up lookups.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable._unique_cell_ids">
<span class="sig-name descname"><span class="pre">_unique_cell_ids</span></span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable._unique_cell_ids" title="Link to this definition"></a></dt>
<dd><p>Cached information about unique cell ids, used to speed up lookups.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spot_table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable"><span class="pre">SpotTable</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_ids</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_labels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cl_to_cid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cid_to_cl</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_polygons</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seg_metadata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.__init__" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spot_table</strong> (<a class="reference internal" href="#sis.spot_table.SpotTable" title="sis.spot_table.SpotTable"><em>SpotTable</em></a>) – The spot table associated with the dataset, containing information
about transcript locations and gene ids.</p></li>
<li><p><strong>cell_ids</strong> (<em>numpy.ndarray</em>) – Array of integer cell ids per spot.</p></li>
<li><p><strong>cell_labels</strong> (<em>numpy.ndarray</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Array of string cell labels per spot.</p></li>
<li><p><strong>cl_to_cid</strong> (<em>dict</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Mapping of cell labels to cell ids.</p></li>
<li><p><strong>cid_to_cl</strong> (<em>dict</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Mapping of cell ids to cell labels.</p></li>
<li><p><strong>cell_polygons</strong> (<em>dict</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Polygons associated with each cell_id. Used to approximate the shapes
of cells and measurements such as volume.</p></li>
<li><p><strong>seg_metadata</strong> (<em>dict</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Metadata about segmentation, e.g. method, parameters, options. Will be
saved in the cell by gene anndata uns if created.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.__getattr__">
<span class="sig-name descname"><span class="pre">__getattr__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.__getattr__" title="Link to this definition"></a></dt>
<dd><p>This method enables SegmentedSpotTable to act as a proxy for
attributes in SpotTable. If the user attempts to look up an attribute
that is not found in SegmentedSpotTable, this method will try to look
for this attribute in the attached SpotTable object instead.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.cell_ids">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">cell_ids</span></span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.cell_ids" title="Link to this definition"></a></dt>
<dd><p>An array of integer cell IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.cell_labels">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">cell_labels</span></span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.cell_labels" title="Link to this definition"></a></dt>
<dd><p>Cell labels are linked to cell_ids, but offer more descriptive naming possibilities. See generate_cell_labels()</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><ul class="simple">
<li><p><em>numpy.ndarray</em> – Array of cell labels</p></li>
<li><p><em>None</em> – If cell_labels have not been set</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.convert_cell_id">
<span class="sig-name descname"><span class="pre">convert_cell_id</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.convert_cell_id" title="Link to this definition"></a></dt>
<dd><p>Convert a cell id to a cell label and vice versa.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>cell_id</strong> (<em>int</em><em> or </em><em>str</em>) – The cell id (type int) or cell label (type str) to query.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>str</em> – If <em>cell_id</em> is an integer, returns the cell label as a string.</p></li>
<li><p><em>int</em> – If <em>cell_id</em> is a string, returns the cell id as an integer.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.convert_cell_label">
<span class="sig-name descname"><span class="pre">convert_cell_label</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_label</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.convert_cell_label" title="Link to this definition"></a></dt>
<dd><p>Alias for convert_cell_id to limit user confusion.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.cell_ids_changed">
<span class="sig-name descname"><span class="pre">cell_ids_changed</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.cell_ids_changed" title="Link to this definition"></a></dt>
<dd><p>Call when self.cell_ids has been modified to invalidate caches.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.unique_cell_ids">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">unique_cell_ids</span></span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.unique_cell_ids" title="Link to this definition"></a></dt>
<dd><p>Create and cache a numpy array of unique cell ids (excluding background)</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Array of unique cell ids excluding background (-1 and 0)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.generate_cell_labels">
<span class="sig-name descname"><span class="pre">generate_cell_labels</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.generate_cell_labels" title="Link to this definition"></a></dt>
<dd><p>Generates cell ids which count up from 1 to the total cell count rather than jumping between integers.
Cell labels are of type string to allow for concatenating a prefix and/or suffix to the id
If no prefix or suffix are specified, a UUID is used as a prefix to ensure that labels are unique</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>prefix</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – String to prepend to all cell labels</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – String to postpend to all cell labels</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.filter_cells">
<span class="sig-name descname"><span class="pre">filter_cells</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">real_cells</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_spot_count</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.filter_cells" title="Link to this definition"></a></dt>
<dd><p>Return a filtered spot table containing only cells matching the filter criteria.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>real_cells</strong> (<em>bool</em><em> or </em><em>None</em>) – If True, include only spots that are assigned to a cell (cell IDs &gt; 0)</p></li>
<li><p><strong>min_spot_count</strong> (<em>int</em><em> or </em><em>None</em>) – Include only spots that are assigned to cells with a minimum number of spots</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>filtered_table</strong> – A filtered spot table containing only cells matching the filter criteria.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable" title="sis.spot_table.SegmentedSpotTable">SegmentedSpotTable</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.cell_by_gene_dense_matrix">
<span class="sig-name descname"><span class="pre">cell_by_gene_dense_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dtype</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'uint16'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.cell_by_gene_dense_matrix" title="Link to this definition"></a></dt>
<dd><p>Return cell-by-gene data in a numpy array</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>dtype</strong> (<em>str</em><em>, </em><em>optional</em>) – The data type of the array.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>A numpy array of shape (n_cells, n_genes) containing the number of transcripts per cell.</p></li>
<li><p>The cell ids used to construct the matrix</p></li>
<li><p>The gene ids used to construct the matrix</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[numpy.ndarray, pandas.Index, pandas.Index]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.cell_by_gene_sparse_matrix">
<span class="sig-name descname"><span class="pre">cell_by_gene_sparse_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dtype</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'uint16'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.cell_by_gene_sparse_matrix" title="Link to this definition"></a></dt>
<dd><p>Return cell-by-gene data in a scipy.sparse.csr_matrix</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>dtype</strong> (<em>str</em><em>, </em><em>optional</em>) – The data type of the matrix.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p><strong>(cell_by_gene, cell_ids, gene_ids)</strong> –</p>
<ul class="simple">
<li><p>A sparse matrix of shape (n_cells, n_genes) containing the number of transcripts per cell.</p></li>
<li><p>The cell ids used to construct the matrix</p></li>
<li><p>The gene ids used to construct the matrix</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[scipy.sparse.csr_matrix, np.ndarray, np.ndarray]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.cell_by_gene_anndata">
<span class="sig-name descname"><span class="pre">cell_by_gene_anndata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x_format</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_dtype</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'uint16'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">additional_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.cell_by_gene_anndata" title="Link to this definition"></a></dt>
<dd><p>Return a cell-by-gene table in AnnData format.</p>
<ul class="simple">
<li><p>Obs includes: area/volume, cell transcipt centroids, cell polygon centroids, cell labels, cell ids</p></li>
<li><p>Var includes: probe name, num cells with reads, num segmented reads, num unsegmented reads</p></li>
<li><p>Uns includes: segmentation metadata, cell polygons, and SIS version</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x_format</strong> (<em>str</em>) – The format of the data matrix, either ‘dense’ or ‘sparse’.</p></li>
<li><p><strong>x_dtype</strong> (<em>str</em><em>, </em><em>optional</em>) – The data type of the matrix.</p></li>
<li><p><strong>additional_obs</strong> (<em>dict</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Additional columns to add to the anndata.obs DataFrame.
Accepts both {key : sequence} and {key : single_value} pairs.
Keys are column names and values are either arrays of the same length as the number of cells
OR a single value to be repeated for all cells.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>adata</strong> – An AnnData object containing the cell-by-gene data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>anndata.AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.cell_bounds">
<span class="sig-name descname"><span class="pre">cell_bounds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.cell_bounds" title="Link to this definition"></a></dt>
<dd><p>Return ((xmin, xmax), (ymin, ymax)) for <em>cell_id</em></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>cell_id</strong> (<em>int</em><em> or </em><em>str</em>) – The cell id (type int) or cell label (type str) to query.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The bounds of the cell formatted as ((xmin, xmax), (ymin, ymax))</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.cell_centroids">
<span class="sig-name descname"><span class="pre">cell_centroids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">use_cell_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.cell_centroids" title="Link to this definition"></a></dt>
<dd><p>Return a Pandas DataFrame of cell centroids calculated as the means of the x,y,z coordinates for each cell</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>use_cell_labels</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use cell labels as the index</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame of cell centroids with columns [‘x’, ‘y’, ‘z’]</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.cell_polygon">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">cell_polygon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_points_array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_inv</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.cell_polygon" title="Link to this definition"></a></dt>
<dd><p>Get 2D alpha shape from a set of points, modified slightly from
<a class="reference external" href="http://blog.thehumangeo.com/2014/05/12/drawing-boundaries-in-python/">http://blog.thehumangeo.com/2014/05/12/drawing-boundaries-in-python/</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cell_points</strong> (<em>np.ndarray</em>) – numpy array of point locations with expected columns [x,y].</p></li>
<li><p><strong>alpha_inv</strong> (<em>float</em>) – parameter that sets the radius filter on the Delaunay triangulation.
Traditionally alpha is defined as 1/radius, and here the function input is inverted for slightly more intuitive use</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>tp</strong> – The alpha shape of the points.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>shapely.geometry.Polygon</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.calculate_optimal_polygon">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">calculate_optimal_polygon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">xy_pos</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_inv</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_inv_coeff</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.3333333333333333</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.calculate_optimal_polygon" title="Link to this definition"></a></dt>
<dd><p>Calculate the optimal polygon for a set of points by increasing alpha_inv until a single polygon containing all points is generated.
After the alpha_inv is decided, the alpha_inv_coeff is applied to get promote a more natural cell shape</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>xy_pos</strong> (<em>np.ndarray</em>) – numpy array of point locations with expected columns [x,y].</p></li>
<li><p><strong>alpha_inv</strong> (<em>float</em>) – parameter that sets the radius filter on the Delaunay triangulation.
Traditionally alpha is defined as 1/radius, and here the function input is inverted for slightly more intuitive use</p></li>
<li><p><strong>alpha_inv_coeff</strong> (<em>float</em><em>, </em><em>optional</em>) – coefficient to apply to alpha_inv to get a more natural cell shape.
Default is 4/3 as we found that to give good results empircally</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>shapely.geometry.Polygon</em> – The polygon which encompasses the given xy coordinate</p></li>
<li><p><em>None</em> – If a polygon cannot be created from the given xy coordinates (e.g. if there are &lt;= 3 points)</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.calculate_cell_polygons">
<span class="sig-name descname"><span class="pre">calculate_cell_polygons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alpha_inv</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">separate_z_planes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cells_to_run</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_inv_coeff</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.3333333333333333</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disable_tqdm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.calculate_cell_polygons" title="Link to this definition"></a></dt>
<dd><p>Calculate the cell polygons for each cell id in self.cell_ids. Add these to the self.cell_polygons dictionary</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alpha_inv</strong> (<em>float</em><em>, </em><em>optional</em>) – Parameter that sets the radius filter on the Delaunay triangulation.
Traditionally alpha is defined as 1/radius, and here the function input is inverted for slightly more intuitive use</p></li>
<li><p><strong>separate_z_planes</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, calculate separate polygons for each z-plane.</p></li>
<li><p><strong>cells_to_run</strong> (<em>np.ndarray</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – If not None, only calculate polygons for these cell ids.</p></li>
<li><p><strong>alpha_inv_coeff</strong> (<em>float</em><em>, </em><em>optional</em>) – Coefficient to apply to alpha_inv to get a more natural cell shape
Default is 4/3 as we found that to give good results empircally</p></li>
<li><p><strong>disable_tqdm</strong> (<em>bool</em>) – If True, disable the progress bar.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.calculate_cell_features">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">calculate_cell_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_polygon</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">z_plane_thickness</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.5</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.calculate_cell_features" title="Link to this definition"></a></dt>
<dd><p>Calculate the area and centroid of a cell polygon.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cell_polygon</strong> (<em>shapely.geometry.Polygon</em>) – The polygon to calculate features for.</p></li>
<li><p><strong>z_plane_thickness</strong> (<em>float</em><em>, </em><em>optional</em>) – The thickness of each z-plane.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary containing the area/volume, center_x, center_y, and potentially center_z of the cell polygon.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.calculate_all_cell_features">
<span class="sig-name descname"><span class="pre">calculate_all_cell_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">z_plane_thickness</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_cell_labels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_both_tags</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disable_tqdm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.calculate_all_cell_features" title="Link to this definition"></a></dt>
<dd><p>Calculate the cell features for each cell with polygons. Return them in a pandas DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>z_plane_thickness</strong> (<em>float</em><em>, </em><em>optional</em>) – The thickness of each z-plane.</p></li>
<li><p><strong>use_cell_labels</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to replace cell ids with cell labels in the DataFrame</p></li>
<li><p><strong>use_both_ids</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to include both cell ids and cell labels in the DataFrame</p></li>
<li><p><strong>disable_tqdm</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, disable the progress bar.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>df</strong> – A DataFrame of cell features with columns [‘cell_id’, ‘volume’, ‘area’, ‘polygon_center_x’, ‘polygon_center_y’, ‘polygon_center_z’]</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.get_geojson_collection">
<span class="sig-name descname"><span class="pre">get_geojson_collection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">use_cell_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.get_geojson_collection" title="Link to this definition"></a></dt>
<dd><p>Create a geojson feature collection from the cell polygons</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>use_cell_labels</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use cell labels as the index</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A geojson feature collection of the cell polygons</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>geojson.FeatureCollection</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.save_cell_polygons">
<span class="sig-name descname"><span class="pre">save_cell_polygons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">save_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_cell_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.save_cell_polygons" title="Link to this definition"></a></dt>
<dd><p>Save a geojson geometry collection from the cell polygons</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>save_path</strong> (<em>Path</em><em> or </em><em>str</em><em>, </em><em>optional</em>) – Path to save the file to. Can be either ‘.geojson’ or ‘.pkl’</p></li>
<li><p><strong>use_cell_labels</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use cell labels as the index</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.load_cell_polygons">
<span class="sig-name descname"><span class="pre">load_cell_polygons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">load_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_ids</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reset_cache</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disable_tqdm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.load_cell_polygons" title="Link to this definition"></a></dt>
<dd><p>Load cell polygons from a geojson feature collection file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>load_path</strong> (<em>Path</em><em> or </em><em>str</em><em>, </em><em>optional</em>) – Path to the file containing the cell polygons. Can be either ‘.geojson’ or ‘.pkl’</p></li>
<li><p><strong>cell_ids</strong> (<em>Path</em><em> or </em><em>str</em><em> or </em><em>list</em><em> or </em><em>np.ndarray</em><em> or </em><em>None</em><em>, </em><em>optional</em><em> [</em><em>BACKWARDS COMPATIBILITY USAGE ONLY</em><em>]</em>) – Path/str to .npy file containing cell ids corresponding to cell polygons for GeometryCollection
list/np.ndarray containing cell ids corresponding to cell polygons for GeometryCollection
Used to assign polygons to cell ids for GeometryCollection b/c it does not store IDs.
This is important if you want to load in a bunch of 2D polygons files w/o reseting the cache as we have no way to determine proper ID
If not provided and using GeometryCollection, it is assumed the cell polygons are in order of the sorted cell ids
GeometryCollections may contain None for uncalculated polygons or may exclude them entirely</p></li>
<li><p><strong>reset_cache</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, reset the stored cell polygons before loading</p></li>
<li><p><strong>disable_tqdm</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, disable the progress bar.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.cell_indices">
<span class="sig-name descname"><span class="pre">cell_indices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_ids</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.cell_indices" title="Link to this definition"></a></dt>
<dd><p>Return indices giving table location of all spots with <em>cell_ids</em></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>cell_ids</strong> (<em>int</em><em> or </em><em>str</em><em> or </em><em>np.ndarray</em>) – The cell ids (type int) or cell labels (type str) to query.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An array of indices corresponding to the spots with the specified cell ids.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.cells_inside_region">
<span class="sig-name descname"><span class="pre">cells_inside_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">xlim</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylim</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.cells_inside_region" title="Link to this definition"></a></dt>
<dd><p>Return IDs of cells that are entirely inside xlim, ylim.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>xlim</strong> (<em>tuple</em>) – The x limits of the region.</p></li>
<li><p><strong>ylim</strong> (<em>tuple</em>) – The y limits of the region.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>cell_ids</strong> – A list of cell ids that are entirely inside the region.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.cell_mask">
<span class="sig-name descname"><span class="pre">cell_mask</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_ids</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.cell_mask" title="Link to this definition"></a></dt>
<dd><p>Return a mask selecting spots that belong to cells in <em>cell_ids</em></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>cell_ids</strong> (<em>int</em><em> or </em><em>str</em><em> or </em><em>np.ndarray</em>) – The cell ids (type int) or cell labels (type str) to query.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>mask</strong> – A boolean mask selecting spots that belong to cells in <em>cell_ids</em>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.cell_indices_within_padding">
<span class="sig-name descname"><span class="pre">cell_indices_within_padding</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">padding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5.0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.cell_indices_within_padding" title="Link to this definition"></a></dt>
<dd><p>Return spot indices all cells that do not come within <em>padding</em> distance of the parent_region bounds.
This is used to exclude cells near the edge of a tile, where the segmentation becomes unreliable.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>padding</strong> (<em>float</em><em>, </em><em>optional</em>) – The padding distance.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>include_inds</strong> – An array of spot indices that do not come within <em>padding</em> distance of the parent_region bounds.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.merge_cells">
<span class="sig-name descname"><span class="pre">merge_cells</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">padding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">union_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.merge_cells" title="Link to this definition"></a></dt>
<dd><p>Merge cell IDs from SpotTable <em>other</em> into self.
Returns a structure describing merge conflicts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#sis.spot_table.SegmentedSpotTable" title="sis.spot_table.SegmentedSpotTable"><em>SegmentedSpotTable</em></a>) – The other SpotTable to merge with.</p></li>
<li><p><strong>padding</strong> (<em>float</em><em>, </em><em>optional</em>) – The padding distance.</p></li>
<li><p><strong>union_threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – The threshold for merging cells based on overlap.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>conflicts</strong> – A list of dictionaries describing merge conflicts.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.set_cell_ids_from_tiles">
<span class="sig-name descname"><span class="pre">set_cell_ids_from_tiles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tiles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">padding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.set_cell_ids_from_tiles" title="Link to this definition"></a></dt>
<dd><p>Overwrite all cell IDs by merging from <em>tiles</em>, which may be a list of SegmentedSpotTable
or SegmentationResult instances.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tiles</strong> (<em>list</em>) – A list of SegmentedSpotTable or SegmentationResult instances.</p></li>
<li><p><strong>padding</strong> (<em>float</em><em>, </em><em>optional</em>) – The padding distance.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>merge_results</strong> – A list of dictionaries describing merge conflicts.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.load_merscope_cell_ids">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_merscope_cell_ids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">csv_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_rows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.load_merscope_cell_ids" title="Link to this definition"></a></dt>
<dd><p>Load the original segmentation for a MERSCOPE dataset.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>csv_file</strong> (<em>str</em>) – Path to the detected transcripts file.</p></li>
<li><p><strong>max_rows</strong> (<em>int</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Maximum number of rows to load from the CSV file.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>cell_ids</strong> – An array of cell ids from the MERSCOPE dataset.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.load_merscope">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_merscope</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">csv_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_rows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.load_merscope" title="Link to this definition"></a></dt>
<dd><p>Load MERSCOPE data from a detected transcripts CSV file, including
the original segmentation. If you are resegmenting the data, prefer
SpotTable.load_merscope.</p>
<p>Note: If cache_file is set, only the raw spot table is cached, not the
cell_ids. This is for consistency with SpotTable.load_merscope.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>csv_file</strong> (<em>str</em>) – Path to the detected transcripts file.</p></li>
<li><p><strong>cache_file</strong> (<em>str</em><em> or </em><em>None</em>) – Path to the detected transcripts cache file, which is an npz file
representing the raw SpotTable (without cell_ids). If passed, will
create a cache file if one does not already exists.</p></li>
<li><p><strong>image_path</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Path to directory containing a MERSCOPE image stack.</p></li>
<li><p><strong>max_rows</strong> (<em>int</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Maximum number of rows to load from the CSV file.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable" title="sis.spot_table.SegmentedSpotTable">sis.spot_table.SegmentedSpotTable</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.load_xenium">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_xenium</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">transcript_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_rows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">z_depth</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">3.0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.load_xenium" title="Link to this definition"></a></dt>
<dd><p>Load Xenium data from a detected transcripts CSV file, including
the original segmentation. If you are resegmenting the data, prefer
SpotTable.load_xenium.</p>
<p>Note: If cache_file is set, only the raw spot table is cached, not the
cell_ids. This is for consistency with SpotTable.load_xenium.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>transcript_file</strong> (<em>str</em>) – Path to the detected transcripts file.</p></li>
<li><p><strong>cache_file</strong> (<em>str</em><em> or </em><em>None</em>) – Path to the detected transcripts cache file, which is an npz file
representing the raw SpotTable (without cell_ids). If passed, will
create a cache file if one does not already exists.</p></li>
<li><p><strong>image_path</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Path to directory containing a Xenium image stack.</p></li>
<li><p><strong>max_rows</strong> (<em>int</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Maximum number of rows to load from the CSV file.</p></li>
<li><p><strong>z_depth</strong> (<em>float</em><em>, </em><em>optional</em>) – Depth (in um) of a imaging layer i.e. z-plane
Used to bin z-positions into discrete planes</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable" title="sis.spot_table.SegmentedSpotTable">sis.spot_table.SegmentedSpotTable</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.load_stereoseq">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_stereoseq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gem_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gem_cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">tuple</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">(('gene',</span> <span class="pre">0),</span> <span class="pre">('x',</span> <span class="pre">1),</span> <span class="pre">('y',</span> <span class="pre">2),</span> <span class="pre">('MIDcounts',</span> <span class="pre">3))</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skiprows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_rows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image_channel</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.load_stereoseq" title="Link to this definition"></a></dt>
<dd><p>Load StereoSeq data from gem file. This can be slow so optionally cache the result to a .npz file.
1/19/2023: New StereoSeq data has cell_ids, add optional cell_cols to add to SpotTable. Also has a flag
for whether the spot was in the main cell or the extended cell (in_cell)
1/3/2024: new cellbin.gem doesn’t have <cite>in_cell</cite>, let’s not worry about it and just specify the column
that the cell ID is in</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.load_npz">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_npz</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">npz_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">images</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="image.html#sis.image.ImageBase" title="sis.image.ImageBase"><span class="pre">ImageBase</span></a><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference internal" href="image.html#sis.image.ImageBase" title="sis.image.ImageBase"><span class="pre">ImageBase</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_pickle</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.load_npz" title="Link to this definition"></a></dt>
<dd><p>Load from an NPZ file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>npz_file</strong> (<em>str</em>) – Path to the npz file.</p></li>
<li><p><strong>images</strong> (<a class="reference internal" href="image.html#sis.image.ImageBase" title="sis.image.ImageBase"><em>ImageBase</em></a><em> or </em><em>list</em><em>[</em><a class="reference internal" href="image.html#sis.image.ImageBase" title="sis.image.ImageBase"><em>ImageBase</em></a><em>] or </em><em>None</em><em>, </em><em>optional</em>) – Image(s) to attach to the SpotTable. Must be loaded separately
since these cannot be stored in the NPZ file.</p></li>
<li><p><strong>allow_pickle</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to allow loading pickled object arrays stored in npy files.
Must be enabled to load dictionaries and cell polygons.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable" title="sis.spot_table.SegmentedSpotTable">sis.spot_table.SegmentedSpotTable</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.save_npz">
<span class="sig-name descname"><span class="pre">save_npz</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">npz_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.save_npz" title="Link to this definition"></a></dt>
<dd><p>Save to an NPZ file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>npz_file</strong> (<em>str</em>) – Output path for the npz file.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.dataframe">
<span class="sig-name descname"><span class="pre">dataframe</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['x',</span> <span class="pre">'y',</span> <span class="pre">'z',</span> <span class="pre">'gene_ids',</span> <span class="pre">'cell_ids']</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.dataframe" title="Link to this definition"></a></dt>
<dd><p>Return a dataframe containing the specified columns.
By default, columns are x, y, z, gene_ids, and cell_ids.
Also available: gene_names</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>cols</strong> (<em>list</em><em>, </em><em>optional</em>) – List of columns to include in the dataframe.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>df</strong> – A pandas dataframe containing the specified columns.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.cell_scatter_plot">
<span class="sig-name descname"><span class="pre">cell_scatter_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.cell_scatter_plot" title="Link to this definition"></a></dt>
<dd><p>Scatter plot of spots colored by cell ID</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>*args</strong> (<em>tuple</em>) – positional arguments for SegmentedSpotTable.scatter_plot()</p></li>
<li><p><strong>**kwds</strong> (<em>dict</em><em>[</em><em>str</em><em>, </em><em>Any</em><em>]</em><em>, </em><em>optional</em>) – keyword arguments for SegmentedSpotTable.scatter_plot()</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.scatter_plot">
<span class="sig-name descname"><span class="pre">scatter_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'x'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'y'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gene_ids'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">z_idx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">z_pos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">palette</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_polygons</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.scatter_plot" title="Link to this definition"></a></dt>
<dd><p>Plot a scatter plot of the spots in this table colored by <em>color</em></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em><em>, </em><em>optional</em>) – The axes to plot the scatter plot on.</p></li>
<li><p><strong>x</strong> (<em>str</em><em>, </em><em>optional</em>) – The attribute to use for the x-axis.</p></li>
<li><p><strong>y</strong> (<em>str</em><em>, </em><em>optional</em>) – The attribute to use for the y-axis.</p></li>
<li><p><strong>color</strong> (<em>str</em><em>, </em><em>optional</em>) – The attribute to use for the color.</p></li>
<li><p><strong>alpha</strong> (<em>float</em><em>, </em><em>optional</em>) – The alpha value for the points.</p></li>
<li><p><strong>size</strong> (<em>float</em><em>, </em><em>optional</em>) – The size of the points.</p></li>
<li><p><strong>z_idx</strong> (<em>int</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Index of the z-plane to plot in the sorted list of z-planes in the SpotTable.
If None and <em>z_pos</em> is None, plot all z-slices.</p></li>
<li><p><strong>z_pos</strong> (<em>int</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Coordinates of the z-plane to plot.
If None and <em>z_pos</em> is None, plot all z-slices.</p></li>
<li><p><strong>palette</strong> (<em>str</em><em> or </em><em>list</em><em>, </em><em>optional</em>) – The palette to use for the colors.</p></li>
<li><p><strong>show_polygons</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, show cell polygons in the plot. If polygons are 3d, only works for individual z-planes</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.cell_palette">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">cell_palette</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cells</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.cell_palette" title="Link to this definition"></a></dt>
<dd><p>Generate a color palette suitable for distinguisging individual cells</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>cells</strong> (<em>list</em>) – List of cell ids.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>palette</strong> – A dictionary mapping cell ids to colors.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">deep</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.copy" title="Link to this definition"></a></dt>
<dd><p>Return a copy of self, optionally with some attributes replaced.
Currently doesn’t support replacing attributes of SpotTable (I’m not
sure what the use case for this would be) but this could be added if
needed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>deep</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, make a deep copy of the attributes.</p></li>
<li><p><strong>**kwds</strong> (<em>dict</em><em>[</em><em>str</em><em>, </em><em>Any</em><em>]</em><em>, </em><em>optional</em>) – Attributes to replace in the copied SegmentedSpotTable.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A copy of self with the specified attributes replaced.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable" title="sis.spot_table.SegmentedSpotTable">SegmentedSpotTable</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.get_subregion">
<span class="sig-name descname"><span class="pre">get_subregion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">xlim</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylim</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">incl_end</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.get_subregion" title="Link to this definition"></a></dt>
<dd><p>Return a SegmentedSpotTable including the subset of this table inside the region xlim, ylim</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>xlim</strong> (<em>tuple</em>) – The x limits of the region.</p></li>
<li><p><strong>ylim</strong> (<em>tuple</em>) – The y limits of the region.</p></li>
<li><p><strong>incl_end</strong> (<em>bool</em><em>, </em><em>optional</em>) – Include all pixels of the image that overlap with the region, rather than just those inside the region.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>seg_subtable</strong> – A SegmentedSpotTable including the subset of this table inside the region xlim, ylim.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable" title="sis.spot_table.SegmentedSpotTable">SegmentedSpotTable</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="sis.spot_table.SegmentedSpotTable.save_xenium_kit_cbg">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">save_xenium_kit_cbg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">expt_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_rows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">z_depth</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">3.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_format</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'sparse'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">additional_obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#sis.spot_table.SegmentedSpotTable.save_xenium_kit_cbg" title="Link to this definition"></a></dt>
<dd><p>Takes the results of a xenium experiment segmented with the Xenium segmentation kit
and saves them into the SIS standard format</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>expt_dir</strong> (<em>Path</em><em> or </em><em>str</em>) – Path to the Xenium experiment directory containing the xenium experiment and its segmentation results</p></li>
<li><p><strong>output_file</strong> (<em>Path</em><em> or </em><em>str</em>) – Path to save the Anndata to</p></li>
<li><p><strong>max_rows</strong> (<em>int</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Maximum number of rows to load from the transcripts file</p></li>
<li><p><strong>z_depth</strong> (<em>float</em><em>, </em><em>optional</em>) – Depth (in um) of a imaging layer i.e. z-plane
Used to bin z-positions into discrete planes</p></li>
<li><p><strong>x_format</strong> (<em>str</em><em>, </em><em>optional</em>) – The format of the data matrix (X in the anndata), either ‘dense’ or ‘sparse’.</p></li>
<li><p><strong>additional_obs</strong> (<em>dict</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Additional columns to add to the anndata.obs DataFrame.
Keys are column names and values are arrays of the same length as the number of cells.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>seg_subtable</strong> – A SegmentedSpotTable including the subset of this table inside the region xlim, ylim.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#sis.spot_table.SegmentedSpotTable" title="sis.spot_table.SegmentedSpotTable">SegmentedSpotTable</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="image.html" class="btn btn-neutral float-right" title="image" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Jacob Quon, Madeleine Hewitt, Luke Campagnola, Brian Long, Stephanie Seeman.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>